<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Contributing" href="contributing.html" /><link rel="prev" title="scranpy" href="index.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2023.09.10 -->
        <title>scran, in Python - scranpy unknown documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">scranpy unknown documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">scranpy unknown documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="api/modules.html">Module Reference</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="api/scranpy.html">scranpy package</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of scranpy package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.aggregation.html">scranpy.aggregation package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.analyze.html">scranpy.analyze package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.batch_correction.html">scranpy.batch_correction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.clustering.html">scranpy.clustering package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.dimensionality_reduction.html">scranpy.dimensionality_reduction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.feature_selection.html">scranpy.feature_selection package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.feature_set_enrichment.html">scranpy.feature_set_enrichment package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.marker_detection.html">scranpy.marker_detection package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.nearest_neighbors.html">scranpy.nearest_neighbors package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.normalization.html">scranpy.normalization package</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/scranpy.quality_control.html">scranpy.quality_control package</a></li>
</ul>
</li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <!-- These are examples of badges you might want to add to your README:
     please update the URLs accordingly

[![Built Status](https://api.cirrus-ci.com/github/<USER>/scranpy.svg?branch=main)](https://cirrus-ci.com/github/<USER>/scranpy)
[![ReadTheDocs](https://readthedocs.org/projects/scranpy/badge/?version=latest)](https://scranpy.readthedocs.io/en/stable/)
[![Coveralls](https://img.shields.io/coveralls/github/<USER>/scranpy/main.svg)](https://coveralls.io/r/<USER>/scranpy)
[![Conda-Forge](https://img.shields.io/conda/vn/conda-forge/scranpy.svg)](https://anaconda.org/conda-forge/scranpy)
[![Twitter](https://img.shields.io/twitter/url/http/shields.io.svg?style=social&label=Twitter)](https://twitter.com/scranpy)
-->
<p><a class="reference external" href="https://pyscaffold.org/"><img alt="Project generated with PyScaffold" src="https://img.shields.io/badge/-PyScaffold-005CA0?logo=pyscaffold" /></a>
<a class="reference external" href="https://pypi.org/project/scranpy/"><img alt="PyPI-Server" src="https://img.shields.io/pypi/v/scranpy.svg" /></a>
<a class="reference external" href="https://pepy.tech/project/scranpy"><img alt="Downloads" src="https://static.pepy.tech/badge/scranpy/month" /></a>
<img alt="Unit tests" src="https://github.com/BiocPy/scranpy/actions/workflows/pypi-test.yml/badge.svg" /></p>
<section class="tex2jax_ignore mathjax_ignore" id="scran-in-python">
<h1>scran, in Python<a class="headerlink" href="#scran-in-python" title="Permalink to this heading">#</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">#</a></h2>
<p>The <strong>scranpy</strong> package provides Python bindings to the single-cell analysis methods in <a class="reference external" href="https://github.com/LTLA/libscran"><strong>libscran</strong></a> and related C++ libraries.
It performs the standard steps in a typical single-cell analysis including quality control, normalization, feature selection, dimensionality reduction, clustering and marker detection.
<strong>scranpy</strong> makes heavy use of the <a class="reference external" href="https://github.com/BiocPy">BiocPy</a> data structures in its user interface,
while it uses the <a class="reference external" href="https://pypi.org/project/mattress"><strong>mattress</strong></a> package to provide a C++ representation of the underlying matrix data.
This package is effectively a mirror of its counterparts in Javascript (<a class="reference external" href="https://npmjs.com/package/scran.js"><strong>scran.js</strong></a>) and R (<a class="reference external" href="https://github.com/LTLA/scran.chan"><strong>scran.chan</strong></a>),
which are based on the same underlying C++ libraries and concepts.</p>
</section>
<section id="quick-start">
<h2>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this heading">#</a></h2>
<p>Let’s load in the famous PBMC 4k dataset from 10X Genomics (available <a class="reference external" href="https://github.com/kanaverse/random-test-files/releases/tag/10x-pbmc-v1.0.0">here</a>):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">singlecellexperiment</span>
<span class="n">sce</span> <span class="o">=</span> <span class="n">singlecellexperiment</span><span class="o">.</span><span class="n">read_tenx_h5</span><span class="p">(</span><span class="s2">&quot;pbmc4k-tenx.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Then we just need to call one of <strong>scranpy</strong>’s <code class="docutils literal notranslate"><span class="pre">analyze()</span></code> functions.
(We do have to tell it what the mitochondrial genes are, though.)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scranpy</span>
<span class="n">options</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">AnalyzeOptions</span><span class="p">()</span>
<span class="n">options</span><span class="o">.</span><span class="n">per_cell_rna_qc_metrics_options</span><span class="o">.</span><span class="n">subsets</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">guess_mito_from_symbols</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="s2">&quot;mt-&quot;</span><span class="p">)</span>
<span class="p">}</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze_sce</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">options</span><span class="o">=</span><span class="n">options</span><span class="p">)</span>
</pre></div>
</div>
<p>This will perform all of the usual steps for a routine single-cell analysis,
as described in Bioconductor’s <a class="reference external" href="https://bioconductor.org/books/OSCA">Orchestrating single cell analysis</a> book.
It returns an object containing clusters, t-SNEs, UMAPs, marker genes, and so on:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">results</span><span class="o">.</span><span class="n">clusters</span>
<span class="n">results</span><span class="o">.</span><span class="n">tsne</span>
<span class="n">results</span><span class="o">.</span><span class="n">umap</span>
<span class="n">results</span><span class="o">.</span><span class="n">rna_markers</span>
</pre></div>
</div>
<p>We won’t go over the theory here as it’s explained more thoroughly in the book.
Check out the <a class="reference external" href="https://biocpy.github.io/scranpy">reference documentation</a> for more details.</p>
</section>
<section id="multiple-batches">
<h2>Multiple batches<a class="headerlink" href="#multiple-batches" title="Permalink to this heading">#</a></h2>
<p>To demonstrate, let’s grab two batches of PBMC datasets from 10X Genomics (again, available <a class="reference external" href="https://github.com/kanaverse/random-test-files/releases/tag/10x-pbmc-v1.0.0">here</a>):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">singlecellexperiment</span>
<span class="n">sce3k</span> <span class="o">=</span> <span class="n">singlecellexperiment</span><span class="o">.</span><span class="n">read_tenx_h5</span><span class="p">(</span><span class="s2">&quot;pbmc3k-tenx.h5&quot;</span><span class="p">)</span>
<span class="n">sce4k</span> <span class="o">=</span> <span class="n">singlecellexperiment</span><span class="o">.</span><span class="n">read_tenx_h5</span><span class="p">(</span><span class="s2">&quot;pbmc4k-tenx.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>They don’t have the same features, so we’ll just take the intersection of their Ensembl IDs before combining them:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">biocutils</span>
<span class="n">common</span> <span class="o">=</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">sce3k</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">],</span> <span class="n">sce4k</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">])</span>
<span class="n">sce3k_common</span> <span class="o">=</span> <span class="n">sce3k</span><span class="p">[</span><span class="n">biocutils</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">common</span><span class="p">,</span> <span class="n">sce3k</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]),</span> <span class="p">:]</span>
<span class="n">sce4k_common</span> <span class="o">=</span> <span class="n">sce4k</span><span class="p">[</span><span class="n">biocutils</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">common</span><span class="p">,</span> <span class="n">sce4k</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]),</span> <span class="p">:]</span>

<span class="kn">import</span> <span class="nn">scipy.sparse</span>
<span class="n">combined</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">sce3k_common</span><span class="o">.</span><span class="n">assay</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">sce4k_common</span><span class="o">.</span><span class="n">assay</span><span class="p">(</span><span class="mi">0</span><span class="p">)))</span>
<span class="n">batch</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;3k&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">sce3k_common</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;4k&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">sce4k_common</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>We can now perform a batch-aware analysis:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scranpy</span>
<span class="n">options</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">AnalyzeOptions</span><span class="p">()</span>
<span class="n">options</span><span class="o">.</span><span class="n">per_cell_rna_qc_metrics_options</span><span class="o">.</span><span class="n">subsets</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">guess_mito_from_symbols</span><span class="p">(</span><span class="n">sce3k_common</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="s2">&quot;mt-&quot;</span><span class="p">)</span>
<span class="p">}</span>
<span class="n">options</span><span class="o">.</span><span class="n">miscellaneous_options</span><span class="o">.</span><span class="n">block</span> <span class="o">=</span> <span class="n">batch</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze</span><span class="p">(</span><span class="n">combined</span><span class="p">,</span> <span class="n">options</span><span class="o">=</span><span class="n">options</span><span class="p">)</span>
</pre></div>
</div>
<p>This yields mostly the same set of results as before, but with an extra MNN-corrected embedding for clustering, visualization, etc.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">results</span><span class="o">.</span><span class="n">mnn</span>
</pre></div>
</div>
</section>
<section id="multiple-modalities">
<h2>Multiple modalities<a class="headerlink" href="#multiple-modalities" title="Permalink to this heading">#</a></h2>
<p>Let’s grab a 10X Genomics immune profiling dataset (see <a class="reference external" href="https://github.com/kanaverse/random-test-files/releases/download/10x-immune-v1.0.0/immune_3.0.0-tenx.h5">here</a>):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">singlecellexperiment</span>
<span class="n">sce</span> <span class="o">=</span> <span class="n">singlecellexperiment</span><span class="o">.</span><span class="n">read_tenx_h5</span><span class="p">(</span><span class="s2">&quot;immune_3.0.0-tenx.h5&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We need to split it to genes and ADTs:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">is_gene</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Gene Expression&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;feature_type&quot;</span><span class="p">]]</span>
<span class="n">gene_data</span> <span class="o">=</span> <span class="n">sce</span><span class="p">[</span><span class="n">is_gene</span><span class="p">,:]</span>
<span class="n">is_adt</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Antibody Capture&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;feature_type&quot;</span><span class="p">]]</span>
<span class="n">adt_data</span> <span class="o">=</span> <span class="n">sce</span><span class="p">[</span><span class="n">is_adt</span><span class="p">,:]</span>
</pre></div>
</div>
<p>And now we can run the analysis:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scranpy</span>
<span class="n">options</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">AnalyzeOptions</span><span class="p">()</span>
<span class="n">options</span><span class="o">.</span><span class="n">per_cell_rna_qc_metrics_options</span><span class="o">.</span><span class="n">subsets</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">guess_mito_from_symbols</span><span class="p">(</span><span class="n">gene_data</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="s2">&quot;mt-&quot;</span><span class="p">)</span>
<span class="p">}</span>
<span class="n">options</span><span class="o">.</span><span class="n">per_cell_adt_qc_metrics_options</span><span class="o">.</span><span class="n">subsets</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;igg&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">n</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;igg&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">adt_data</span><span class="o">.</span><span class="n">row_data</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span>
<span class="p">}</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze_se</span><span class="p">(</span><span class="n">gene_data</span><span class="p">,</span> <span class="n">adt_se</span> <span class="o">=</span> <span class="n">adt_data</span><span class="p">,</span> <span class="n">options</span><span class="o">=</span><span class="n">options</span><span class="p">)</span>
</pre></div>
</div>
<p>This returns ADT-specific results in the relevant fields, as well as a set of combined PCs for use in clustering, visualization, etc.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">results</span><span class="o">.</span><span class="n">adt_size_factors</span>
<span class="n">results</span><span class="o">.</span><span class="n">adt_markers</span>
<span class="n">results</span><span class="o">.</span><span class="n">combined_pcs</span>
</pre></div>
</div>
</section>
<section id="customizing-the-analysis">
<h2>Customizing the analysis<a class="headerlink" href="#customizing-the-analysis" title="Permalink to this heading">#</a></h2>
<p>Most parameters can be changed by setting the relevant fields in the <code class="docutils literal notranslate"><span class="pre">AnalyzeOptions</span></code> object.
For example, we can modify the number of neighbors and resolution used for graph-based clustering:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">options</span><span class="o">.</span><span class="n">build_snn_graph_options</span><span class="o">.</span><span class="n">num_neighbors</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">options</span><span class="o">.</span><span class="n">miscellaneous_options</span><span class="o">.</span><span class="n">snn_graph_multilevel_resolution</span> <span class="o">=</span> <span class="mi">2</span>
</pre></div>
</div>
<p>Or we can fiddle the the various dimensionality reduction parameters:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">options</span><span class="o">.</span><span class="n">run_pca_options</span><span class="o">.</span><span class="n">rank</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">options</span><span class="o">.</span><span class="n">run_tsne_options</span><span class="o">.</span><span class="n">perplexity</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">options</span><span class="o">.</span><span class="n">run_umap_options</span><span class="o">.</span><span class="n">min_dist</span> <span class="o">=</span> <span class="mf">0.5</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">AnalyzeOptions</span></code> has a few convenience methods to easily set the same parameter across multiple <code class="docutils literal notranslate"><span class="pre">*_options</span></code> attributes.
For example, to enable parallel processing in every step:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">options</span><span class="o">.</span><span class="n">set_threads</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<p>Advanced users can even obtain the sequence of steps used internally by <code class="docutils literal notranslate"><span class="pre">analyze()</span></code> by calling it with <code class="docutils literal notranslate"><span class="pre">dry_run</span> <span class="pre">=</span> <span class="pre">True</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">commands</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">dry_run</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">commands</span><span class="p">)</span>
<span class="c1">## import scranpy</span>
<span class="c1">## import numpy</span>
<span class="c1">## </span>
<span class="c1">## results = AnalyzeResults()</span>
<span class="c1">## ...</span>
</pre></div>
</div>
<p>Users can then add, remove or replace steps as desired.</p>
</section>
<section id="developer-notes">
<h2>Developer Notes<a class="headerlink" href="#developer-notes" title="Permalink to this heading">#</a></h2>
<p>Steps to setup dependencies -</p>
<ul class="simple">
<li><p>initialize git submodules in <code class="docutils literal notranslate"><span class="pre">extern/libscran</span></code>.</p></li>
<li><p>run <code class="docutils literal notranslate"><span class="pre">cmake</span> <span class="pre">.</span></code> inside the <code class="docutils literal notranslate"><span class="pre">extern/knncolle</span></code> to download the annoy library. a future version of this will use a cmake to setup the extern directory.</p></li>
</ul>
<p>First one needs to build the extern library, this would generate a shared object file to <code class="docutils literal notranslate"><span class="pre">src/scranpy/core-[*].so</span></code></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>setup.py<span class="w"> </span>build_ext<span class="w"> </span>--inplace
</pre></div>
</div>
<p>For typical development workflows, run this for tests</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>setup.py<span class="w"> </span>build_ext<span class="w"> </span>--inplace<span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span>tox
</pre></div>
</div>
<p>To rebuild the <strong>ctypes</strong> bindings <a class="reference external" href="https://github.com/BiocPy/ctypes-wrapper"><strong>cpptypes</strong></a>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>cpptypes<span class="w"> </span>src/scranpy/lib<span class="w"> </span>--py<span class="w"> </span>src/scranpy/_cpphelpers.py<span class="w"> </span>--cpp<span class="w"> </span>src/scranpy/lib/bindings.cpp<span class="w"> </span>--dll<span class="w"> </span>_core
</pre></div>
</div>
<p>To rebuild the <a class="reference internal" href="#src/scranpy/analysis_dry.py"><span class="xref myst">dry run analysis source code</span></a>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>./scripts/dryrun.py<span class="w"> </span>src/scranpy/analyze/live_analyze.py<span class="w"> </span>&gt;<span class="w"> </span>src/scranpy/analyze/dry_analyze.py
</pre></div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="contributing.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Contributing</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Home</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Jayaram Kancherla
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">scran, in Python</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#quick-start">Quick start</a></li>
<li><a class="reference internal" href="#multiple-batches">Multiple batches</a></li>
<li><a class="reference internal" href="#multiple-modalities">Multiple modalities</a></li>
<li><a class="reference internal" href="#customizing-the-analysis">Customizing the analysis</a></li>
<li><a class="reference internal" href="#developer-notes">Developer Notes</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=680b02e7"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    </body>
</html>