<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="scranpy.batch_correction package" href="scranpy.batch_correction.html" /><link rel="prev" title="scranpy.aggregation package" href="scranpy.aggregation.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>scranpy.analyze package - scranpy unknown documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">scranpy unknown documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">scranpy unknown documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="modules.html">Module Reference</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current has-children"><a class="reference internal" href="scranpy.html">scranpy package</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of scranpy package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="scranpy.aggregation.html">scranpy.aggregation package</a></li>
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">scranpy.analyze package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.batch_correction.html">scranpy.batch_correction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.clustering.html">scranpy.clustering package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.dimensionality_reduction.html">scranpy.dimensionality_reduction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.feature_selection.html">scranpy.feature_selection package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.feature_set_enrichment.html">scranpy.feature_set_enrichment package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.marker_detection.html">scranpy.marker_detection package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.nearest_neighbors.html">scranpy.nearest_neighbors package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.normalization.html">scranpy.normalization package</a></li>
<li class="toctree-l3"><a class="reference internal" href="scranpy.quality_control.html">scranpy.quality_control package</a></li>
</ul>
</li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="scranpy-analyze-package">
<h1>scranpy.analyze package<a class="headerlink" href="#scranpy-analyze-package" title="Permalink to this heading">#</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading">#</a></h2>
</section>
<section id="module-scranpy.analyze.AnalyzeOptions">
<span id="scranpy-analyze-analyzeoptions-module"></span><h2>scranpy.analyze.AnalyzeOptions module<a class="headerlink" href="#module-scranpy.analyze.AnalyzeOptions" title="Permalink to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.analyze.AnalyzeOptions.</span></span><span class="sig-name descname"><span class="pre">AnalyzeOptions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">per_cell_rna_qc_metrics_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suggest_rna_qc_filters_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_rna_qc_filter_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">per_cell_adt_qc_metrics_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suggest_adt_qc_filters_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_adt_qc_filter_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">per_cell_crispr_qc_metrics_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suggest_crispr_qc_filters_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_crispr_qc_filter_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_cells_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_log_norm_counts_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grouped_size_factors_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_log_norm_counts_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_log_norm_counts_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_hvgs_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_gene_variances_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_run_pca_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_run_pca_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_run_pca_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combine_embeddings_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnn_correct_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_neighbor_index_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_tsne_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_umap_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">find_nearest_neighbors_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_snn_graph_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_score_markers_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_score_markers_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_score_markers_options=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">miscellaneous_options=&lt;factory&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/AnalyzeOptions.html#AnalyzeOptions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions" title="Permalink to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Optional parameters for all <a class="reference internal" href="#scranpy.analyze.analyze.analyze" title="scranpy.analyze.analyze.analyze"><code class="xref py py-meth docutils literal notranslate"><span class="pre">analyze()</span></code></a> steps.</p>
<p>Optional parameters for each function are named after the function with the <code class="docutils literal notranslate"><span class="pre">_options</span></code> suffix.
In most cases, these can be modified directly to refine the behavior of the
<a class="reference internal" href="#scranpy.analyze.analyze.analyze" title="scranpy.analyze.analyze.analyze"><code class="xref py py-meth docutils literal notranslate"><span class="pre">analyze()</span></code></a> function. However, for a few options, it usually makes more sense
to set them across multiple parameter objects simultaneously;
check out the setter methods of this class for more details.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_rna_qc_metrics_options">
<span class="sig-name descname"><span class="pre">per_cell_rna_qc_metrics_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_rna_qc_metrics_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.per_cell_rna_qc_metrics.per_cell_rna_qc_metrics" title="scranpy.quality_control.per_cell_rna_qc_metrics.per_cell_rna_qc_metrics"><code class="xref py py-meth docutils literal notranslate"><span class="pre">per_cell_rna_qc_metrics()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_rna_qc_filters_options">
<span class="sig-name descname"><span class="pre">suggest_rna_qc_filters_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_rna_qc_filters_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_rna_qc_filters.suggest_rna_qc_filters" title="scranpy.quality_control.suggest_rna_qc_filters.suggest_rna_qc_filters"><code class="xref py py-meth docutils literal notranslate"><span class="pre">suggest_rna_qc_filters()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_rna_qc_filter_options">
<span class="sig-name descname"><span class="pre">create_rna_qc_filter_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_rna_qc_filter_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_rna_qc_filter.create_rna_qc_filter" title="scranpy.quality_control.create_rna_qc_filter.create_rna_qc_filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_rna_qc_filter()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_adt_qc_metrics_options">
<span class="sig-name descname"><span class="pre">per_cell_adt_qc_metrics_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_adt_qc_metrics_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <code class="xref py py-meth docutils literal notranslate"><span class="pre">per_cell_adt_qc_metrics()</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_adt_qc_filters_options">
<span class="sig-name descname"><span class="pre">suggest_adt_qc_filters_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_adt_qc_filters_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_adt_qc_filters.suggest_adt_qc_filters" title="scranpy.quality_control.suggest_adt_qc_filters.suggest_adt_qc_filters"><code class="xref py py-meth docutils literal notranslate"><span class="pre">suggest_adt_qc_filters()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_adt_qc_filter_options">
<span class="sig-name descname"><span class="pre">create_adt_qc_filter_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_adt_qc_filter_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_adt_qc_filter.create_adt_qc_filter" title="scranpy.quality_control.create_adt_qc_filter.create_adt_qc_filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_adt_qc_filter()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_crispr_qc_metrics_options">
<span class="sig-name descname"><span class="pre">per_cell_crispr_qc_metrics_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_crispr_qc_metrics_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to
<a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.per_cell_crispr_qc_metrics.per_cell_crispr_qc_metrics" title="scranpy.quality_control.per_cell_crispr_qc_metrics.per_cell_crispr_qc_metrics"><code class="xref py py-meth docutils literal notranslate"><span class="pre">per_cell_crispr_qc_metrics()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_crispr_qc_filters_options">
<span class="sig-name descname"><span class="pre">suggest_crispr_qc_filters_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_crispr_qc_filters_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_crispr_qc_filters.suggest_crispr_qc_filters" title="scranpy.quality_control.suggest_crispr_qc_filters.suggest_crispr_qc_filters"><code class="xref py py-meth docutils literal notranslate"><span class="pre">suggest_crispr_qc_filters()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_crispr_qc_filter_options">
<span class="sig-name descname"><span class="pre">create_crispr_qc_filter_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_crispr_qc_filter_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_crispr_qc_filter.create_crispr_qc_filter" title="scranpy.quality_control.create_crispr_qc_filter.create_crispr_qc_filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_crispr_qc_filter()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.filter_cells_options">
<span class="sig-name descname"><span class="pre">filter_cells_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.filter_cells_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.filter_cells.filter_cells" title="scranpy.quality_control.filter_cells.filter_cells"><code class="xref py py-meth docutils literal notranslate"><span class="pre">filter_cells()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_log_norm_counts_options">
<span class="sig-name descname"><span class="pre">rna_log_norm_counts_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_log_norm_counts_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.normalization.html#scranpy.normalization.log_norm_counts.log_norm_counts" title="scranpy.normalization.log_norm_counts.log_norm_counts"><code class="xref py py-meth docutils literal notranslate"><span class="pre">log_norm_counts()</span></code></a>
for the RNA count matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.grouped_size_factors_options">
<span class="sig-name descname"><span class="pre">grouped_size_factors_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.grouped_size_factors_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.normalization.html#scranpy.normalization.grouped_size_factors.grouped_size_factors" title="scranpy.normalization.grouped_size_factors.grouped_size_factors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">grouped_size_factors()</span></code></a>
to compute ADT size factors.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_log_norm_counts_options">
<span class="sig-name descname"><span class="pre">adt_log_norm_counts_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_log_norm_counts_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.normalization.html#scranpy.normalization.log_norm_counts.log_norm_counts" title="scranpy.normalization.log_norm_counts.log_norm_counts"><code class="xref py py-meth docutils literal notranslate"><span class="pre">log_norm_counts()</span></code></a>
for the ADT count matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_log_norm_counts_options">
<span class="sig-name descname"><span class="pre">crispr_log_norm_counts_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_log_norm_counts_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.normalization.html#scranpy.normalization.log_norm_counts.log_norm_counts" title="scranpy.normalization.log_norm_counts.log_norm_counts"><code class="xref py py-meth docutils literal notranslate"><span class="pre">log_norm_counts()</span></code></a>
for the CRISPR count matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.choose_hvgs_options">
<span class="sig-name descname"><span class="pre">choose_hvgs_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.choose_hvgs_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.feature_selection.html#scranpy.feature_selection.choose_hvgs.choose_hvgs" title="scranpy.feature_selection.choose_hvgs.choose_hvgs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">choose_hvgs()</span></code></a>
to choose highly variable genes for the RNA data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.model_gene_variances_options">
<span class="sig-name descname"><span class="pre">model_gene_variances_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.model_gene_variances_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.feature_selection.html#scranpy.feature_selection.model_gene_variances.model_gene_variances" title="scranpy.feature_selection.model_gene_variances.model_gene_variances"><code class="xref py py-meth docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>
to model per-gene variances for the RNA data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_run_pca_options">
<span class="sig-name descname"><span class="pre">rna_run_pca_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_run_pca_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a>
for the RNA log-expression matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_run_pca_options">
<span class="sig-name descname"><span class="pre">adt_run_pca_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_run_pca_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a>
for the ADT log-expression matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_run_pca_options">
<span class="sig-name descname"><span class="pre">crispr_run_pca_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_run_pca_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a>
for the CRISPR log-expression matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.mnn_correct_options">
<span class="sig-name descname"><span class="pre">mnn_correct_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.mnn_correct_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.batch_correction.html#scranpy.batch_correction.mnn_correct.mnn_correct" title="scranpy.batch_correction.mnn_correct.mnn_correct"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mnn_correct()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.build_neighbor_index_options">
<span class="sig-name descname"><span class="pre">build_neighbor_index_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.build_neighbor_index_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.build_neighbor_index.build_neighbor_index" title="scranpy.nearest_neighbors.build_neighbor_index.build_neighbor_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_neighbor_index()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.find_nearest_neighbors_options">
<span class="sig-name descname"><span class="pre">find_nearest_neighbors_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.find_nearest_neighbors_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.find_nearest_neighbors.find_nearest_neighbors" title="scranpy.nearest_neighbors.find_nearest_neighbors.find_nearest_neighbors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">find_nearest_neighbors()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.run_tsne_options">
<span class="sig-name descname"><span class="pre">run_tsne_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.run_tsne_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_tsne.run_tsne" title="scranpy.dimensionality_reduction.run_tsne.run_tsne"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.run_umap_options">
<span class="sig-name descname"><span class="pre">run_umap_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.run_umap_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_umap.run_umap" title="scranpy.dimensionality_reduction.run_umap.run_umap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.build_snn_graph_options">
<span class="sig-name descname"><span class="pre">build_snn_graph_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.build_snn_graph_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.clustering.html#scranpy.clustering.build_snn_graph.build_snn_graph" title="scranpy.clustering.build_snn_graph.build_snn_graph"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_score_markers_options">
<span class="sig-name descname"><span class="pre">rna_score_markers_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_score_markers_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.score_markers" title="scranpy.marker_detection.score_markers.score_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">score_markers()</span></code></a>
for the RNA log-expression values.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_score_markers_options">
<span class="sig-name descname"><span class="pre">adt_score_markers_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_score_markers_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.score_markers" title="scranpy.marker_detection.score_markers.score_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">score_markers()</span></code></a>
for the ADT log-abundances.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_score_markers_options">
<span class="sig-name descname"><span class="pre">crispr_score_markers_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_score_markers_options" title="Permalink to this definition">#</a></dt>
<dd><p>Options to pass to <a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.score_markers" title="scranpy.marker_detection.score_markers.score_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">score_markers()</span></code></a>
for the CRISPR log-abundances.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.miscellaneous_options">
<span class="sig-name descname"><span class="pre">miscellaneous_options</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.miscellaneous_options" title="Permalink to this definition">#</a></dt>
<dd><p>Further options that are not associated with any single function call.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'adt_log_norm_counts_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,</span> <span class="pre">'adt_run_pca_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,</span> <span class="pre">'adt_score_markers_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,</span> <span class="pre">'build_neighbor_index_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.nearest_neighbors.build_neighbor_index.BuildNeighborIndexOptions'&gt;,</span> <span class="pre">'build_snn_graph_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.clustering.build_snn_graph.BuildSnnGraphOptions'&gt;,</span> <span class="pre">'choose_hvgs_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.feature_selection.choose_hvgs.ChooseHvgsOptions'&gt;,</span> <span class="pre">'combine_embeddings_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.combine_embeddings.CombineEmbeddingsOptions'&gt;,</span> <span class="pre">'create_adt_qc_filter_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.create_adt_qc_filter.CreateAdtQcFilterOptions'&gt;,</span> <span class="pre">'create_crispr_qc_filter_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.create_crispr_qc_filter.CreateCrisprQcFilterOptions'&gt;,</span> <span class="pre">'create_rna_qc_filter_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.create_rna_qc_filter.CreateRnaQcFilterOptions'&gt;,</span> <span class="pre">'crispr_log_norm_counts_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,</span> <span class="pre">'crispr_run_pca_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,</span> <span class="pre">'crispr_score_markers_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,</span> <span class="pre">'filter_cells_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.filter_cells.FilterCellsOptions'&gt;,</span> <span class="pre">'find_nearest_neighbors_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.nearest_neighbors.find_nearest_neighbors.FindNearestNeighborsOptions'&gt;,</span> <span class="pre">'grouped_size_factors_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.normalization.grouped_size_factors.GroupedSizeFactorsOptions'&gt;,</span> <span class="pre">'miscellaneous_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.analyze.AnalyzeOptions.MiscellaneousOptions'&gt;,</span> <span class="pre">'mnn_correct_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.batch_correction.mnn_correct.MnnCorrectOptions'&gt;,</span> <span class="pre">'model_gene_variances_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.feature_selection.model_gene_variances.ModelGeneVariancesOptions'&gt;,</span> <span class="pre">'per_cell_adt_qc_metrics_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_adt_qc_metrics.PerCellAdtQcMetricsOptions'&gt;,</span> <span class="pre">'per_cell_crispr_qc_metrics_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_crispr_qc_metrics.PerCellCrisprQcMetricsOptions'&gt;,</span> <span class="pre">'per_cell_rna_qc_metrics_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_rna_qc_metrics.PerCellRnaQcMetricsOptions'&gt;,</span> <span class="pre">'rna_log_norm_counts_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,</span> <span class="pre">'rna_run_pca_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,</span> <span class="pre">'rna_score_markers_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,</span> <span class="pre">'run_tsne_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_tsne.RunTsneOptions'&gt;,</span> <span class="pre">'run_umap_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_umap.RunUmapOptions'&gt;,</span> <span class="pre">'suggest_adt_qc_filters_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_adt_qc_filters.SuggestAdtQcFiltersOptions'&gt;,</span> <span class="pre">'suggest_crispr_qc_filters_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_crispr_qc_filters.SuggestCrisprQcFiltersOptions'&gt;,</span> <span class="pre">'suggest_rna_qc_filters_options':</span> <span class="pre">&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_rna_qc_filters.SuggestRnaQcFiltersOptions'&gt;}</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__annotations__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'adt_log_norm_counts_options':</span> <span class="pre">Field(name='adt_log_norm_counts_options',type=&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'adt_run_pca_options':</span> <span class="pre">Field(name='adt_run_pca_options',type=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'adt_score_markers_options':</span> <span class="pre">Field(name='adt_score_markers_options',type=&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'build_neighbor_index_options':</span> <span class="pre">Field(name='build_neighbor_index_options',type=&lt;class</span> <span class="pre">'scranpy.nearest_neighbors.build_neighbor_index.BuildNeighborIndexOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.nearest_neighbors.build_neighbor_index.BuildNeighborIndexOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'build_snn_graph_options':</span> <span class="pre">Field(name='build_snn_graph_options',type=&lt;class</span> <span class="pre">'scranpy.clustering.build_snn_graph.BuildSnnGraphOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.clustering.build_snn_graph.BuildSnnGraphOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'choose_hvgs_options':</span> <span class="pre">Field(name='choose_hvgs_options',type=&lt;class</span> <span class="pre">'scranpy.feature_selection.choose_hvgs.ChooseHvgsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.feature_selection.choose_hvgs.ChooseHvgsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'combine_embeddings_options':</span> <span class="pre">Field(name='combine_embeddings_options',type=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.combine_embeddings.CombineEmbeddingsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.combine_embeddings.CombineEmbeddingsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'create_adt_qc_filter_options':</span> <span class="pre">Field(name='create_adt_qc_filter_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.create_adt_qc_filter.CreateAdtQcFilterOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.create_adt_qc_filter.CreateAdtQcFilterOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'create_crispr_qc_filter_options':</span> <span class="pre">Field(name='create_crispr_qc_filter_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.create_crispr_qc_filter.CreateCrisprQcFilterOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.create_crispr_qc_filter.CreateCrisprQcFilterOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'create_rna_qc_filter_options':</span> <span class="pre">Field(name='create_rna_qc_filter_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.create_rna_qc_filter.CreateRnaQcFilterOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.create_rna_qc_filter.CreateRnaQcFilterOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_log_norm_counts_options':</span> <span class="pre">Field(name='crispr_log_norm_counts_options',type=&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_run_pca_options':</span> <span class="pre">Field(name='crispr_run_pca_options',type=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_score_markers_options':</span> <span class="pre">Field(name='crispr_score_markers_options',type=&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'filter_cells_options':</span> <span class="pre">Field(name='filter_cells_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.filter_cells.FilterCellsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.filter_cells.FilterCellsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'find_nearest_neighbors_options':</span> <span class="pre">Field(name='find_nearest_neighbors_options',type=&lt;class</span> <span class="pre">'scranpy.nearest_neighbors.find_nearest_neighbors.FindNearestNeighborsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.nearest_neighbors.find_nearest_neighbors.FindNearestNeighborsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'grouped_size_factors_options':</span> <span class="pre">Field(name='grouped_size_factors_options',type=&lt;class</span> <span class="pre">'scranpy.normalization.grouped_size_factors.GroupedSizeFactorsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.normalization.grouped_size_factors.GroupedSizeFactorsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'miscellaneous_options':</span> <span class="pre">Field(name='miscellaneous_options',type=&lt;class</span> <span class="pre">'scranpy.analyze.AnalyzeOptions.MiscellaneousOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.analyze.AnalyzeOptions.MiscellaneousOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'mnn_correct_options':</span> <span class="pre">Field(name='mnn_correct_options',type=&lt;class</span> <span class="pre">'scranpy.batch_correction.mnn_correct.MnnCorrectOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.batch_correction.mnn_correct.MnnCorrectOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'model_gene_variances_options':</span> <span class="pre">Field(name='model_gene_variances_options',type=&lt;class</span> <span class="pre">'scranpy.feature_selection.model_gene_variances.ModelGeneVariancesOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.feature_selection.model_gene_variances.ModelGeneVariancesOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'per_cell_adt_qc_metrics_options':</span> <span class="pre">Field(name='per_cell_adt_qc_metrics_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_adt_qc_metrics.PerCellAdtQcMetricsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_adt_qc_metrics.PerCellAdtQcMetricsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'per_cell_crispr_qc_metrics_options':</span> <span class="pre">Field(name='per_cell_crispr_qc_metrics_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_crispr_qc_metrics.PerCellCrisprQcMetricsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_crispr_qc_metrics.PerCellCrisprQcMetricsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'per_cell_rna_qc_metrics_options':</span> <span class="pre">Field(name='per_cell_rna_qc_metrics_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_rna_qc_metrics.PerCellRnaQcMetricsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.per_cell_rna_qc_metrics.PerCellRnaQcMetricsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_log_norm_counts_options':</span> <span class="pre">Field(name='rna_log_norm_counts_options',type=&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.normalization.log_norm_counts.LogNormCountsOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_run_pca_options':</span> <span class="pre">Field(name='rna_run_pca_options',type=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_pca.RunPcaOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_score_markers_options':</span> <span class="pre">Field(name='rna_score_markers_options',type=&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.marker_detection.score_markers.ScoreMarkersOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'run_tsne_options':</span> <span class="pre">Field(name='run_tsne_options',type=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_tsne.RunTsneOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_tsne.RunTsneOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'run_umap_options':</span> <span class="pre">Field(name='run_umap_options',type=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_umap.RunUmapOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.dimensionality_reduction.run_umap.RunUmapOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'suggest_adt_qc_filters_options':</span> <span class="pre">Field(name='suggest_adt_qc_filters_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_adt_qc_filters.SuggestAdtQcFiltersOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_adt_qc_filters.SuggestAdtQcFiltersOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'suggest_crispr_qc_filters_options':</span> <span class="pre">Field(name='suggest_crispr_qc_filters_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_crispr_qc_filters.SuggestCrisprQcFiltersOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_crispr_qc_filters.SuggestCrisprQcFiltersOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'suggest_rna_qc_filters_options':</span> <span class="pre">Field(name='suggest_rna_qc_filters_options',type=&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_rna_qc_filters.SuggestRnaQcFiltersOptions'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;class</span> <span class="pre">'scranpy.quality_control.suggest_rna_qc_filters.SuggestRnaQcFiltersOptions'&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__dataclass_fields__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__dataclass_params__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__eq__" title="Permalink to this definition">#</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__hash__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__repr__" title="Permalink to this definition">#</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id0">
<span class="sig-name descname"><span class="pre">adt_log_norm_counts_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.normalization.html#scranpy.normalization.log_norm_counts.LogNormCountsOptions" title="scranpy.normalization.log_norm_counts.LogNormCountsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">LogNormCountsOptions</span></code></a></em><a class="headerlink" href="#id0" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id1">
<span class="sig-name descname"><span class="pre">adt_run_pca_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.RunPcaOptions" title="scranpy.dimensionality_reduction.run_pca.RunPcaOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaOptions</span></code></a></em><a class="headerlink" href="#id1" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id2">
<span class="sig-name descname"><span class="pre">adt_score_markers_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.ScoreMarkersOptions" title="scranpy.marker_detection.score_markers.ScoreMarkersOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">ScoreMarkersOptions</span></code></a></em><a class="headerlink" href="#id2" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id3">
<span class="sig-name descname"><span class="pre">build_neighbor_index_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.build_neighbor_index.BuildNeighborIndexOptions" title="scranpy.nearest_neighbors.build_neighbor_index.BuildNeighborIndexOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">BuildNeighborIndexOptions</span></code></a></em><a class="headerlink" href="#id3" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id4">
<span class="sig-name descname"><span class="pre">build_snn_graph_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.clustering.html#scranpy.clustering.build_snn_graph.BuildSnnGraphOptions" title="scranpy.clustering.build_snn_graph.BuildSnnGraphOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">BuildSnnGraphOptions</span></code></a></em><a class="headerlink" href="#id4" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id5">
<span class="sig-name descname"><span class="pre">choose_hvgs_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.feature_selection.html#scranpy.feature_selection.choose_hvgs.ChooseHvgsOptions" title="scranpy.feature_selection.choose_hvgs.ChooseHvgsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">ChooseHvgsOptions</span></code></a></em><a class="headerlink" href="#id5" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.combine_embeddings_options">
<span class="sig-name descname"><span class="pre">combine_embeddings_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.combine_embeddings.CombineEmbeddingsOptions" title="scranpy.dimensionality_reduction.combine_embeddings.CombineEmbeddingsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">CombineEmbeddingsOptions</span></code></a></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.combine_embeddings_options" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id6">
<span class="sig-name descname"><span class="pre">create_adt_qc_filter_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_adt_qc_filter.CreateAdtQcFilterOptions" title="scranpy.quality_control.create_adt_qc_filter.CreateAdtQcFilterOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">CreateAdtQcFilterOptions</span></code></a></em><a class="headerlink" href="#id6" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id7">
<span class="sig-name descname"><span class="pre">create_crispr_qc_filter_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_crispr_qc_filter.CreateCrisprQcFilterOptions" title="scranpy.quality_control.create_crispr_qc_filter.CreateCrisprQcFilterOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">CreateCrisprQcFilterOptions</span></code></a></em><a class="headerlink" href="#id7" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id8">
<span class="sig-name descname"><span class="pre">create_rna_qc_filter_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_rna_qc_filter.CreateRnaQcFilterOptions" title="scranpy.quality_control.create_rna_qc_filter.CreateRnaQcFilterOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">CreateRnaQcFilterOptions</span></code></a></em><a class="headerlink" href="#id8" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id9">
<span class="sig-name descname"><span class="pre">crispr_log_norm_counts_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.normalization.html#scranpy.normalization.log_norm_counts.LogNormCountsOptions" title="scranpy.normalization.log_norm_counts.LogNormCountsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">LogNormCountsOptions</span></code></a></em><a class="headerlink" href="#id9" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id10">
<span class="sig-name descname"><span class="pre">crispr_run_pca_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.RunPcaOptions" title="scranpy.dimensionality_reduction.run_pca.RunPcaOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaOptions</span></code></a></em><a class="headerlink" href="#id10" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id11">
<span class="sig-name descname"><span class="pre">crispr_score_markers_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.ScoreMarkersOptions" title="scranpy.marker_detection.score_markers.ScoreMarkersOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">ScoreMarkersOptions</span></code></a></em><a class="headerlink" href="#id11" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id12">
<span class="sig-name descname"><span class="pre">filter_cells_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.filter_cells.FilterCellsOptions" title="scranpy.quality_control.filter_cells.FilterCellsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">FilterCellsOptions</span></code></a></em><a class="headerlink" href="#id12" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id13">
<span class="sig-name descname"><span class="pre">find_nearest_neighbors_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.find_nearest_neighbors.FindNearestNeighborsOptions" title="scranpy.nearest_neighbors.find_nearest_neighbors.FindNearestNeighborsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindNearestNeighborsOptions</span></code></a></em><a class="headerlink" href="#id13" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id14">
<span class="sig-name descname"><span class="pre">grouped_size_factors_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.normalization.html#scranpy.normalization.grouped_size_factors.GroupedSizeFactorsOptions" title="scranpy.normalization.grouped_size_factors.GroupedSizeFactorsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">GroupedSizeFactorsOptions</span></code></a></em><a class="headerlink" href="#id14" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id15">
<span class="sig-name descname"><span class="pre">miscellaneous_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions" title="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">MiscellaneousOptions</span></code></a></em><a class="headerlink" href="#id15" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id16">
<span class="sig-name descname"><span class="pre">mnn_correct_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.batch_correction.html#scranpy.batch_correction.mnn_correct.MnnCorrectOptions" title="scranpy.batch_correction.mnn_correct.MnnCorrectOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">MnnCorrectOptions</span></code></a></em><a class="headerlink" href="#id16" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id17">
<span class="sig-name descname"><span class="pre">model_gene_variances_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.feature_selection.html#scranpy.feature_selection.model_gene_variances.ModelGeneVariancesOptions" title="scranpy.feature_selection.model_gene_variances.ModelGeneVariancesOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">ModelGeneVariancesOptions</span></code></a></em><a class="headerlink" href="#id17" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id18">
<span class="sig-name descname"><span class="pre">per_cell_adt_qc_metrics_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.per_cell_adt_qc_metrics.PerCellAdtQcMetricsOptions" title="scranpy.quality_control.per_cell_adt_qc_metrics.PerCellAdtQcMetricsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">PerCellAdtQcMetricsOptions</span></code></a></em><a class="headerlink" href="#id18" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id19">
<span class="sig-name descname"><span class="pre">per_cell_crispr_qc_metrics_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.per_cell_crispr_qc_metrics.PerCellCrisprQcMetricsOptions" title="scranpy.quality_control.per_cell_crispr_qc_metrics.PerCellCrisprQcMetricsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">PerCellCrisprQcMetricsOptions</span></code></a></em><a class="headerlink" href="#id19" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id20">
<span class="sig-name descname"><span class="pre">per_cell_rna_qc_metrics_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.per_cell_rna_qc_metrics.PerCellRnaQcMetricsOptions" title="scranpy.quality_control.per_cell_rna_qc_metrics.PerCellRnaQcMetricsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">PerCellRnaQcMetricsOptions</span></code></a></em><a class="headerlink" href="#id20" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id21">
<span class="sig-name descname"><span class="pre">rna_log_norm_counts_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.normalization.html#scranpy.normalization.log_norm_counts.LogNormCountsOptions" title="scranpy.normalization.log_norm_counts.LogNormCountsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">LogNormCountsOptions</span></code></a></em><a class="headerlink" href="#id21" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id22">
<span class="sig-name descname"><span class="pre">rna_run_pca_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.RunPcaOptions" title="scranpy.dimensionality_reduction.run_pca.RunPcaOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaOptions</span></code></a></em><a class="headerlink" href="#id22" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id23">
<span class="sig-name descname"><span class="pre">rna_score_markers_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.ScoreMarkersOptions" title="scranpy.marker_detection.score_markers.ScoreMarkersOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">ScoreMarkersOptions</span></code></a></em><a class="headerlink" href="#id23" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id24">
<span class="sig-name descname"><span class="pre">run_tsne_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_tsne.RunTsneOptions" title="scranpy.dimensionality_reduction.run_tsne.RunTsneOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunTsneOptions</span></code></a></em><a class="headerlink" href="#id24" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id25">
<span class="sig-name descname"><span class="pre">run_umap_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_umap.RunUmapOptions" title="scranpy.dimensionality_reduction.run_umap.RunUmapOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunUmapOptions</span></code></a></em><a class="headerlink" href="#id25" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.set_seed">
<span class="sig-name descname"><span class="pre">set_seed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/AnalyzeOptions.html#AnalyzeOptions.set_seed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.set_seed" title="Permalink to this definition">#</a></dt>
<dd><p>Set seed for RNG. This calls the method of the same name for
<a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a>,
<a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_tsne.run_tsne" title="scranpy.dimensionality_reduction.run_tsne.run_tsne"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>,
<a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_umap.run_umap" title="scranpy.dimensionality_reduction.run_umap.run_umap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>)  Seed for random number generation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.AnalyzeOptions.set_threads">
<span class="sig-name descname"><span class="pre">set_threads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/AnalyzeOptions.html#AnalyzeOptions.set_threads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.set_threads" title="Permalink to this definition">#</a></dt>
<dd><p>Set number of threads to use. This calls the method of the same name for
<code class="xref py py-meth docutils literal notranslate"><span class="pre">per_cell_rna_qc_metrics()</span></code>,
<code class="xref py py-meth docutils literal notranslate"><span class="pre">()</span></code>,
<a class="reference internal" href="scranpy.feature_selection.html#module-scranpy.feature_selection.model_gene_variances" title="scranpy.feature_selection.model_gene_variances"><code class="xref py py-meth docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>,
<a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a>,
<a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.find_nearest_neighbors.find_nearest_neighbors" title="scranpy.nearest_neighbors.find_nearest_neighbors.find_nearest_neighbors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">find_nearest_neighbors()</span></code></a>.</p>
<p>The number of threads provided to
<a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.find_nearest_neighbors.find_nearest_neighbors" title="scranpy.nearest_neighbors.find_nearest_neighbors.find_nearest_neighbors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">find_nearest_neighbors()</span></code></a>
is also used in <a class="reference internal" href="#module-scranpy.analyze.run_neighbor_suite" title="scranpy.analyze.run_neighbor_suite"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_neighbor_suite()</span></code></a>,
which determines the thread allocation to subsequent steps like
<a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_tsne.run_tsne" title="scranpy.dimensionality_reduction.run_tsne.run_tsne"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>
and
<a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.score_markers" title="scranpy.marker_detection.score_markers.score_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.
In all cases, thread utilization will not exceed the limit specified here in <code class="docutils literal notranslate"><span class="pre">num_threads</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>)  Number of threads. Defaults to 1.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id26">
<span class="sig-name descname"><span class="pre">suggest_adt_qc_filters_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_adt_qc_filters.SuggestAdtQcFiltersOptions" title="scranpy.quality_control.suggest_adt_qc_filters.SuggestAdtQcFiltersOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestAdtQcFiltersOptions</span></code></a></em><a class="headerlink" href="#id26" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id27">
<span class="sig-name descname"><span class="pre">suggest_crispr_qc_filters_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_crispr_qc_filters.SuggestCrisprQcFiltersOptions" title="scranpy.quality_control.suggest_crispr_qc_filters.SuggestCrisprQcFiltersOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestCrisprQcFiltersOptions</span></code></a></em><a class="headerlink" href="#id27" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id28">
<span class="sig-name descname"><span class="pre">suggest_rna_qc_filters_options</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_rna_qc_filters.SuggestRnaQcFiltersOptions" title="scranpy.quality_control.suggest_rna_qc_filters.SuggestRnaQcFiltersOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestRnaQcFiltersOptions</span></code></a></em><a class="headerlink" href="#id28" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.analyze.AnalyzeOptions.</span></span><span class="sig-name descname"><span class="pre">MiscellaneousOptions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_feature_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_feature_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_feature_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_on_rna_qc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_on_adt_qc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_on_crispr_qc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snn_graph_multilevel_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/AnalyzeOptions.html#MiscellaneousOptions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions" title="Permalink to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Miscellaneous options for <a class="reference internal" href="#scranpy.analyze.analyze.analyze" title="scranpy.analyze.analyze.analyze"><code class="xref py py-meth docutils literal notranslate"><span class="pre">analyze()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.cell_names">
<span class="sig-name descname"><span class="pre">cell_names</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.cell_names" title="Permalink to this definition">#</a></dt>
<dd><p>Names for all cells in the dataset, to be added to any per-cell data frames.
This should have the same length as the number of columns in each data matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.rna_feature_names">
<span class="sig-name descname"><span class="pre">rna_feature_names</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.rna_feature_names" title="Permalink to this definition">#</a></dt>
<dd><p>Names for all features in the RNA data.
This should have the same length as the number of rows in the RNA count matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.adt_feature_names">
<span class="sig-name descname"><span class="pre">adt_feature_names</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.adt_feature_names" title="Permalink to this definition">#</a></dt>
<dd><p>Names for all tags in the ADT data.
This should have the same length as the number of rows in the ADT count matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.crispr_feature_names">
<span class="sig-name descname"><span class="pre">crispr_feature_names</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.crispr_feature_names" title="Permalink to this definition">#</a></dt>
<dd><p>Names for all guides in the CRISPR data.
This should have the same length as the number of rows in the CRISPR count matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_rna_qc">
<span class="sig-name descname"><span class="pre">filter_on_rna_qc</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_rna_qc" title="Permalink to this definition">#</a></dt>
<dd><p>Whether to filter cells on the RNA-based quality control metrics,
when RNA data is available.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_adt_qc">
<span class="sig-name descname"><span class="pre">filter_on_adt_qc</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_adt_qc" title="Permalink to this definition">#</a></dt>
<dd><p>Whether to filter cells on the ADT-based quality control metrics,
when ADT data is available.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_crispr_qc">
<span class="sig-name descname"><span class="pre">filter_on_crispr_qc</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_crispr_qc" title="Permalink to this definition">#</a></dt>
<dd><p>Whether to filter cells on the CRISPR-based quality control metrics,
when CRISPR data is available.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.snn_graph_multilevel_resolution">
<span class="sig-name descname"><span class="pre">snn_graph_multilevel_resolution</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.snn_graph_multilevel_resolution" title="Permalink to this definition">#</a></dt>
<dd><p>Resolution to use for multi-level clustering of the SNN graph.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.block">
<span class="sig-name descname"><span class="pre">block</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.block" title="Permalink to this definition">#</a></dt>
<dd><p>Block assignment for each cell.
This should have length equal to the total number of cells in the dataset, before any quality control
is applied.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'adt_feature_names':</span> <span class="pre">typing.Optional[typing.Sequence[str]],</span> <span class="pre">'block':</span> <span class="pre">typing.Optional[typing.Sequence],</span> <span class="pre">'cell_names':</span> <span class="pre">typing.Optional[typing.Sequence[str]],</span> <span class="pre">'crispr_feature_names':</span> <span class="pre">typing.Optional[typing.Sequence[str]],</span> <span class="pre">'filter_on_adt_qc':</span> <span class="pre">&lt;class</span> <span class="pre">'bool'&gt;,</span> <span class="pre">'filter_on_crispr_qc':</span> <span class="pre">&lt;class</span> <span class="pre">'bool'&gt;,</span> <span class="pre">'filter_on_rna_qc':</span> <span class="pre">&lt;class</span> <span class="pre">'bool'&gt;,</span> <span class="pre">'rna_feature_names':</span> <span class="pre">typing.Optional[typing.Sequence[str]],</span> <span class="pre">'snn_graph_multilevel_resolution':</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;}</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__annotations__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'adt_feature_names':</span> <span class="pre">Field(name='adt_feature_names',type=typing.Optional[typing.Sequence[str]],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'block':</span> <span class="pre">Field(name='block',type=typing.Optional[typing.Sequence],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'cell_names':</span> <span class="pre">Field(name='cell_names',type=typing.Optional[typing.Sequence[str]],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_feature_names':</span> <span class="pre">Field(name='crispr_feature_names',type=typing.Optional[typing.Sequence[str]],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'filter_on_adt_qc':</span> <span class="pre">Field(name='filter_on_adt_qc',type=&lt;class</span> <span class="pre">'bool'&gt;,default=True,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'filter_on_crispr_qc':</span> <span class="pre">Field(name='filter_on_crispr_qc',type=&lt;class</span> <span class="pre">'bool'&gt;,default=True,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'filter_on_rna_qc':</span> <span class="pre">Field(name='filter_on_rna_qc',type=&lt;class</span> <span class="pre">'bool'&gt;,default=True,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_feature_names':</span> <span class="pre">Field(name='rna_feature_names',type=typing.Optional[typing.Sequence[str]],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'snn_graph_multilevel_resolution':</span> <span class="pre">Field(name='snn_graph_multilevel_resolution',type=&lt;class</span> <span class="pre">'int'&gt;,default=1,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__dataclass_fields__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__dataclass_params__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__eq__" title="Permalink to this definition">#</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__hash__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__repr__" title="Permalink to this definition">#</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id29">
<span class="sig-name descname"><span class="pre">adt_feature_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a><span class="pre">]]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id29" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id30">
<span class="sig-name descname"><span class="pre">block</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id30" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id31">
<span class="sig-name descname"><span class="pre">cell_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a><span class="pre">]]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id31" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id32">
<span class="sig-name descname"><span class="pre">crispr_feature_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a><span class="pre">]]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id32" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id33">
<span class="sig-name descname"><span class="pre">filter_on_adt_qc</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">True</span></em><a class="headerlink" href="#id33" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id34">
<span class="sig-name descname"><span class="pre">filter_on_crispr_qc</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">True</span></em><a class="headerlink" href="#id34" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id35">
<span class="sig-name descname"><span class="pre">filter_on_rna_qc</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">True</span></em><a class="headerlink" href="#id35" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id36">
<span class="sig-name descname"><span class="pre">rna_feature_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a><span class="pre">]]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id36" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id37">
<span class="sig-name descname"><span class="pre">snn_graph_multilevel_resolution</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span></em><a class="headerlink" href="#id37" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-scranpy.analyze.AnalyzeResults">
<span id="scranpy-analyze-analyzeresults-module"></span><h2>scranpy.analyze.AnalyzeResults module<a class="headerlink" href="#module-scranpy.analyze.AnalyzeResults" title="Permalink to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.analyze.AnalyzeResults.</span></span><span class="sig-name descname"><span class="pre">AnalyzeResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_quality_control_metrics</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_quality_control_thresholds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_quality_control_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_quality_control_metrics</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_quality_control_thresholds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_quality_control_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_quality_control_metrics</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_quality_control_thresholds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_quality_control_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_control_retained</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_size_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_size_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_size_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_variances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvgs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_pca</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_pca</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_pca</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combined_pcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tsne</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snn_graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_markers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_markers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_markers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/AnalyzeResults.html#AnalyzeResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults" title="Permalink to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results across all analyis steps from <a class="reference internal" href="#scranpy.analyze.analyze.analyze" title="scranpy.analyze.analyze.analyze"><code class="xref py py-meth docutils literal notranslate"><span class="pre">analyze()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_metrics">
<span class="sig-name descname"><span class="pre">rna_quality_control_metrics</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_metrics" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.per_cell_rna_qc_metrics.per_cell_rna_qc_metrics" title="scranpy.quality_control.per_cell_rna_qc_metrics.per_cell_rna_qc_metrics"><code class="xref py py-meth docutils literal notranslate"><span class="pre">per_cell_rna_qc_metrics()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_thresholds">
<span class="sig-name descname"><span class="pre">rna_quality_control_thresholds</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_thresholds" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_rna_qc_filters.suggest_rna_qc_filters" title="scranpy.quality_control.suggest_rna_qc_filters.suggest_rna_qc_filters"><code class="xref py py-meth docutils literal notranslate"><span class="pre">suggest_rna_qc_filters()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_filter">
<span class="sig-name descname"><span class="pre">rna_quality_control_filter</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_filter" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_rna_qc_filter.create_rna_qc_filter" title="scranpy.quality_control.create_rna_qc_filter.create_rna_qc_filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_rna_qc_filter()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_metrics">
<span class="sig-name descname"><span class="pre">adt_quality_control_metrics</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_metrics" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.per_cell_adt_qc_metrics.per_cell_adt_qc_metrics" title="scranpy.quality_control.per_cell_adt_qc_metrics.per_cell_adt_qc_metrics"><code class="xref py py-meth docutils literal notranslate"><span class="pre">per_cell_adt_qc_metrics()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_thresholds">
<span class="sig-name descname"><span class="pre">adt_quality_control_thresholds</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_thresholds" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_adt_qc_filters.suggest_adt_qc_filters" title="scranpy.quality_control.suggest_adt_qc_filters.suggest_adt_qc_filters"><code class="xref py py-meth docutils literal notranslate"><span class="pre">suggest_adt_qc_filters()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_filter">
<span class="sig-name descname"><span class="pre">adt_quality_control_filter</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_filter" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_adt_qc_filter.create_adt_qc_filter" title="scranpy.quality_control.create_adt_qc_filter.create_adt_qc_filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_adt_qc_filter()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_metrics">
<span class="sig-name descname"><span class="pre">crispr_quality_control_metrics</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_metrics" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.per_cell_crispr_qc_metrics.per_cell_crispr_qc_metrics" title="scranpy.quality_control.per_cell_crispr_qc_metrics.per_cell_crispr_qc_metrics"><code class="xref py py-meth docutils literal notranslate"><span class="pre">per_cell_crispr_qc_metrics()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_thresholds">
<span class="sig-name descname"><span class="pre">crispr_quality_control_thresholds</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_thresholds" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.suggest_crispr_qc_filters.suggest_crispr_qc_filters" title="scranpy.quality_control.suggest_crispr_qc_filters.suggest_crispr_qc_filters"><code class="xref py py-meth docutils literal notranslate"><span class="pre">suggest_crispr_qc_filters()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_filter">
<span class="sig-name descname"><span class="pre">crispr_quality_control_filter</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_filter" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.quality_control.html#scranpy.quality_control.create_crispr_qc_filter.create_crispr_qc_filter" title="scranpy.quality_control.create_crispr_qc_filter.create_crispr_qc_filter"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_crispr_qc_filter()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.quality_control_retained">
<span class="sig-name descname"><span class="pre">quality_control_retained</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.quality_control_retained" title="Permalink to this definition">#</a></dt>
<dd><p>Array of length equal to the number of cells in the dataset before quality filtering,
indicating whether each cell should be retained.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_size_factors">
<span class="sig-name descname"><span class="pre">rna_size_factors</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_size_factors" title="Permalink to this definition">#</a></dt>
<dd><p>Array of length equal to the number of cells in the dataset after quality filtering,
containing the size factor from the RNA data for each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_size_factors">
<span class="sig-name descname"><span class="pre">adt_size_factors</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_size_factors" title="Permalink to this definition">#</a></dt>
<dd><p>Array of length equal to the number of cells in the dataset after quality filtering,
containing the size factor from the ADT data for each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_size_factors">
<span class="sig-name descname"><span class="pre">crispr_size_factors</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_size_factors" title="Permalink to this definition">#</a></dt>
<dd><p>Array of length equal to the number of cells in the dataset after quality filtering,
containing the size factor from the CRISPR data for each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.gene_variances">
<span class="sig-name descname"><span class="pre">gene_variances</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.gene_variances" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.feature_selection.html#scranpy.feature_selection.model_gene_variances.model_gene_variances" title="scranpy.feature_selection.model_gene_variances.model_gene_variances"><code class="xref py py-meth docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.hvgs">
<span class="sig-name descname"><span class="pre">hvgs</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.hvgs" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.feature_selection.html#scranpy.feature_selection.choose_hvgs.choose_hvgs" title="scranpy.feature_selection.choose_hvgs.choose_hvgs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">choose_hvgs()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_pca">
<span class="sig-name descname"><span class="pre">rna_pca</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_pca" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a> on the RNA data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_pca">
<span class="sig-name descname"><span class="pre">adt_pca</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_pca" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a> on the ADT data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_pca">
<span class="sig-name descname"><span class="pre">crispr_pca</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_pca" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a> on the CRISPR data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.combined_pcs">
<span class="sig-name descname"><span class="pre">combined_pcs</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.combined_pcs" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.combine_embeddings.combine_embeddings" title="scranpy.dimensionality_reduction.combine_embeddings.combine_embeddings"><code class="xref py py-meth docutils literal notranslate"><span class="pre">combine_embeddings()</span></code></a>
on the principal components for multiple modalities.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.mnn">
<span class="sig-name descname"><span class="pre">mnn</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.mnn" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.batch_correction.html#scranpy.batch_correction.mnn_correct.mnn_correct" title="scranpy.batch_correction.mnn_correct.mnn_correct"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mnn_correct()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.tsne">
<span class="sig-name descname"><span class="pre">tsne</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.tsne" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_tsne.run_tsne" title="scranpy.dimensionality_reduction.run_tsne.run_tsne"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.umap">
<span class="sig-name descname"><span class="pre">umap</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.umap" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_umap.run_umap" title="scranpy.dimensionality_reduction.run_umap.run_umap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.snn_graph">
<span class="sig-name descname"><span class="pre">snn_graph</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.snn_graph" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.clustering.html#scranpy.clustering.build_snn_graph.build_snn_graph" title="scranpy.clustering.build_snn_graph.build_snn_graph"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.clusters">
<span class="sig-name descname"><span class="pre">clusters</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.clusters" title="Permalink to this definition">#</a></dt>
<dd><p>List of length equal to the number of cells in the (filtered) dataset,
containing the cluster assignment for each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_markers">
<span class="sig-name descname"><span class="pre">rna_markers</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_markers" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.score_markers" title="scranpy.marker_detection.score_markers.score_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">score_markers()</span></code></a>
on the RNA data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_markers">
<span class="sig-name descname"><span class="pre">adt_markers</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_markers" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.score_markers" title="scranpy.marker_detection.score_markers.score_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">score_markers()</span></code></a>
on the ADT data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_markers">
<span class="sig-name descname"><span class="pre">crispr_markers</span></span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_markers" title="Permalink to this definition">#</a></dt>
<dd><p>Output of <a class="reference internal" href="scranpy.marker_detection.html#scranpy.marker_detection.score_markers.score_markers" title="scranpy.marker_detection.score_markers.score_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">score_markers()</span></code></a>
on the CRISPR data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'adt_markers':</span> <span class="pre">typing.Optional[typing.Mapping],</span> <span class="pre">'adt_pca':</span> <span class="pre">typing.Optional[scranpy.dimensionality_reduction.run_pca.PcaResult],</span> <span class="pre">'adt_quality_control_filter':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'adt_quality_control_metrics':</span> <span class="pre">typing.Optional[biocframe.BiocFrame.BiocFrame],</span> <span class="pre">'adt_quality_control_thresholds':</span> <span class="pre">typing.Optional[biocframe.BiocFrame.BiocFrame],</span> <span class="pre">'adt_size_factors':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'clusters':</span> <span class="pre">typing.Optional[list],</span> <span class="pre">'combined_pcs':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'crispr_markers':</span> <span class="pre">typing.Optional[typing.Mapping],</span> <span class="pre">'crispr_pca':</span> <span class="pre">typing.Optional[scranpy.dimensionality_reduction.run_pca.PcaResult],</span> <span class="pre">'crispr_quality_control_filter':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'crispr_quality_control_metrics':</span> <span class="pre">typing.Optional[biocframe.BiocFrame.BiocFrame],</span> <span class="pre">'crispr_quality_control_thresholds':</span> <span class="pre">typing.Optional[biocframe.BiocFrame.BiocFrame],</span> <span class="pre">'crispr_size_factors':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'gene_variances':</span> <span class="pre">typing.Optional[biocframe.BiocFrame.BiocFrame],</span> <span class="pre">'hvgs':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'mnn':</span> <span class="pre">typing.Optional[scranpy.batch_correction.mnn_correct.MnnCorrectResult],</span> <span class="pre">'quality_control_retained':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'rna_markers':</span> <span class="pre">typing.Optional[typing.Mapping],</span> <span class="pre">'rna_pca':</span> <span class="pre">typing.Optional[scranpy.dimensionality_reduction.run_pca.PcaResult],</span> <span class="pre">'rna_quality_control_filter':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'rna_quality_control_metrics':</span> <span class="pre">typing.Optional[biocframe.BiocFrame.BiocFrame],</span> <span class="pre">'rna_quality_control_thresholds':</span> <span class="pre">typing.Optional[biocframe.BiocFrame.BiocFrame],</span> <span class="pre">'rna_size_factors':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'snn_graph':</span> <span class="pre">typing.Optional[igraph.Graph],</span> <span class="pre">'tsne':</span> <span class="pre">typing.Optional[scranpy.dimensionality_reduction.run_tsne.TsneEmbedding],</span> <span class="pre">'umap':</span> <span class="pre">typing.Optional[scranpy.dimensionality_reduction.run_umap.UmapEmbedding]}</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__annotations__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'adt_markers':</span> <span class="pre">Field(name='adt_markers',type=typing.Optional[typing.Mapping],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'adt_pca':</span> <span class="pre">Field(name='adt_pca',type=typing.Optional[scranpy.dimensionality_reduction.run_pca.PcaResult],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'adt_quality_control_filter':</span> <span class="pre">Field(name='adt_quality_control_filter',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'adt_quality_control_metrics':</span> <span class="pre">Field(name='adt_quality_control_metrics',type=typing.Optional[biocframe.BiocFrame.BiocFrame],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'adt_quality_control_thresholds':</span> <span class="pre">Field(name='adt_quality_control_thresholds',type=typing.Optional[biocframe.BiocFrame.BiocFrame],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'adt_size_factors':</span> <span class="pre">Field(name='adt_size_factors',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'clusters':</span> <span class="pre">Field(name='clusters',type=typing.Optional[list],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'combined_pcs':</span> <span class="pre">Field(name='combined_pcs',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_markers':</span> <span class="pre">Field(name='crispr_markers',type=typing.Optional[typing.Mapping],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_pca':</span> <span class="pre">Field(name='crispr_pca',type=typing.Optional[scranpy.dimensionality_reduction.run_pca.PcaResult],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_quality_control_filter':</span> <span class="pre">Field(name='crispr_quality_control_filter',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_quality_control_metrics':</span> <span class="pre">Field(name='crispr_quality_control_metrics',type=typing.Optional[biocframe.BiocFrame.BiocFrame],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_quality_control_thresholds':</span> <span class="pre">Field(name='crispr_quality_control_thresholds',type=typing.Optional[biocframe.BiocFrame.BiocFrame],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'crispr_size_factors':</span> <span class="pre">Field(name='crispr_size_factors',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'gene_variances':</span> <span class="pre">Field(name='gene_variances',type=typing.Optional[biocframe.BiocFrame.BiocFrame],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'hvgs':</span> <span class="pre">Field(name='hvgs',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'mnn':</span> <span class="pre">Field(name='mnn',type=typing.Optional[scranpy.batch_correction.mnn_correct.MnnCorrectResult],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'quality_control_retained':</span> <span class="pre">Field(name='quality_control_retained',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_markers':</span> <span class="pre">Field(name='rna_markers',type=typing.Optional[typing.Mapping],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_pca':</span> <span class="pre">Field(name='rna_pca',type=typing.Optional[scranpy.dimensionality_reduction.run_pca.PcaResult],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_quality_control_filter':</span> <span class="pre">Field(name='rna_quality_control_filter',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_quality_control_metrics':</span> <span class="pre">Field(name='rna_quality_control_metrics',type=typing.Optional[biocframe.BiocFrame.BiocFrame],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_quality_control_thresholds':</span> <span class="pre">Field(name='rna_quality_control_thresholds',type=typing.Optional[biocframe.BiocFrame.BiocFrame],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'rna_size_factors':</span> <span class="pre">Field(name='rna_size_factors',type=typing.Optional[numpy.ndarray],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'snn_graph':</span> <span class="pre">Field(name='snn_graph',type=typing.Optional[igraph.Graph],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'tsne':</span> <span class="pre">Field(name='tsne',type=typing.Optional[scranpy.dimensionality_reduction.run_tsne.TsneEmbedding],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD),</span> <span class="pre">'umap':</span> <span class="pre">Field(name='umap',type=typing.Optional[scranpy.dimensionality_reduction.run_umap.UmapEmbedding],default=None,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__dataclass_fields__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__dataclass_params__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__eq__" title="Permalink to this definition">#</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__hash__" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__repr__" title="Permalink to this definition">#</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id38">
<span class="sig-name descname"><span class="pre">adt_markers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Mapping" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id38" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id39">
<span class="sig-name descname"><span class="pre">adt_pca</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.PcaResult" title="scranpy.dimensionality_reduction.run_pca.PcaResult"><code class="xref py py-class docutils literal notranslate"><span class="pre">PcaResult</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id39" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id40">
<span class="sig-name descname"><span class="pre">adt_quality_control_filter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id40" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id41">
<span class="sig-name descname"><span class="pre">adt_quality_control_metrics</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.5.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id41" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id42">
<span class="sig-name descname"><span class="pre">adt_quality_control_thresholds</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.5.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id42" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id43">
<span class="sig-name descname"><span class="pre">adt_size_factors</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id43" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id44">
<span class="sig-name descname"><span class="pre">clusters</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id44" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id45">
<span class="sig-name descname"><span class="pre">combined_pcs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id45" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id46">
<span class="sig-name descname"><span class="pre">crispr_markers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Mapping" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id46" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id47">
<span class="sig-name descname"><span class="pre">crispr_pca</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.PcaResult" title="scranpy.dimensionality_reduction.run_pca.PcaResult"><code class="xref py py-class docutils literal notranslate"><span class="pre">PcaResult</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id47" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id48">
<span class="sig-name descname"><span class="pre">crispr_quality_control_filter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id48" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id49">
<span class="sig-name descname"><span class="pre">crispr_quality_control_metrics</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.5.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id49" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id50">
<span class="sig-name descname"><span class="pre">crispr_quality_control_thresholds</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.5.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id50" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id51">
<span class="sig-name descname"><span class="pre">crispr_size_factors</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id51" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id52">
<span class="sig-name descname"><span class="pre">gene_variances</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.5.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id52" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id53">
<span class="sig-name descname"><span class="pre">hvgs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id53" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id54">
<span class="sig-name descname"><span class="pre">mnn</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="scranpy.batch_correction.html#scranpy.batch_correction.mnn_correct.MnnCorrectResult" title="scranpy.batch_correction.mnn_correct.MnnCorrectResult"><code class="xref py py-class docutils literal notranslate"><span class="pre">MnnCorrectResult</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id54" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id55">
<span class="sig-name descname"><span class="pre">quality_control_retained</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id55" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id56">
<span class="sig-name descname"><span class="pre">rna_markers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Mapping" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id56" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id57">
<span class="sig-name descname"><span class="pre">rna_pca</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.PcaResult" title="scranpy.dimensionality_reduction.run_pca.PcaResult"><code class="xref py py-class docutils literal notranslate"><span class="pre">PcaResult</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id57" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id58">
<span class="sig-name descname"><span class="pre">rna_quality_control_filter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id58" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id59">
<span class="sig-name descname"><span class="pre">rna_quality_control_metrics</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.5.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id59" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id60">
<span class="sig-name descname"><span class="pre">rna_quality_control_thresholds</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.5.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id60" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id61">
<span class="sig-name descname"><span class="pre">rna_size_factors</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id61" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id62">
<span class="sig-name descname"><span class="pre">snn_graph</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">Graph</span></code><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id62" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.AnalyzeResults.to_sce">
<span class="sig-name descname"><span class="pre">to_sce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/AnalyzeResults.html#AnalyzeResults.to_sce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.to_sce" title="Permalink to this definition">#</a></dt>
<dd><p>Save results as a <code class="xref py py-class docutils literal notranslate"><span class="pre">singlecellexperiment.SingleCellExperiment</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong>  Input object. usually a matrix of raw counts.</p></li>
<li><p><strong>assay</strong>  Assay name for the matrix.
Defaults to counts.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.4.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An SCE with the results.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id63">
<span class="sig-name descname"><span class="pre">tsne</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_tsne.TsneEmbedding" title="scranpy.dimensionality_reduction.run_tsne.TsneEmbedding"><code class="xref py py-class docutils literal notranslate"><span class="pre">TsneEmbedding</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id63" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id64">
<span class="sig-name descname"><span class="pre">umap</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_umap.UmapEmbedding" title="scranpy.dimensionality_reduction.run_umap.UmapEmbedding"><code class="xref py py-class docutils literal notranslate"><span class="pre">UmapEmbedding</span></code></a><span class="pre">]</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#id64" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-scranpy.analyze.analyze">
<span id="scranpy-analyze-analyze-module"></span><h2>scranpy.analyze.analyze module<a class="headerlink" href="#module-scranpy.analyze.analyze" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze.analyze.analyze">
<span class="sig-prename descclassname"><span class="pre">scranpy.analyze.analyze.</span></span><span class="sig-name descname"><span class="pre">analyze</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">AnalyzeOptions(per_cell_rna_qc_metrics_options=PerCellRnaQcMetricsOptions(subsets=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_rna_qc_filters_options=SuggestRnaQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_rna_qc_filter_options=CreateRnaQcFilterOptions(block=None),</span> <span class="pre">per_cell_adt_qc_metrics_options=PerCellAdtQcMetricsOptions(subsets=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_adt_qc_filters_options=SuggestAdtQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_adt_qc_filter_options=CreateAdtQcFilterOptions(block=None),</span> <span class="pre">per_cell_crispr_qc_metrics_options=PerCellCrisprQcMetricsOptions(assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_crispr_qc_filters_options=SuggestCrisprQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_crispr_qc_filter_options=CreateCrisprQcFilterOptions(block=None),</span> <span class="pre">filter_cells_options=FilterCellsOptions(discard=True,</span> <span class="pre">intersect=False,</span> <span class="pre">with_retain_vector=False,</span> <span class="pre">delayed=True),</span> <span class="pre">rna_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">grouped_size_factors_options=GroupedSizeFactorsOptions(rank=25,</span> <span class="pre">groups=None,</span> <span class="pre">block=None,</span> <span class="pre">initial_factors=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">adt_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">crispr_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">choose_hvgs_options=ChooseHvgsOptions(number=2500),</span> <span class="pre">model_gene_variances_options=ModelGeneVariancesOptions(block=None,</span> <span class="pre">span=0.3,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">rna_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">adt_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">crispr_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">combine_embeddings_options=CombineEmbeddingsOptions(neighbors=20,</span> <span class="pre">approximate=True,</span> <span class="pre">weights=None,</span> <span class="pre">num_threads=True),</span> <span class="pre">mnn_correct_options=MnnCorrectOptions(k=15,</span> <span class="pre">approximate=True,</span> <span class="pre">order=None,</span> <span class="pre">reference_policy='max-rss',</span> <span class="pre">num_mads=3,</span> <span class="pre">mass_cap=-1,</span> <span class="pre">num_threads=1),</span> <span class="pre">build_neighbor_index_options=BuildNeighborIndexOptions(approximate=True),</span> <span class="pre">run_tsne_options=RunTsneOptions(max_iterations=500,</span> <span class="pre">initialize_tsne=InitializeTsneOptions(perplexity=30,</span> <span class="pre">seed=42,</span> <span class="pre">num_threads=1)),</span> <span class="pre">run_umap_options=RunUmapOptions(initialize_umap=InitializeUmapOptions(min_dist=0.1,</span> <span class="pre">num_neighbors=15,</span> <span class="pre">num_epochs=500,</span> <span class="pre">seed=42,</span> <span class="pre">num_threads=1)),</span> <span class="pre">find_nearest_neighbors_options=FindNearestNeighborsOptions(num_threads=1),</span> <span class="pre">build_snn_graph_options=BuildSnnGraphOptions(num_neighbors=15,</span> <span class="pre">weight_scheme='ranked',</span> <span class="pre">build_neighbor_index_options=BuildNeighborIndexOptions(approximate=True),</span> <span class="pre">num_threads=1),</span> <span class="pre">rna_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">adt_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">crispr_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">miscellaneous_options=MiscellaneousOptions(cell_names=None,</span> <span class="pre">rna_feature_names=None,</span> <span class="pre">adt_feature_names=None,</span> <span class="pre">crispr_feature_names=None,</span> <span class="pre">filter_on_rna_qc=True,</span> <span class="pre">filter_on_adt_qc=True,</span> <span class="pre">filter_on_crispr_qc=True,</span> <span class="pre">snn_graph_multilevel_resolution=1,</span> <span class="pre">block=None))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dry_run</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/analyze.html#analyze"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.analyze.analyze" title="Permalink to this definition">#</a></dt>
<dd><p>Run a routine analysis workflow for single-cell datasets. This supports
RNA-seq, ADT and CRISPR modalities for the same cells. Steps include:</p>
<ul class="simple">
<li><p>Removal of low-quality cells</p></li>
<li><p>Normalization and log-transformation</p></li>
<li><p>Variance modelling across genes</p></li>
<li><p>PCA on highly variable genes</p></li>
<li><p>Combined embeddings across modalities</p></li>
<li><p>Batch correction with MNN</p></li>
<li><p>Graph-based clustering</p></li>
<li><p>Dimensionality reductions for visualization</p></li>
<li><p>Marker detection for each cluster</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rna_matrix</strong>  Count matrix for RNA data.
Alternatively None if no RNA data is available.</p></li>
<li><p><strong>adt_matrix</strong>  Count matrix for the ADT data.
Alternatively None if no ADT data is available.</p></li>
<li><p><strong>crispr_matrix</strong>  Count matrix for the CRISPR data.
Alternatively None if no CRISPR data is available.</p></li>
<li><p><strong>options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions" title="scranpy.analyze.AnalyzeOptions.AnalyzeOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeOptions</span></code></a></span>)  Optional analysis parameters.</p></li>
<li><p><strong>dry_run</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>)  Whether to perform a dry run.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/exceptions.html#NotImplementedError" title="(in Python v3.9)"><strong>NotImplementedError</strong></a>  If <code class="docutils literal notranslate"><span class="pre">matrix</span></code> is not an expected type.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults" title="scranpy.analyze.AnalyzeResults.AnalyzeResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>If <code class="docutils literal notranslate"><span class="pre">dry_run</span> <span class="pre">=</span> <span class="pre">False</span></code>, a <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults" title="scranpy.analyze.AnalyzeResults.AnalyzeResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a> object is returned
containing well, the analysis results, obviously.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">dry_run</span> <span class="pre">=</span> <span class="pre">True</span></code>, a string is returned containing all the steps required to perform the analysis.</p>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze.analyze.analyze_sce">
<span class="sig-prename descclassname"><span class="pre">scranpy.analyze.analyze.</span></span><span class="sig-name descname"><span class="pre">analyze_sce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sce</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_exp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_exp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_exp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">AnalyzeOptions(per_cell_rna_qc_metrics_options=PerCellRnaQcMetricsOptions(subsets=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_rna_qc_filters_options=SuggestRnaQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_rna_qc_filter_options=CreateRnaQcFilterOptions(block=None),</span> <span class="pre">per_cell_adt_qc_metrics_options=PerCellAdtQcMetricsOptions(subsets=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_adt_qc_filters_options=SuggestAdtQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_adt_qc_filter_options=CreateAdtQcFilterOptions(block=None),</span> <span class="pre">per_cell_crispr_qc_metrics_options=PerCellCrisprQcMetricsOptions(assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_crispr_qc_filters_options=SuggestCrisprQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_crispr_qc_filter_options=CreateCrisprQcFilterOptions(block=None),</span> <span class="pre">filter_cells_options=FilterCellsOptions(discard=True,</span> <span class="pre">intersect=False,</span> <span class="pre">with_retain_vector=False,</span> <span class="pre">delayed=True),</span> <span class="pre">rna_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">grouped_size_factors_options=GroupedSizeFactorsOptions(rank=25,</span> <span class="pre">groups=None,</span> <span class="pre">block=None,</span> <span class="pre">initial_factors=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">adt_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">crispr_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">choose_hvgs_options=ChooseHvgsOptions(number=2500),</span> <span class="pre">model_gene_variances_options=ModelGeneVariancesOptions(block=None,</span> <span class="pre">span=0.3,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">rna_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">adt_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">crispr_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">combine_embeddings_options=CombineEmbeddingsOptions(neighbors=20,</span> <span class="pre">approximate=True,</span> <span class="pre">weights=None,</span> <span class="pre">num_threads=True),</span> <span class="pre">mnn_correct_options=MnnCorrectOptions(k=15,</span> <span class="pre">approximate=True,</span> <span class="pre">order=None,</span> <span class="pre">reference_policy='max-rss',</span> <span class="pre">num_mads=3,</span> <span class="pre">mass_cap=-1,</span> <span class="pre">num_threads=1),</span> <span class="pre">build_neighbor_index_options=BuildNeighborIndexOptions(approximate=True),</span> <span class="pre">run_tsne_options=RunTsneOptions(max_iterations=500,</span> <span class="pre">initialize_tsne=InitializeTsneOptions(perplexity=30,</span> <span class="pre">seed=42,</span> <span class="pre">num_threads=1)),</span> <span class="pre">run_umap_options=RunUmapOptions(initialize_umap=InitializeUmapOptions(min_dist=0.1,</span> <span class="pre">num_neighbors=15,</span> <span class="pre">num_epochs=500,</span> <span class="pre">seed=42,</span> <span class="pre">num_threads=1)),</span> <span class="pre">find_nearest_neighbors_options=FindNearestNeighborsOptions(num_threads=1),</span> <span class="pre">build_snn_graph_options=BuildSnnGraphOptions(num_neighbors=15,</span> <span class="pre">weight_scheme='ranked',</span> <span class="pre">build_neighbor_index_options=BuildNeighborIndexOptions(approximate=True),</span> <span class="pre">num_threads=1),</span> <span class="pre">rna_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">adt_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">crispr_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">miscellaneous_options=MiscellaneousOptions(cell_names=None,</span> <span class="pre">rna_feature_names=None,</span> <span class="pre">adt_feature_names=None,</span> <span class="pre">crispr_feature_names=None,</span> <span class="pre">filter_on_rna_qc=True,</span> <span class="pre">filter_on_adt_qc=True,</span> <span class="pre">filter_on_crispr_qc=True,</span> <span class="pre">snn_graph_multilevel_resolution=1,</span> <span class="pre">block=None))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dry_run</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/analyze.html#analyze_sce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.analyze.analyze_sce" title="Permalink to this definition">#</a></dt>
<dd><p>Convenience wrapper around <a class="reference internal" href="#scranpy.analyze.analyze.analyze" title="scranpy.analyze.analyze.analyze"><code class="xref py py-meth docutils literal notranslate"><span class="pre">analyze()</span></code></a> for
<a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.4.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> inputs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sce</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.4.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span>)  A <code class="xref py py-class docutils literal notranslate"><span class="pre">singlecellexperiment.SingleCellExperiment</span></code> object,
possibly with data from other modalities in its alternative experiments.</p></li>
<li><p><strong>rna_exp</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>)  String or index specifying the alternative experiment containing the RNA data.
An empty string is assumed to refer to the main experiment.
If None, we assume that no RNA data is available.</p></li>
<li><p><strong>adt_exp</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>)  String or index specifying the alternative experiment containing the ADT data.
An empty string is assumed to refer to the main experiment.
If None, we assume that no RNA data is available.</p></li>
<li><p><strong>crispr_exp</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>)  String or index specifying the alternative experiment containing the CRISPR data.
An empty string is assumed to refer to the main experiment.
If None, we assume that no RNA data is available.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>)  Assay containing the count data in each SummarizedExperiment.</p></li>
<li><p><strong>options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions" title="scranpy.analyze.AnalyzeOptions.AnalyzeOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeOptions</span></code></a></span>)  Optional analysis parameters.</p></li>
<li><p><strong>dry_run</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>)  Whether to perform a dry run.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a>  If SCE does not contain a <code class="docutils literal notranslate"><span class="pre">assay</span></code> matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults" title="scranpy.analyze.AnalyzeResults.AnalyzeResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>If <code class="docutils literal notranslate"><span class="pre">dry_run</span> <span class="pre">=</span> <span class="pre">False</span></code>, a <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults" title="scranpy.analyze.AnalyzeResults.AnalyzeResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a> object is returned
containing well, the analysis results, obviously.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">dry_run</span> <span class="pre">=</span> <span class="pre">True</span></code>, a string is returned containing all the steps required to perform the analysis.</p>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze.analyze.analyze_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.analyze.analyze.</span></span><span class="sig-name descname"><span class="pre">analyze_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_se</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_se</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_se</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">AnalyzeOptions(per_cell_rna_qc_metrics_options=PerCellRnaQcMetricsOptions(subsets=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_rna_qc_filters_options=SuggestRnaQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_rna_qc_filter_options=CreateRnaQcFilterOptions(block=None),</span> <span class="pre">per_cell_adt_qc_metrics_options=PerCellAdtQcMetricsOptions(subsets=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_adt_qc_filters_options=SuggestAdtQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_adt_qc_filter_options=CreateAdtQcFilterOptions(block=None),</span> <span class="pre">per_cell_crispr_qc_metrics_options=PerCellCrisprQcMetricsOptions(assay_type=0,</span> <span class="pre">cell_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">suggest_crispr_qc_filters_options=SuggestCrisprQcFiltersOptions(block=None,</span> <span class="pre">num_mads=3,</span> <span class="pre">custom_thresholds=None),</span> <span class="pre">create_crispr_qc_filter_options=CreateCrisprQcFilterOptions(block=None),</span> <span class="pre">filter_cells_options=FilterCellsOptions(discard=True,</span> <span class="pre">intersect=False,</span> <span class="pre">with_retain_vector=False,</span> <span class="pre">delayed=True),</span> <span class="pre">rna_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">grouped_size_factors_options=GroupedSizeFactorsOptions(rank=25,</span> <span class="pre">groups=None,</span> <span class="pre">block=None,</span> <span class="pre">initial_factors=None,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">adt_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">crispr_log_norm_counts_options=LogNormCountsOptions(block=None,</span> <span class="pre">size_factors=None,</span> <span class="pre">center=True,</span> <span class="pre">center_size_factors_options=CenterSizeFactorsOptions(block=None,</span> <span class="pre">in_place=False,</span> <span class="pre">allow_zeros=False,</span> <span class="pre">allow_non_finite=False),</span> <span class="pre">delayed=True,</span> <span class="pre">with_size_factors=False,</span> <span class="pre">assay_type=0,</span> <span class="pre">num_threads=1),</span> <span class="pre">choose_hvgs_options=ChooseHvgsOptions(number=2500),</span> <span class="pre">model_gene_variances_options=ModelGeneVariancesOptions(block=None,</span> <span class="pre">span=0.3,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">rna_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">adt_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">crispr_run_pca_options=RunPcaOptions(rank=25,</span> <span class="pre">subset=None,</span> <span class="pre">block=None,</span> <span class="pre">scale=False,</span> <span class="pre">block_method='project',</span> <span class="pre">block_weights=True,</span> <span class="pre">num_threads=1,</span> <span class="pre">assay_type='logcounts'),</span> <span class="pre">combine_embeddings_options=CombineEmbeddingsOptions(neighbors=20,</span> <span class="pre">approximate=True,</span> <span class="pre">weights=None,</span> <span class="pre">num_threads=True),</span> <span class="pre">mnn_correct_options=MnnCorrectOptions(k=15,</span> <span class="pre">approximate=True,</span> <span class="pre">order=None,</span> <span class="pre">reference_policy='max-rss',</span> <span class="pre">num_mads=3,</span> <span class="pre">mass_cap=-1,</span> <span class="pre">num_threads=1),</span> <span class="pre">build_neighbor_index_options=BuildNeighborIndexOptions(approximate=True),</span> <span class="pre">run_tsne_options=RunTsneOptions(max_iterations=500,</span> <span class="pre">initialize_tsne=InitializeTsneOptions(perplexity=30,</span> <span class="pre">seed=42,</span> <span class="pre">num_threads=1)),</span> <span class="pre">run_umap_options=RunUmapOptions(initialize_umap=InitializeUmapOptions(min_dist=0.1,</span> <span class="pre">num_neighbors=15,</span> <span class="pre">num_epochs=500,</span> <span class="pre">seed=42,</span> <span class="pre">num_threads=1)),</span> <span class="pre">find_nearest_neighbors_options=FindNearestNeighborsOptions(num_threads=1),</span> <span class="pre">build_snn_graph_options=BuildSnnGraphOptions(num_neighbors=15,</span> <span class="pre">weight_scheme='ranked',</span> <span class="pre">build_neighbor_index_options=BuildNeighborIndexOptions(approximate=True),</span> <span class="pre">num_threads=1),</span> <span class="pre">rna_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">adt_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">crispr_score_markers_options=ScoreMarkersOptions(block=None,</span> <span class="pre">threshold=0,</span> <span class="pre">compute_auc=True,</span> <span class="pre">assay_type='logcounts',</span> <span class="pre">feature_names=None,</span> <span class="pre">num_threads=1),</span> <span class="pre">miscellaneous_options=MiscellaneousOptions(cell_names=None,</span> <span class="pre">rna_feature_names=None,</span> <span class="pre">adt_feature_names=None,</span> <span class="pre">crispr_feature_names=None,</span> <span class="pre">filter_on_rna_qc=True,</span> <span class="pre">filter_on_adt_qc=True,</span> <span class="pre">filter_on_crispr_qc=True,</span> <span class="pre">snn_graph_multilevel_resolution=1,</span> <span class="pre">block=None))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dry_run</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/analyze.html#analyze_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.analyze.analyze_se" title="Permalink to this definition">#</a></dt>
<dd><p>Convenience wrapper around <a class="reference internal" href="#scranpy.analyze.analyze.analyze" title="scranpy.analyze.analyze.analyze"><code class="xref py py-meth docutils literal notranslate"><span class="pre">analyze()</span></code></a> for
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.4.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> inputs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rna_se</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.4.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>]</span>)  SummarizedExperiment containing RNA data.
Alternatively None if no RNA data is available.</p></li>
<li><p><strong>adt_se</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.4.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>]</span>)  SummarizedExperiment containing ADT data.
Alternatively None if no ADT data is available.</p></li>
<li><p><strong>crispr_se</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.4.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>]</span>)  SummarizedExperiment containing CRISPR data.
Alternatively None if no CRISPR data is available.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>)  Assay containing the count data in each SummarizedExperiment.</p></li>
<li><p><strong>options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions" title="scranpy.analyze.AnalyzeOptions.AnalyzeOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeOptions</span></code></a></span>)  Optional analysis parameters.</p></li>
<li><p><strong>dry_run</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>)  Whether to perform a dry run.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/exceptions.html#NotImplementedError" title="(in Python v3.9)"><strong>NotImplementedError</strong></a>  If <code class="docutils literal notranslate"><span class="pre">matrix</span></code> is not an expected type.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults" title="scranpy.analyze.AnalyzeResults.AnalyzeResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>If <code class="docutils literal notranslate"><span class="pre">dry_run</span> <span class="pre">=</span> <span class="pre">False</span></code>, a <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults" title="scranpy.analyze.AnalyzeResults.AnalyzeResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a> object is returned
containing well, the analysis results, obviously.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">dry_run</span> <span class="pre">=</span> <span class="pre">True</span></code>, a string is returned containing all the steps required to perform the analysis.</p>
</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scranpy.analyze.dry_analyze">
<span id="scranpy-analyze-dry-analyze-module"></span><h2>scranpy.analyze.dry_analyze module<a class="headerlink" href="#module-scranpy.analyze.dry_analyze" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze.dry_analyze.dry_analyze">
<span class="sig-prename descclassname"><span class="pre">scranpy.analyze.dry_analyze.</span></span><span class="sig-name descname"><span class="pre">dry_analyze</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/dry_analyze.html#dry_analyze"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.dry_analyze.dry_analyze" title="Permalink to this definition">#</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scranpy.analyze.live_analyze">
<span id="scranpy-analyze-live-analyze-module"></span><h2>scranpy.analyze.live_analyze module<a class="headerlink" href="#module-scranpy.analyze.live_analyze" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze.live_analyze.live_analyze">
<span class="sig-prename descclassname"><span class="pre">scranpy.analyze.live_analyze.</span></span><span class="sig-name descname"><span class="pre">live_analyze</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/live_analyze.html#live_analyze"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.live_analyze.live_analyze" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

</section>
<section id="module-scranpy.analyze.run_neighbor_suite">
<span id="scranpy-analyze-run-neighbor-suite-module"></span><h2>scranpy.analyze.run_neighbor_suite module<a class="headerlink" href="#module-scranpy.analyze.run_neighbor_suite" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze.run_neighbor_suite.run_neighbor_suite">
<span class="sig-prename descclassname"><span class="pre">scranpy.analyze.run_neighbor_suite.</span></span><span class="sig-name descname"><span class="pre">run_neighbor_suite</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">principal_components</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_neighbor_index_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">BuildNeighborIndexOptions(approximate=True)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">find_nearest_neighbors_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">FindNearestNeighborsOptions(num_threads=1)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_umap_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">RunUmapOptions(initialize_umap=InitializeUmapOptions(min_dist=0.1,</span> <span class="pre">num_neighbors=15,</span> <span class="pre">num_epochs=500,</span> <span class="pre">seed=42,</span> <span class="pre">num_threads=1))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_tsne_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">RunTsneOptions(max_iterations=500,</span> <span class="pre">initialize_tsne=InitializeTsneOptions(perplexity=30,</span> <span class="pre">seed=42,</span> <span class="pre">num_threads=1))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_snn_graph_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">BuildSnnGraphOptions(num_neighbors=15,</span> <span class="pre">weight_scheme='ranked',</span> <span class="pre">build_neighbor_index_options=BuildNeighborIndexOptions(approximate=True),</span> <span class="pre">num_threads=1)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/run_neighbor_suite.html#run_neighbor_suite"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.run_neighbor_suite.run_neighbor_suite" title="Permalink to this definition">#</a></dt>
<dd><p>Run the suite of nearest neighbor methods together. This builds the index once and re-uses it for all methods.
Given enough threads, it also runs all post-neighbor-detection functions in parallel, as none of them depend on each
other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>principal_components</strong>  Matrix of principal components where rows are cells and columns are PCs.
Thi is usually produced by <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_pca.run_pca" title="scranpy.dimensionality_reduction.run_pca.run_pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p></li>
<li><p><strong>build_neighbor_index_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.build_neighbor_index.BuildNeighborIndexOptions" title="scranpy.nearest_neighbors.build_neighbor_index.BuildNeighborIndexOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">BuildNeighborIndexOptions</span></code></a></span>)  Optional arguments to pass to
<a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.build_neighbor_index.build_neighbor_index" title="scranpy.nearest_neighbors.build_neighbor_index.build_neighbor_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_neighbor_index()</span></code></a>.</p></li>
<li><p><strong>find_nearest_neighbors_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.find_nearest_neighbors.FindNearestNeighborsOptions" title="scranpy.nearest_neighbors.find_nearest_neighbors.FindNearestNeighborsOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindNearestNeighborsOptions</span></code></a></span>)  Optional arguments to pass to
<a class="reference internal" href="scranpy.nearest_neighbors.html#scranpy.nearest_neighbors.find_nearest_neighbors.find_nearest_neighbors" title="scranpy.nearest_neighbors.find_nearest_neighbors.find_nearest_neighbors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">find_nearest_neighbors()</span></code></a>.</p></li>
<li><p><strong>run_umap_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_umap.RunUmapOptions" title="scranpy.dimensionality_reduction.run_umap.RunUmapOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunUmapOptions</span></code></a></span>)  Optional arguments to pass to <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_umap.run_umap" title="scranpy.dimensionality_reduction.run_umap.run_umap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p></li>
<li><p><strong>run_tsne_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_tsne.RunTsneOptions" title="scranpy.dimensionality_reduction.run_tsne.RunTsneOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunTsneOptions</span></code></a></span>)  Optional arguments to pass to <a class="reference internal" href="scranpy.dimensionality_reduction.html#scranpy.dimensionality_reduction.run_tsne.run_tsne" title="scranpy.dimensionality_reduction.run_tsne.run_tsne"><code class="xref py py-meth docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.</p></li>
<li><p><strong>build_snn_graph_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="scranpy.clustering.html#scranpy.clustering.build_snn_graph.BuildSnnGraphOptions" title="scranpy.clustering.build_snn_graph.BuildSnnGraphOptions"><code class="xref py py-class docutils literal notranslate"><span class="pre">BuildSnnGraphOptions</span></code></a></span>)  Optional arguments to pass to <a class="reference internal" href="scranpy.clustering.html#scranpy.clustering.build_snn_graph.build_snn_graph" title="scranpy.clustering.build_snn_graph.build_snn_graph"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>)  Number of threads to use for the parallel execution of UMAP, t-SNE and SNN graph construction.
This overrides the specified number of threads in <code class="docutils literal notranslate"><span class="pre">run_umap</span></code>, <code class="docutils literal notranslate"><span class="pre">run_tsne</span></code> and <code class="docutils literal notranslate"><span class="pre">build_snn_graph</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A function that takes no arguments and returns a tuple containing the t-SNE and UMAP coordinates.</p></li>
<li><p>The shared nearest neighbor graph from <a class="reference internal" href="scranpy.clustering.html#scranpy.clustering.build_snn_graph.build_snn_graph" title="scranpy.clustering.build_snn_graph.build_snn_graph"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.</p></li>
<li><p>The number of remaining threads.</p></li>
</ul>
<p>The idea is that the number of remaining threads can be used to perform tasks on the main thread
(e.g., clustering, marker detection) while the t-SNE and UMAP are still being computed;
once all tasks on the main thread have completed, the first function can be called to obtain the coordinates.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>A tuple containing, in order</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scranpy.analyze.update">
<span id="scranpy-analyze-update-module"></span><h2>scranpy.analyze.update module<a class="headerlink" href="#module-scranpy.analyze.update" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze.update.update">
<span class="sig-prename descclassname"><span class="pre">scranpy.analyze.update.</span></span><span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">options</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze/update.html#update"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.update.update" title="Permalink to this definition">#</a></dt>
<dd><p>Convenience function to update the settings on an <code class="docutils literal notranslate"><span class="pre">*Options</span></code> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>options</strong>  Any of the <code class="docutils literal notranslate"><span class="pre">*Options</span></code> object.</p></li>
<li><p><strong>kwargs</strong>  Key-value pairs of settings to replace.</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Results:</dt><dd><p>A copy of <code class="docutils literal notranslate"><span class="pre">options</span></code> with replaced settings.
Note that the input <code class="docutils literal notranslate"><span class="pre">options</span></code> itself is unchanged.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scranpy.analyze">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-scranpy.analyze" title="Permalink to this heading">#</a></h2>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="scranpy.batch_correction.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">scranpy.batch_correction package</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="scranpy.aggregation.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">scranpy.aggregation package</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Jayaram Kancherla
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">scranpy.analyze package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-scranpy.analyze.AnalyzeOptions">scranpy.analyze.AnalyzeOptions module</a><ul>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_rna_qc_metrics_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.per_cell_rna_qc_metrics_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_rna_qc_filters_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.suggest_rna_qc_filters_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_rna_qc_filter_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.create_rna_qc_filter_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_adt_qc_metrics_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.per_cell_adt_qc_metrics_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_adt_qc_filters_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.suggest_adt_qc_filters_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_adt_qc_filter_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.create_adt_qc_filter_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.per_cell_crispr_qc_metrics_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.per_cell_crispr_qc_metrics_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.suggest_crispr_qc_filters_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.suggest_crispr_qc_filters_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.create_crispr_qc_filter_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.create_crispr_qc_filter_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.filter_cells_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.filter_cells_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_log_norm_counts_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.rna_log_norm_counts_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.grouped_size_factors_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.grouped_size_factors_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_log_norm_counts_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.adt_log_norm_counts_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_log_norm_counts_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.crispr_log_norm_counts_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.choose_hvgs_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.choose_hvgs_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.model_gene_variances_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.model_gene_variances_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_run_pca_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.rna_run_pca_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_run_pca_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.adt_run_pca_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_run_pca_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.crispr_run_pca_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.mnn_correct_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.mnn_correct_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.build_neighbor_index_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.build_neighbor_index_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.find_nearest_neighbors_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.find_nearest_neighbors_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.run_tsne_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.run_tsne_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.run_umap_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.run_umap_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.build_snn_graph_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.build_snn_graph_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.rna_score_markers_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.rna_score_markers_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.adt_score_markers_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.adt_score_markers_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.crispr_score_markers_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.crispr_score_markers_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.miscellaneous_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.miscellaneous_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__annotations__"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__eq__"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__hash__"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.__repr__"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#id0"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.adt_log_norm_counts_options</span></code></a></li>
<li><a class="reference internal" href="#id1"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.adt_run_pca_options</span></code></a></li>
<li><a class="reference internal" href="#id2"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.adt_score_markers_options</span></code></a></li>
<li><a class="reference internal" href="#id3"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.build_neighbor_index_options</span></code></a></li>
<li><a class="reference internal" href="#id4"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.build_snn_graph_options</span></code></a></li>
<li><a class="reference internal" href="#id5"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.choose_hvgs_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.combine_embeddings_options"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.combine_embeddings_options</span></code></a></li>
<li><a class="reference internal" href="#id6"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.create_adt_qc_filter_options</span></code></a></li>
<li><a class="reference internal" href="#id7"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.create_crispr_qc_filter_options</span></code></a></li>
<li><a class="reference internal" href="#id8"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.create_rna_qc_filter_options</span></code></a></li>
<li><a class="reference internal" href="#id9"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.crispr_log_norm_counts_options</span></code></a></li>
<li><a class="reference internal" href="#id10"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.crispr_run_pca_options</span></code></a></li>
<li><a class="reference internal" href="#id11"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.crispr_score_markers_options</span></code></a></li>
<li><a class="reference internal" href="#id12"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.filter_cells_options</span></code></a></li>
<li><a class="reference internal" href="#id13"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.find_nearest_neighbors_options</span></code></a></li>
<li><a class="reference internal" href="#id14"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.grouped_size_factors_options</span></code></a></li>
<li><a class="reference internal" href="#id15"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.miscellaneous_options</span></code></a></li>
<li><a class="reference internal" href="#id16"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.mnn_correct_options</span></code></a></li>
<li><a class="reference internal" href="#id17"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.model_gene_variances_options</span></code></a></li>
<li><a class="reference internal" href="#id18"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.per_cell_adt_qc_metrics_options</span></code></a></li>
<li><a class="reference internal" href="#id19"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.per_cell_crispr_qc_metrics_options</span></code></a></li>
<li><a class="reference internal" href="#id20"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.per_cell_rna_qc_metrics_options</span></code></a></li>
<li><a class="reference internal" href="#id21"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.rna_log_norm_counts_options</span></code></a></li>
<li><a class="reference internal" href="#id22"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.rna_run_pca_options</span></code></a></li>
<li><a class="reference internal" href="#id23"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.rna_score_markers_options</span></code></a></li>
<li><a class="reference internal" href="#id24"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.run_tsne_options</span></code></a></li>
<li><a class="reference internal" href="#id25"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.run_umap_options</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.set_seed"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.set_seed()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.AnalyzeOptions.set_threads"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.set_threads()</span></code></a></li>
<li><a class="reference internal" href="#id26"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.suggest_adt_qc_filters_options</span></code></a></li>
<li><a class="reference internal" href="#id27"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.suggest_crispr_qc_filters_options</span></code></a></li>
<li><a class="reference internal" href="#id28"><code class="docutils literal notranslate"><span class="pre">AnalyzeOptions.suggest_rna_qc_filters_options</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.cell_names"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.cell_names</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.rna_feature_names"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.rna_feature_names</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.adt_feature_names"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.adt_feature_names</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.crispr_feature_names"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.crispr_feature_names</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_rna_qc"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.filter_on_rna_qc</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_adt_qc"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.filter_on_adt_qc</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.filter_on_crispr_qc"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.filter_on_crispr_qc</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.snn_graph_multilevel_resolution"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.snn_graph_multilevel_resolution</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.block"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.block</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__annotations__"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__eq__"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__hash__"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeOptions.MiscellaneousOptions.__repr__"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#id29"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.adt_feature_names</span></code></a></li>
<li><a class="reference internal" href="#id30"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.block</span></code></a></li>
<li><a class="reference internal" href="#id31"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.cell_names</span></code></a></li>
<li><a class="reference internal" href="#id32"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.crispr_feature_names</span></code></a></li>
<li><a class="reference internal" href="#id33"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.filter_on_adt_qc</span></code></a></li>
<li><a class="reference internal" href="#id34"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.filter_on_crispr_qc</span></code></a></li>
<li><a class="reference internal" href="#id35"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.filter_on_rna_qc</span></code></a></li>
<li><a class="reference internal" href="#id36"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.rna_feature_names</span></code></a></li>
<li><a class="reference internal" href="#id37"><code class="docutils literal notranslate"><span class="pre">MiscellaneousOptions.snn_graph_multilevel_resolution</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.analyze.AnalyzeResults">scranpy.analyze.AnalyzeResults module</a><ul>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_metrics"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_quality_control_metrics</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_thresholds"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_quality_control_thresholds</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_quality_control_filter"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_quality_control_filter</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_metrics"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_quality_control_metrics</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_thresholds"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_quality_control_thresholds</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_quality_control_filter"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_quality_control_filter</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_metrics"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_quality_control_metrics</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_thresholds"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_quality_control_thresholds</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_quality_control_filter"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_quality_control_filter</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.quality_control_retained"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.quality_control_retained</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_size_factors"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_size_factors</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_size_factors"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_size_factors</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_size_factors"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_size_factors</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.gene_variances"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.gene_variances</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.hvgs"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.hvgs</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_pca"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_pca</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_pca"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_pca</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_pca"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_pca</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.combined_pcs"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.combined_pcs</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.mnn"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.mnn</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.tsne"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.tsne</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.umap"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.umap</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.snn_graph"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.snn_graph</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.clusters"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.clusters</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.rna_markers"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_markers</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.adt_markers"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_markers</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.crispr_markers"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_markers</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__eq__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__hash__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.__repr__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#id38"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_markers</span></code></a></li>
<li><a class="reference internal" href="#id39"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_pca</span></code></a></li>
<li><a class="reference internal" href="#id40"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_quality_control_filter</span></code></a></li>
<li><a class="reference internal" href="#id41"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_quality_control_metrics</span></code></a></li>
<li><a class="reference internal" href="#id42"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_quality_control_thresholds</span></code></a></li>
<li><a class="reference internal" href="#id43"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_size_factors</span></code></a></li>
<li><a class="reference internal" href="#id44"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.clusters</span></code></a></li>
<li><a class="reference internal" href="#id45"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.combined_pcs</span></code></a></li>
<li><a class="reference internal" href="#id46"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_markers</span></code></a></li>
<li><a class="reference internal" href="#id47"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_pca</span></code></a></li>
<li><a class="reference internal" href="#id48"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_quality_control_filter</span></code></a></li>
<li><a class="reference internal" href="#id49"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_quality_control_metrics</span></code></a></li>
<li><a class="reference internal" href="#id50"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_quality_control_thresholds</span></code></a></li>
<li><a class="reference internal" href="#id51"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_size_factors</span></code></a></li>
<li><a class="reference internal" href="#id52"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.gene_variances</span></code></a></li>
<li><a class="reference internal" href="#id53"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.hvgs</span></code></a></li>
<li><a class="reference internal" href="#id54"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.mnn</span></code></a></li>
<li><a class="reference internal" href="#id55"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.quality_control_retained</span></code></a></li>
<li><a class="reference internal" href="#id56"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_markers</span></code></a></li>
<li><a class="reference internal" href="#id57"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_pca</span></code></a></li>
<li><a class="reference internal" href="#id58"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_quality_control_filter</span></code></a></li>
<li><a class="reference internal" href="#id59"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_quality_control_metrics</span></code></a></li>
<li><a class="reference internal" href="#id60"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_quality_control_thresholds</span></code></a></li>
<li><a class="reference internal" href="#id61"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_size_factors</span></code></a></li>
<li><a class="reference internal" href="#id62"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.snn_graph</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.AnalyzeResults.to_sce"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.to_sce()</span></code></a></li>
<li><a class="reference internal" href="#id63"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.tsne</span></code></a></li>
<li><a class="reference internal" href="#id64"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.umap</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.analyze.analyze">scranpy.analyze.analyze module</a><ul>
<li><a class="reference internal" href="#scranpy.analyze.analyze.analyze"><code class="docutils literal notranslate"><span class="pre">analyze()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.analyze.analyze_sce"><code class="docutils literal notranslate"><span class="pre">analyze_sce()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.analyze.analyze_se"><code class="docutils literal notranslate"><span class="pre">analyze_se()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.analyze.dry_analyze">scranpy.analyze.dry_analyze module</a><ul>
<li><a class="reference internal" href="#scranpy.analyze.dry_analyze.dry_analyze"><code class="docutils literal notranslate"><span class="pre">dry_analyze()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.analyze.live_analyze">scranpy.analyze.live_analyze module</a><ul>
<li><a class="reference internal" href="#scranpy.analyze.live_analyze.live_analyze"><code class="docutils literal notranslate"><span class="pre">live_analyze()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.analyze.run_neighbor_suite">scranpy.analyze.run_neighbor_suite module</a><ul>
<li><a class="reference internal" href="#scranpy.analyze.run_neighbor_suite.run_neighbor_suite"><code class="docutils literal notranslate"><span class="pre">run_neighbor_suite()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.analyze.update">scranpy.analyze.update module</a><ul>
<li><a class="reference internal" href="#scranpy.analyze.update.update"><code class="docutils literal notranslate"><span class="pre">update()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.analyze">Module contents</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=680b02e7"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/scripts/furo.js?v=32e29ea5"></script>
    </body>
</html>