<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="prev" title="scranpy" href="modules.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.08.06 -->
        <title>scranpy package - scranpy 0.2.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">scranpy 0.2.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">scranpy 0.2.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="modules.html">Module Reference</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">scranpy package</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/api/scranpy.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="scranpy-package">
<h1>scranpy package<a class="headerlink" href="#scranpy-package" title="Permalink to this heading">¶</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading">¶</a></h2>
</section>
<section id="module-scranpy.adt_quality_control">
<span id="scranpy-adt-quality-control-module"></span><h2>scranpy.adt_quality_control module<a class="headerlink" href="#module-scranpy.adt_quality_control" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.adt_quality_control.</span></span><span class="sig-name descname"><span class="pre">ComputeAdtQcMetricsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">detected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_sum</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/adt_quality_control.html#ComputeAdtQcMetricsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.adt_quality_control.compute_adt_qc_metrics" title="scranpy.adt_quality_control.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'detected':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'subset_sum':</span> <span class="pre">&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,</span> <span class="pre">'sum':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'detected':</span> <span class="pre">Field(name='detected',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'subset_sum':</span> <span class="pre">Field(name='subset_sum',type=&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'sum':</span> <span class="pre">Field(name='sum',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('sum',</span> <span class="pre">'detected',</span> <span class="pre">'subset_sum')</span></em><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.detected">
<span class="sig-name descname"><span class="pre">detected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.detected" title="Permalink to this definition">¶</a></dt>
<dd><p>Integer array of length equal to the number of cells, containing the number of detected ADTs in each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.subset_sum">
<span class="sig-name descname"><span class="pre">subset_sum</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></em><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.subset_sum" title="Permalink to this definition">¶</a></dt>
<dd><p>List of length equal to the number of <code class="docutils literal notranslate"><span class="pre">subsets</span></code> in <a class="reference internal" href="#scranpy.adt_quality_control.compute_adt_qc_metrics" title="scranpy.adt_quality_control.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.
Each element corresponds to a subset of ADTs and is a NumPy array of length equal to the number of cells.
Each entry of the array contains the sum of counts for that subset in each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.sum">
<span class="sig-name descname"><span class="pre">sum</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of cells, containing the sum of counts across all ADTs for each cell.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.to_biocframe">
<span class="sig-name descname"><span class="pre">to_biocframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flatten</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/adt_quality_control.html#ComputeAdtQcMetricsResults.to_biocframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.to_biocframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the results into a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>flatten</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to flatten the subset sums into separate columns.
If <code class="docutils literal notranslate"><span class="pre">True</span></code>, each entry of <a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.subset_sum" title="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.subset_sum"><code class="xref py py-attr docutils literal notranslate"><span class="pre">subset_sum</span></code></a> is represented by a <code class="docutils literal notranslate"><span class="pre">subset_sum_&lt;NAME&gt;</span></code> column,
where <code class="docutils literal notranslate"><span class="pre">&lt;NAME&gt;</span></code> is the the name of each entry (if available) or its index (otherwise).
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, <a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.subset_sum" title="scranpy.adt_quality_control.ComputeAdtQcMetricsResults.subset_sum"><code class="xref py py-attr docutils literal notranslate"><span class="pre">subset_sum</span></code></a> is represented by a nested <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> where each row corresponds to a cell and each column is one of the metrics.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.adt_quality_control.</span></span><span class="sig-name descname"><span class="pre">SuggestAdtQcThresholdsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">detected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_sum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/adt_quality_control.html#SuggestAdtQcThresholdsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'block':</span> <span class="pre">typing.Optional[list],</span> <span class="pre">'detected':</span> <span class="pre">typing.Union[biocutils.NamedList.NamedList,</span> <span class="pre">float],</span> <span class="pre">'subset_sum':</span> <span class="pre">&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;}</span></em><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'block':</span> <span class="pre">Field(name='block',type=typing.Optional[list],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'detected':</span> <span class="pre">Field(name='detected',type=typing.Union[biocutils.NamedList.NamedList,</span> <span class="pre">float],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'subset_sum':</span> <span class="pre">Field(name='subset_sum',type=&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('detected',</span> <span class="pre">'subset_sum',</span> <span class="pre">'block')</span></em><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.block">
<span class="sig-name descname"><span class="pre">block</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.block" title="Permalink to this definition">¶</a></dt>
<dd><p>Levels of the blocking factor.
Each entry corresponds to a element of <a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.detected" title="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.detected"><code class="xref py py-attr docutils literal notranslate"><span class="pre">detected</span></code></a>, etc., if <code class="docutils literal notranslate"><span class="pre">block</span></code> was provided in <a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.
This is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> if no blocking was performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.detected">
<span class="sig-name descname"><span class="pre">detected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">,</span> <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.detected" title="Permalink to this definition">¶</a></dt>
<dd><p>Threshold on the number of detected ADTs.
Cells with lower numbers of detected ADTs are considered to be of low quality.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided in <a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>, a list is returned containing a separate threshold for each level of the factor.
Otherwise, a single float is returned containing the threshold for all cells.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.subset_sum">
<span class="sig-name descname"><span class="pre">subset_sum</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></em><a class="headerlink" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.subset_sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Thresholds on the sum of counts in each ADT subset.
Each element of the list corresponds to a ADT subset.
Cells with higher sums than the threshold for any subset are considered to be of low quality.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided in <a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>, each entry of the returned list is another <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>  containing a separate threshold for each level.
Otherwise, each entry of the list is a single float containing the threshold for all cells.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.compute_adt_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.adt_quality_control.</span></span><span class="sig-name descname"><span class="pre">compute_adt_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/adt_quality_control.html#compute_adt_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.adt_quality_control.compute_adt_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute quality control metrics from ADT count data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object containing ADT counts.</p></li>
<li><p><strong>subsets</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>, <a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – <p>Subsets of ADTs corresponding to control features like IgGs.
This may be either:</p>
<ul>
<li><p>A list of arrays.
Each array corresponds to an ADT subset and can either contain boolean or integer values.
For booleans, the array should be of length equal to the number of rows, and values should be truthy for rows that belong in the subset.
For integers, each element of the array is treated the row index of an ADT in the subset.</p></li>
<li><p>A dictionary where keys are the names of each ADT subset and the values are arrays as described above.</p></li>
<li><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> where each element is an array as described above, possibly with names.</p></li>
</ul>
</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults" title="scranpy.adt_quality_control.ComputeAdtQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>QC metrics computed from the ADT count matrix for each cell.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_adt_qc_metrics</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind these QC metrics.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.filter_adt_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.adt_quality_control.</span></span><span class="sig-name descname"><span class="pre">filter_adt_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/adt_quality_control.html#filter_adt_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.adt_quality_control.filter_adt_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter for high-quality cells based on ADT-derived QC metrics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thresholds</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults" title="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults</span></code></a></span>) – Filter thresholds on the QC metrics, typically computed with <a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults" title="scranpy.adt_quality_control.ComputeAdtQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults</span></code></a></span>) – ADT-derived QC metrics, typically computed with <a class="reference internal" href="#scranpy.adt_quality_control.compute_adt_qc_metrics" title="scranpy.adt_quality_control.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
The levels should be a subset of those used in <a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A NumPy vector of length equal to the number of cells in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>, containing truthy values for putative high-quality cells.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.adt_quality_control.suggest_adt_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.adt_quality_control.</span></span><span class="sig-name descname"><span class="pre">suggest_adt_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_detected_drop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/adt_quality_control.html#suggest_adt_qc_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Suggest filter thresholds for the ADT-derived QC metrics, typically generated from <a class="reference internal" href="#scranpy.adt_quality_control.compute_adt_qc_metrics" title="scranpy.adt_quality_control.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults" title="scranpy.adt_quality_control.ComputeAdtQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults</span></code></a></span>) – ADT-derived QC metrics from <a class="reference internal" href="#scranpy.adt_quality_control.compute_adt_qc_metrics" title="scranpy.adt_quality_control.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>min_detected_drop</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum proportional drop in the number of detected ADTs to consider a cell to be of low quality.
Specifically, the filter threshold on <code class="docutils literal notranslate"><span class="pre">metrics.detected</span></code> must be no higher than the product of <code class="docutils literal notranslate"><span class="pre">min_detected_drop</span></code> and the median number of ADTs, regardless of <code class="docutils literal notranslate"><span class="pre">num_mads</span></code>.</p></li>
<li><p><strong>num_mads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number of MADs from the median to define the threshold for outliers in each QC metric.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults" title="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Suggested filters on the relevant QC metrics.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_adt_qc_filters</span></code> and <code class="docutils literal notranslate"><span class="pre">compute_adt_qc_filters_blocked</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind the suggested filters.</p>
</dd></dl>

</section>
<section id="module-scranpy.aggregate_across_cells">
<span id="scranpy-aggregate-across-cells-module"></span><h2>scranpy.aggregate_across_cells module<a class="headerlink" href="#module-scranpy.aggregate_across_cells" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.aggregate_across_cells.</span></span><span class="sig-name descname"><span class="pre">AggregateAcrossCellsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">detected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combinations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/aggregate_across_cells.html#AggregateAcrossCellsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.aggregate_across_cells.aggregate_across_cells" title="scranpy.aggregate_across_cells.aggregate_across_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'combinations':</span> <span class="pre">&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,</span> <span class="pre">'counts':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'detected':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'index':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'sum':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'combinations':</span> <span class="pre">Field(name='combinations',type=&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'counts':</span> <span class="pre">Field(name='counts',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'detected':</span> <span class="pre">Field(name='detected',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'index':</span> <span class="pre">Field(name='index',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'sum':</span> <span class="pre">Field(name='sum',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('sum',</span> <span class="pre">'detected',</span> <span class="pre">'combinations',</span> <span class="pre">'counts',</span> <span class="pre">'index')</span></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations">
<span class="sig-name descname"><span class="pre">combinations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations" title="Permalink to this definition">¶</a></dt>
<dd><p>Sorted and unique combination of levels across all <code class="docutils literal notranslate"><span class="pre">factors</span></code> in <a class="reference internal" href="#scranpy.aggregate_across_cells.aggregate_across_cells" title="scranpy.aggregate_across_cells.aggregate_across_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a>.
Each entry of the list is another list that corresponds to an entry of <code class="docutils literal notranslate"><span class="pre">factors</span></code>, where the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th combination is defined as the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th elements of all inner lists.
Combinations are in the same order as the columns of <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.sum" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.sum"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sum</span></code></a> and <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.detected" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.detected"><code class="xref py py-attr docutils literal notranslate"><span class="pre">detected</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of cells associated with each combination.
Each entry corresponds to a combination in <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">combinations</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.detected">
<span class="sig-name descname"><span class="pre">detected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.detected" title="Permalink to this definition">¶</a></dt>
<dd><p>Integer matrix where each row corresponds to a gene and each column corresponds to a unique combination of grouping levels.
Each entry contains the number of cells with detected expression in that combination.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.index">
<span class="sig-name descname"><span class="pre">index</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Integer vector of length equal to the number of cells.
This specifies the combination in <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">combinations</span></code></a> associated with each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.sum">
<span class="sig-name descname"><span class="pre">sum</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point matrix where each row corresponds to a gene and each column corresponds to a unique combination of grouping levels.
Each matrix entry contains the summed expression across all cells with that combination.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.to_summarizedexperiment">
<span class="sig-name descname"><span class="pre">to_summarizedexperiment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/aggregate_across_cells.html#AggregateAcrossCellsResults.to_summarizedexperiment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.to_summarizedexperiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the results to a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>include_counts</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to include <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.counts" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.counts"><code class="xref py py-attr docutils literal notranslate"><span class="pre">counts</span></code></a> in the column data.
Users may need to set this to <code class="docutils literal notranslate"><span class="pre">False</span></code> if a <code class="docutils literal notranslate"><span class="pre">&quot;counts&quot;</span></code> factor is present in <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">combinations</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
where <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.sum" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.sum"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sum</span></code></a> and <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.detected" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.detected"><code class="xref py py-attr docutils literal notranslate"><span class="pre">detected</span></code></a> are assays and <a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">combinations</span></code></a> is stored in the column data.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells.aggregate_across_cells">
<span class="sig-prename descclassname"><span class="pre">scranpy.aggregate_across_cells.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/aggregate_across_cells.html#aggregate_across_cells"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.aggregate_across_cells.aggregate_across_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Aggregate expression values across cells based on one or more grouping factors.
This is primarily used to create pseudo-bulk profiles for each cluster/sample combination.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
Values are expected to be counts.</p></li>
<li><p><strong>factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – One or more grouping factors, see <a class="reference internal" href="#scranpy.combine_factors.combine_factors" title="scranpy.combine_factors.combine_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_factors()</span></code></a>.
If this is a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, any names will be retained in the output.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for aggregation.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults" title="scranpy.aggregate_across_cells.AggregateAcrossCellsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Results of the aggregation, including the sum and the number of detected cells in each group for each gene.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">aggregate_across_cells</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_aggregate">scran_aggregate</a> C++ library, which implements the aggregation.</p>
</dd></dl>

</section>
<section id="module-scranpy.aggregate_across_genes">
<span id="scranpy-aggregate-across-genes-module"></span><h2>scranpy.aggregate_across_genes module<a class="headerlink" href="#module-scranpy.aggregate_across_genes" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.aggregate_across_genes.aggregate_across_genes">
<span class="sig-prename descclassname"><span class="pre">scranpy.aggregate_across_genes.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">average</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/aggregate_across_genes.html#aggregate_across_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.aggregate_across_genes.aggregate_across_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Aggregate expression values across genes, potentially with weights.
This is typically used to summarize expression values for gene sets into a single per-cell score.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – Matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
Values are expected to be log-expression values.</p></li>
<li><p><strong>sets</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of integer arrays containing the row indices of genes in each set.
Alternatively, each entry may be a tuple of length 2, containing an integer vector (row indices) and a numeric vector (weights).
If this is a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, the names will be preserved in the output.</p></li>
<li><p><strong>average</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the average rather than the sum.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to be used for aggregation.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of length equal to that of <code class="docutils literal notranslate"><span class="pre">sets</span></code>.
Each entry is a numeric vector of length equal to the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>,
containing the (weighted) sum/mean of expression values for the corresponding set across all cells.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">aggregate_across_genes</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_aggregate">scran_aggregate</a> C++ library, which implements the aggregation.</p>
</dd></dl>

</section>
<section id="module-scranpy.analyze">
<span id="scranpy-analyze-module"></span><h2>scranpy.analyze module<a class="headerlink" href="#module-scranpy.analyze" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.analyze.</span></span><span class="sig-name descname"><span class="pre">AnalyzeResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_qc_metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_qc_thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_qc_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_qc_metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_qc_thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_qc_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_qc_metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_qc_thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_qc_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combined_qc_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_filtered</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_filtered</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_filtered</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_normalized</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_normalized</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_normalized</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_gene_variances</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_highly_variable_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_pca</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_pca</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_pca</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combined_pca</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mnn_corrected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tsne</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snn_graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph_clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kmeans_clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_markers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_markers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_markers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rna_row_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adt_row_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crispr_row_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze.html#AnalyzeResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.AnalyzeResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <code class="xref py py-func docutils literal notranslate"><span class="pre">analyse()</span></code>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'adt_filtered':</span> <span class="pre">typing.Optional[delayedarray.DelayedArray.DelayedArray],</span> <span class="pre">'adt_markers':</span> <span class="pre">typing.Optional[scranpy.run_pca.RunPcaResults],</span> <span class="pre">'adt_normalized':</span> <span class="pre">typing.Optional[delayedarray.DelayedArray.DelayedArray],</span> <span class="pre">'adt_pca':</span> <span class="pre">typing.Optional[scranpy.run_pca.RunPcaResults],</span> <span class="pre">'adt_qc_filter':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'adt_qc_metrics':</span> <span class="pre">typing.Optional[scranpy.adt_quality_control.ComputeAdtQcMetricsResults],</span> <span class="pre">'adt_qc_thresholds':</span> <span class="pre">typing.Optional[scranpy.adt_quality_control.SuggestAdtQcThresholdsResults],</span> <span class="pre">'adt_row_names':</span> <span class="pre">&lt;class</span> <span class="pre">'biocutils.Names.Names'&gt;,</span> <span class="pre">'adt_size_factors':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'block':</span> <span class="pre">typing.Optional[typing.Sequence],</span> <span class="pre">'clusters':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'column_names':</span> <span class="pre">&lt;class</span> <span class="pre">'biocutils.Names.Names'&gt;,</span> <span class="pre">'combined_pca':</span> <span class="pre">typing.Union[typing.Literal['rna_pca',</span> <span class="pre">'adt_pca',</span> <span class="pre">'crispr_pca'],</span> <span class="pre">scranpy.scale_by_neighbors.ScaleByNeighborsResults],</span> <span class="pre">'combined_qc_filter':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'crispr_filtered':</span> <span class="pre">typing.Optional[delayedarray.DelayedArray.DelayedArray],</span> <span class="pre">'crispr_markers':</span> <span class="pre">typing.Optional[scranpy.run_pca.RunPcaResults],</span> <span class="pre">'crispr_normalized':</span> <span class="pre">typing.Optional[delayedarray.DelayedArray.DelayedArray],</span> <span class="pre">'crispr_pca':</span> <span class="pre">typing.Optional[scranpy.run_pca.RunPcaResults],</span> <span class="pre">'crispr_qc_filter':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'crispr_qc_metrics':</span> <span class="pre">typing.Optional[scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults],</span> <span class="pre">'crispr_qc_thresholds':</span> <span class="pre">typing.Optional[scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults],</span> <span class="pre">'crispr_row_names':</span> <span class="pre">&lt;class</span> <span class="pre">'biocutils.Names.Names'&gt;,</span> <span class="pre">'crispr_size_factors':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'graph_clusters':</span> <span class="pre">typing.Optional[scranpy.cluster_graph.ClusterGraphResults],</span> <span class="pre">'kmeans_clusters':</span> <span class="pre">typing.Optional[scranpy.cluster_graph.ClusterGraphResults],</span> <span class="pre">'mnn_corrected':</span> <span class="pre">typing.Optional[scranpy.correct_mnn.CorrectMnnResults],</span> <span class="pre">'rna_filtered':</span> <span class="pre">typing.Optional[delayedarray.DelayedArray.DelayedArray],</span> <span class="pre">'rna_gene_variances':</span> <span class="pre">typing.Optional[scranpy.model_gene_variances.ModelGeneVariancesResults],</span> <span class="pre">'rna_highly_variable_genes':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'rna_markers':</span> <span class="pre">typing.Optional[scranpy.run_pca.RunPcaResults],</span> <span class="pre">'rna_normalized':</span> <span class="pre">typing.Optional[delayedarray.DelayedArray.DelayedArray],</span> <span class="pre">'rna_pca':</span> <span class="pre">typing.Optional[scranpy.run_pca.RunPcaResults],</span> <span class="pre">'rna_qc_filter':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'rna_qc_metrics':</span> <span class="pre">typing.Optional[scranpy.rna_quality_control.ComputeRnaQcMetricsResults],</span> <span class="pre">'rna_qc_thresholds':</span> <span class="pre">typing.Optional[scranpy.rna_quality_control.SuggestRnaQcThresholdsResults],</span> <span class="pre">'rna_row_names':</span> <span class="pre">typing.Optional[biocutils.Names.Names],</span> <span class="pre">'rna_size_factors':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'snn_graph':</span> <span class="pre">typing.Optional[scranpy.build_snn_graph.GraphComponents],</span> <span class="pre">'tsne':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'umap':</span> <span class="pre">typing.Optional[numpy.ndarray]}</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'adt_filtered':</span> <span class="pre">Field(name='adt_filtered',type=typing.Optional[delayedarray.DelayedArray.DelayedArray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'adt_markers':</span> <span class="pre">Field(name='adt_markers',type=typing.Optional[scranpy.run_pca.RunPcaResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'adt_normalized':</span> <span class="pre">Field(name='adt_normalized',type=typing.Optional[delayedarray.DelayedArray.DelayedArray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'adt_pca':</span> <span class="pre">Field(name='adt_pca',type=typing.Optional[scranpy.run_pca.RunPcaResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'adt_qc_filter':</span> <span class="pre">Field(name='adt_qc_filter',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'adt_qc_metrics':</span> <span class="pre">Field(name='adt_qc_metrics',type=typing.Optional[scranpy.adt_quality_control.ComputeAdtQcMetricsResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'adt_qc_thresholds':</span> <span class="pre">Field(name='adt_qc_thresholds',type=typing.Optional[scranpy.adt_quality_control.SuggestAdtQcThresholdsResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'adt_row_names':</span> <span class="pre">Field(name='adt_row_names',type=&lt;class</span> <span class="pre">'biocutils.Names.Names'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'adt_size_factors':</span> <span class="pre">Field(name='adt_size_factors',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'block':</span> <span class="pre">Field(name='block',type=typing.Optional[typing.Sequence],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'clusters':</span> <span class="pre">Field(name='clusters',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'column_names':</span> <span class="pre">Field(name='column_names',type=&lt;class</span> <span class="pre">'biocutils.Names.Names'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'combined_pca':</span> <span class="pre">Field(name='combined_pca',type=typing.Union[typing.Literal['rna_pca',</span> <span class="pre">'adt_pca',</span> <span class="pre">'crispr_pca'],</span> <span class="pre">scranpy.scale_by_neighbors.ScaleByNeighborsResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'combined_qc_filter':</span> <span class="pre">Field(name='combined_qc_filter',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_filtered':</span> <span class="pre">Field(name='crispr_filtered',type=typing.Optional[delayedarray.DelayedArray.DelayedArray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_markers':</span> <span class="pre">Field(name='crispr_markers',type=typing.Optional[scranpy.run_pca.RunPcaResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_normalized':</span> <span class="pre">Field(name='crispr_normalized',type=typing.Optional[delayedarray.DelayedArray.DelayedArray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_pca':</span> <span class="pre">Field(name='crispr_pca',type=typing.Optional[scranpy.run_pca.RunPcaResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_qc_filter':</span> <span class="pre">Field(name='crispr_qc_filter',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_qc_metrics':</span> <span class="pre">Field(name='crispr_qc_metrics',type=typing.Optional[scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_qc_thresholds':</span> <span class="pre">Field(name='crispr_qc_thresholds',type=typing.Optional[scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_row_names':</span> <span class="pre">Field(name='crispr_row_names',type=&lt;class</span> <span class="pre">'biocutils.Names.Names'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'crispr_size_factors':</span> <span class="pre">Field(name='crispr_size_factors',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'graph_clusters':</span> <span class="pre">Field(name='graph_clusters',type=typing.Optional[scranpy.cluster_graph.ClusterGraphResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'kmeans_clusters':</span> <span class="pre">Field(name='kmeans_clusters',type=typing.Optional[scranpy.cluster_graph.ClusterGraphResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'mnn_corrected':</span> <span class="pre">Field(name='mnn_corrected',type=typing.Optional[scranpy.correct_mnn.CorrectMnnResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_filtered':</span> <span class="pre">Field(name='rna_filtered',type=typing.Optional[delayedarray.DelayedArray.DelayedArray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_gene_variances':</span> <span class="pre">Field(name='rna_gene_variances',type=typing.Optional[scranpy.model_gene_variances.ModelGeneVariancesResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_highly_variable_genes':</span> <span class="pre">Field(name='rna_highly_variable_genes',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_markers':</span> <span class="pre">Field(name='rna_markers',type=typing.Optional[scranpy.run_pca.RunPcaResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_normalized':</span> <span class="pre">Field(name='rna_normalized',type=typing.Optional[delayedarray.DelayedArray.DelayedArray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_pca':</span> <span class="pre">Field(name='rna_pca',type=typing.Optional[scranpy.run_pca.RunPcaResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_qc_filter':</span> <span class="pre">Field(name='rna_qc_filter',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_qc_metrics':</span> <span class="pre">Field(name='rna_qc_metrics',type=typing.Optional[scranpy.rna_quality_control.ComputeRnaQcMetricsResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_qc_thresholds':</span> <span class="pre">Field(name='rna_qc_thresholds',type=typing.Optional[scranpy.rna_quality_control.SuggestRnaQcThresholdsResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_row_names':</span> <span class="pre">Field(name='rna_row_names',type=typing.Optional[biocutils.Names.Names],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rna_size_factors':</span> <span class="pre">Field(name='rna_size_factors',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'snn_graph':</span> <span class="pre">Field(name='snn_graph',type=typing.Optional[scranpy.build_snn_graph.GraphComponents],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'tsne':</span> <span class="pre">Field(name='tsne',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'umap':</span> <span class="pre">Field(name='umap',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('rna_qc_metrics',</span> <span class="pre">'rna_qc_thresholds',</span> <span class="pre">'rna_qc_filter',</span> <span class="pre">'adt_qc_metrics',</span> <span class="pre">'adt_qc_thresholds',</span> <span class="pre">'adt_qc_filter',</span> <span class="pre">'crispr_qc_metrics',</span> <span class="pre">'crispr_qc_thresholds',</span> <span class="pre">'crispr_qc_filter',</span> <span class="pre">'combined_qc_filter',</span> <span class="pre">'rna_filtered',</span> <span class="pre">'adt_filtered',</span> <span class="pre">'crispr_filtered',</span> <span class="pre">'rna_size_factors',</span> <span class="pre">'rna_normalized',</span> <span class="pre">'adt_size_factors',</span> <span class="pre">'adt_normalized',</span> <span class="pre">'crispr_size_factors',</span> <span class="pre">'crispr_normalized',</span> <span class="pre">'rna_gene_variances',</span> <span class="pre">'rna_highly_variable_genes',</span> <span class="pre">'rna_pca',</span> <span class="pre">'adt_pca',</span> <span class="pre">'crispr_pca',</span> <span class="pre">'combined_pca',</span> <span class="pre">'block',</span> <span class="pre">'mnn_corrected',</span> <span class="pre">'tsne',</span> <span class="pre">'umap',</span> <span class="pre">'snn_graph',</span> <span class="pre">'graph_clusters',</span> <span class="pre">'kmeans_clusters',</span> <span class="pre">'clusters',</span> <span class="pre">'rna_markers',</span> <span class="pre">'adt_markers',</span> <span class="pre">'crispr_markers',</span> <span class="pre">'rna_row_names',</span> <span class="pre">'adt_row_names',</span> <span class="pre">'crispr_row_names',</span> <span class="pre">'column_names')</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_filtered">
<span class="sig-name descname"><span class="pre">adt_filtered</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_filtered" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix of ADT counts that has been filtered to only contain the high-quality cells in <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.combined_qc_filter" title="scranpy.analyze.AnalyzeResults.combined_qc_filter"><code class="xref py py-attr docutils literal notranslate"><span class="pre">combined_qc_filter</span></code></a>.
If ADT data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_markers">
<span class="sig-name descname"><span class="pre">adt_markers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.run_pca.RunPcaResults" title="scranpy.run_pca.RunPcaResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of calling <a class="reference internal" href="#scranpy.score_markers.score_markers" title="scranpy.score_markers.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a> on the ADT log-expression matrix.
If ADT data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.
This will also be <code class="docutils literal notranslate"><span class="pre">None</span></code> if no suitable clusterings are available.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_normalized">
<span class="sig-name descname"><span class="pre">adt_normalized</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_normalized" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix of (log-)normalized expression values derived from ADT counts, as computed by <a class="reference internal" href="#scranpy.normalize_counts.normalize_counts" title="scranpy.normalize_counts.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a> using <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_size_factors" title="scranpy.analyze.AnalyzeResults.adt_size_factors"><code class="xref py py-attr docutils literal notranslate"><span class="pre">adt_size_factors</span></code></a>.
If ADT data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_pca">
<span class="sig-name descname"><span class="pre">adt_pca</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.run_pca.RunPcaResults" title="scranpy.run_pca.RunPcaResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of calling <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a> on the ADT log-expression matrix.
If ADT data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_qc_filter">
<span class="sig-name descname"><span class="pre">adt_qc_filter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_qc_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.adt_quality_control.filter_adt_qc_metrics" title="scranpy.adt_quality_control.filter_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter_adt_qc_metrics()</span></code></a>.
If ADT data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_qc_metrics">
<span class="sig-name descname"><span class="pre">adt_qc_metrics</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults" title="scranpy.adt_quality_control.ComputeAdtQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.adt_quality_control.compute_adt_qc_metrics" title="scranpy.adt_quality_control.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.
If ADT data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_qc_thresholds">
<span class="sig-name descname"><span class="pre">adt_qc_thresholds</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults" title="scranpy.adt_quality_control.SuggestAdtQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.
If ADT data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_row_names">
<span class="sig-name descname"><span class="pre">adt_row_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Names</span></code></a></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_row_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Names for the tags in the ADT data.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if ADT data is not available or no names were supplied to <a class="reference internal" href="#scranpy.analyze.analyze" title="scranpy.analyze.analyze"><code class="xref py py-func docutils literal notranslate"><span class="pre">analyze()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.adt_size_factors">
<span class="sig-name descname"><span class="pre">adt_size_factors</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.adt_size_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Size factors for the ADT count matrix, computed using the CLRm1 method.
Size factors are centered with <a class="reference internal" href="#scranpy.center_size_factors.center_size_factors" title="scranpy.center_size_factors.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.
If ADT data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.block">
<span class="sig-name descname"><span class="pre">block</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.block" title="Permalink to this definition">¶</a></dt>
<dd><p>Sequence containing the blocking factor for all cells (after filtering, if <code class="docutils literal notranslate"><span class="pre">filter_cells</span> <span class="pre">=</span> <span class="pre">True</span></code> in <a class="reference internal" href="#scranpy.analyze.analyze" title="scranpy.analyze.analyze"><code class="xref py py-func docutils literal notranslate"><span class="pre">analyze()</span></code></a>).
This is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> if no blocking factor was supplied.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.clusters">
<span class="sig-name descname"><span class="pre">clusters</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Array containing a cluster assignment for each cell (after filtering, if <code class="docutils literal notranslate"><span class="pre">filter_cells</span> <span class="pre">=</span> <span class="pre">True</span></code> in <a class="reference internal" href="#scranpy.analyze.analyze" title="scranpy.analyze.analyze"><code class="xref py py-func docutils literal notranslate"><span class="pre">analyze()</span></code></a>).
This may be derived from <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.graph_clusters" title="scranpy.analyze.AnalyzeResults.graph_clusters"><code class="xref py py-attr docutils literal notranslate"><span class="pre">graph_clusters</span></code></a> or <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.kmeans_clusters" title="scranpy.analyze.AnalyzeResults.kmeans_clusters"><code class="xref py py-attr docutils literal notranslate"><span class="pre">kmeans_clusters</span></code></a>, depending on the choice of <code class="docutils literal notranslate"><span class="pre">clusters_for_markers</span></code> in <a class="reference internal" href="#scranpy.analyze.analyze" title="scranpy.analyze.analyze"><code class="xref py py-func docutils literal notranslate"><span class="pre">analyze()</span></code></a>.
If no suitable clusterings are available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.column_names">
<span class="sig-name descname"><span class="pre">column_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Names</span></code></a></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.column_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Names for the cells.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if no names were supplied in <a class="reference internal" href="#scranpy.analyze.analyze" title="scranpy.analyze.analyze"><code class="xref py py-func docutils literal notranslate"><span class="pre">analyze()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.combined_pca">
<span class="sig-name descname"><span class="pre">combined_pca</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a><span class="pre">[</span><code class="docutils literal notranslate"><span class="pre">'rna_pca'</span></code><span class="pre">,</span> <code class="docutils literal notranslate"><span class="pre">'adt_pca'</span></code><span class="pre">,</span> <code class="docutils literal notranslate"><span class="pre">'crispr_pca'</span></code><span class="pre">],</span> <a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults" title="scranpy.scale_by_neighbors.ScaleByNeighborsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ScaleByNeighborsResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.combined_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>If only one modality is used for the downstream analysis, this is a string specifying the attribute containing the components to be used.
If multiple modalities are to be combined for downstream analysis, this contains the results of <a class="reference internal" href="#scranpy.scale_by_neighbors.scale_by_neighbors" title="scranpy.scale_by_neighbors.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a> on the PCs of those modalities.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.combined_qc_filter">
<span class="sig-name descname"><span class="pre">combined_qc_filter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.combined_qc_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of booleans indicating which cells are of high quality and should be retained for downstream analyses.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_filtered">
<span class="sig-name descname"><span class="pre">crispr_filtered</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_filtered" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix of CRISPR counts that has been filtered to only contain the high-quality cells in <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.combined_qc_filter" title="scranpy.analyze.AnalyzeResults.combined_qc_filter"><code class="xref py py-attr docutils literal notranslate"><span class="pre">combined_qc_filter</span></code></a>.
If CRISPR data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_markers">
<span class="sig-name descname"><span class="pre">crispr_markers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.run_pca.RunPcaResults" title="scranpy.run_pca.RunPcaResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of calling <a class="reference internal" href="#scranpy.score_markers.score_markers" title="scranpy.score_markers.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a> on the CRISPR log-expression matrix.
If CRISPR data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.
This will also be <code class="docutils literal notranslate"><span class="pre">None</span></code> if no suitable clusterings are available.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_normalized">
<span class="sig-name descname"><span class="pre">crispr_normalized</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_normalized" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix of (log-)normalized expression values derived from CRISPR counts, as computed by <a class="reference internal" href="#scranpy.normalize_counts.normalize_counts" title="scranpy.normalize_counts.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a> using <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_size_factors" title="scranpy.analyze.AnalyzeResults.crispr_size_factors"><code class="xref py py-attr docutils literal notranslate"><span class="pre">crispr_size_factors</span></code></a>.
If CRISPR data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_pca">
<span class="sig-name descname"><span class="pre">crispr_pca</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.run_pca.RunPcaResults" title="scranpy.run_pca.RunPcaResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of calling <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a> on the CRISPR log-expression matrix.
If CRISPR data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_qc_filter">
<span class="sig-name descname"><span class="pre">crispr_qc_filter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_qc_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.crispr_quality_control.filter_crispr_qc_metrics" title="scranpy.crispr_quality_control.filter_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter_crispr_qc_metrics()</span></code></a>.
If CRISPR data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_qc_metrics">
<span class="sig-name descname"><span class="pre">crispr_qc_metrics</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults" title="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.crispr_quality_control.compute_crispr_qc_metrics" title="scranpy.crispr_quality_control.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.
If CRISPR data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_qc_thresholds">
<span class="sig-name descname"><span class="pre">crispr_qc_thresholds</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults" title="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds" title="scranpy.crispr_quality_control.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.
If CRISPR data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_row_names">
<span class="sig-name descname"><span class="pre">crispr_row_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Names</span></code></a></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_row_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Names for the guides in the CRISPR data.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if CRISPR data is not available or no names were supplied to <a class="reference internal" href="#scranpy.analyze.analyze" title="scranpy.analyze.analyze"><code class="xref py py-func docutils literal notranslate"><span class="pre">analyze()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.crispr_size_factors">
<span class="sig-name descname"><span class="pre">crispr_size_factors</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.crispr_size_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Size factors for the CRISPR count matrix, derived from the sum of counts for each cell.
Size factors are centered with <a class="reference internal" href="#scranpy.center_size_factors.center_size_factors" title="scranpy.center_size_factors.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.
If CRISPR data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.graph_clusters">
<span class="sig-name descname"><span class="pre">graph_clusters</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults" title="scranpy.cluster_graph.ClusterGraphResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.graph_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if graph-based clustering was not performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.kmeans_clusters">
<span class="sig-name descname"><span class="pre">kmeans_clusters</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults" title="scranpy.cluster_graph.ClusterGraphResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.kmeans_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.cluster_kmeans.cluster_kmeans" title="scranpy.cluster_kmeans.cluster_kmeans"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if k-means clustering was not performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.mnn_corrected">
<span class="sig-name descname"><span class="pre">mnn_corrected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults" title="scranpy.correct_mnn.CorrectMnnResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">CorrectMnnResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.mnn_corrected" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.correct_mnn.correct_mnn" title="scranpy.correct_mnn.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a>.
If no blocking factor is supplied, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_filtered">
<span class="sig-name descname"><span class="pre">rna_filtered</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_filtered" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix of RNA counts that has been filtered to only contain the high-quality cells in <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.combined_qc_filter" title="scranpy.analyze.AnalyzeResults.combined_qc_filter"><code class="xref py py-attr docutils literal notranslate"><span class="pre">combined_qc_filter</span></code></a>.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_gene_variances">
<span class="sig-name descname"><span class="pre">rna_gene_variances</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults" title="scranpy.model_gene_variances.ModelGeneVariancesResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ModelGeneVariancesResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_gene_variances" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.model_gene_variances.model_gene_variances" title="scranpy.model_gene_variances.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_highly_variable_genes">
<span class="sig-name descname"><span class="pre">rna_highly_variable_genes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_highly_variable_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.choose_highly_variable_genes.choose_highly_variable_genes" title="scranpy.choose_highly_variable_genes.choose_highly_variable_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">choose_highly_variable_genes()</span></code></a>.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_markers">
<span class="sig-name descname"><span class="pre">rna_markers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.run_pca.RunPcaResults" title="scranpy.run_pca.RunPcaResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of calling <a class="reference internal" href="#scranpy.score_markers.score_markers" title="scranpy.score_markers.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a> on the RNA log-expression matrix.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.
This will also be <code class="docutils literal notranslate"><span class="pre">None</span></code> if no suitable clusterings are available.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_normalized">
<span class="sig-name descname"><span class="pre">rna_normalized</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_normalized" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix of (log-)normalized expression values derived from RNA counts, as computed by <a class="reference internal" href="#scranpy.normalize_counts.normalize_counts" title="scranpy.normalize_counts.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a> using <a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_size_factors" title="scranpy.analyze.AnalyzeResults.rna_size_factors"><code class="xref py py-attr docutils literal notranslate"><span class="pre">rna_size_factors</span></code></a>.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_pca">
<span class="sig-name descname"><span class="pre">rna_pca</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.run_pca.RunPcaResults" title="scranpy.run_pca.RunPcaResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of calling <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a> on the RNA log-expression matrix.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_qc_filter">
<span class="sig-name descname"><span class="pre">rna_qc_filter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_qc_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.rna_quality_control.filter_rna_qc_metrics" title="scranpy.rna_quality_control.filter_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter_rna_qc_metrics()</span></code></a>.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_qc_metrics">
<span class="sig-name descname"><span class="pre">rna_qc_metrics</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults" title="scranpy.rna_quality_control.ComputeRnaQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.rna_quality_control.compute_rna_qc_metrics" title="scranpy.rna_quality_control.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_qc_thresholds">
<span class="sig-name descname"><span class="pre">rna_qc_thresholds</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults" title="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_row_names">
<span class="sig-name descname"><span class="pre">rna_row_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Names</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_row_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Names for the genes in the RNA data.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if RNA data is not available or no names were supplied to <a class="reference internal" href="#scranpy.analyze.analyze" title="scranpy.analyze.analyze"><code class="xref py py-func docutils literal notranslate"><span class="pre">analyze()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.rna_size_factors">
<span class="sig-name descname"><span class="pre">rna_size_factors</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.rna_size_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Size factors for the RNA count matrix, derived from the sum of counts for each cell.
Size factors are centered with <a class="reference internal" href="#scranpy.center_size_factors.center_size_factors" title="scranpy.center_size_factors.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.
If RNA data is not available, this is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> instead.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.snn_graph">
<span class="sig-name descname"><span class="pre">snn_graph</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents" title="scranpy.build_snn_graph.GraphComponents"><code class="xref py py-class docutils literal notranslate"><span class="pre">GraphComponents</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.snn_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.build_snn_graph.build_snn_graph" title="scranpy.build_snn_graph.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if graph-based clustering was not performed.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.to_singlecellexperiment">
<span class="sig-name descname"><span class="pre">to_singlecellexperiment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">main_modality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flatten_qc_subsets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_per_block_variances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze.html#AnalyzeResults.to_singlecellexperiment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.to_singlecellexperiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the results into a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.5.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>main_modality</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'rna'</span></code>, <code class="docutils literal notranslate"><span class="pre">'adt'</span></code>, <code class="docutils literal notranslate"><span class="pre">'crispr'</span></code>]]</span>) – Modality to use as the main experiment.
If other modalities are present, they are stored in the alternative experiments.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, it defaults to RNA, then ADT, then CRISPR, depending on which modalities are available.</p></li>
<li><p><strong>flatten_qc_subsets</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to flatten QC feature subsets,
see the <a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.to_biocframe" title="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.to_biocframe"><code class="xref py py-meth docutils literal notranslate"><span class="pre">to_biocframe()</span></code></a> method of the <a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults" title="scranpy.rna_quality_control.ComputeRnaQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults</span></code></a> class for more details.</p></li>
<li><p><strong>include_per_block_variances</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the per-block variances,
see the <a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.to_biocframe" title="scranpy.model_gene_variances.ModelGeneVariancesResults.to_biocframe"><code class="xref py py-meth docutils literal notranslate"><span class="pre">to_biocframe()</span></code></a> method of the <a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults" title="scranpy.model_gene_variances.ModelGeneVariancesResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ModelGeneVariancesResults</span></code></a> class for more details.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.5.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> containing the filtered and normalized matrices in the assays.
QC metrics, size factors and clustering results are stored in the column data.
PCA and other low-dimensional embeddings are stored in the reduced dimensions.
Additional modalities are stored as alternative experiments.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.tsne">
<span class="sig-name descname"><span class="pre">tsne</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.tsne" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.run_tsne.run_tsne" title="scranpy.run_tsne.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if t-SNE was not performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.analyze.AnalyzeResults.umap">
<span class="sig-name descname"><span class="pre">umap</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.analyze.AnalyzeResults.umap" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.run_umap.run_umap" title="scranpy.run_umap.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if UMAP was not performed.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze.analyze">
<span class="sig-prename descclassname"><span class="pre">scranpy.analyze.</span></span><span class="sig-name descname"><span class="pre">analyze</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">rna_x,</span> <span class="pre">adt_x=None,</span> <span class="pre">crispr_x=None,</span> <span class="pre">block=None,</span> <span class="pre">rna_subsets=[],</span> <span class="pre">adt_subsets=[],</span> <span class="pre">suggest_rna_qc_thresholds_options={},</span> <span class="pre">suggest_adt_qc_thresholds_options={},</span> <span class="pre">suggest_crispr_qc_thresholds_options={},</span> <span class="pre">filter_cells=True,</span> <span class="pre">center_size_factors_options={},</span> <span class="pre">compute_clrm1_factors_options={},</span> <span class="pre">normalize_counts_options={},</span> <span class="pre">model_gene_variances_options={},</span> <span class="pre">choose_highly_variable_genes_options={},</span> <span class="pre">run_pca_options={},</span> <span class="pre">use_rna_pcs=True,</span> <span class="pre">use_adt_pcs=True,</span> <span class="pre">use_crispr_pcs=True,</span> <span class="pre">scale_by_neighbors_options={},</span> <span class="pre">correct_mnn_options={},</span> <span class="pre">run_umap_options={},</span> <span class="pre">run_tsne_options={},</span> <span class="pre">build_snn_graph_options={},</span> <span class="pre">cluster_graph_options={},</span> <span class="pre">run_all_neighbor_steps_options={},</span> <span class="pre">kmeans_clusters=None,</span> <span class="pre">cluster_kmeans_options={},</span> <span class="pre">clusters_for_markers=['graph',</span> <span class="pre">'kmeans'],</span> <span class="pre">score_markers_options={},</span> <span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;,</span> <span class="pre">rna_assay=0,</span> <span class="pre">adt_assay=0,</span> <span class="pre">crispr_assay=0,</span> <span class="pre">num_threads=3</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/analyze.html#analyze"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze.analyze" title="Permalink to this definition">¶</a></dt>
<dd><p>Run through a simple single-cell analysis pipeline, starting from a count matrix and ending with clusters, visualizations and markers.
This also supports integration of multiple modalities and correction of batch effects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rna_x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>]</span>) – <p>A matrix-like object containing RNA counts.
This should have the same number of columns as the other <code class="docutils literal notranslate"><span class="pre">*_x</span></code> arguments.</p>
<p>Alternatively, a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> class containing such a matrix in its <code class="docutils literal notranslate"><span class="pre">rna_assay</span></code>.</p>
<p>Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if no RNA counts are available.</p>
</p></li>
<li><p><strong>adt_x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>]</span>) – <p>A matrix-like object containing ADT counts.
This should have the same number of columns as the other <code class="docutils literal notranslate"><span class="pre">*_x</span></code> arguments.</p>
<p>Alternatively, a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> class containing such a matrix in its <code class="docutils literal notranslate"><span class="pre">adt_assay</span></code>.</p>
<p>Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if no ADT counts are available.</p>
</p></li>
<li><p><strong>crispr_x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>]</span>) – <p>A matrix-like object containing CRISPR counts.
This should have the same number of columns as the other <code class="docutils literal notranslate"><span class="pre">*_x</span></code> arguments.</p>
<p>Alternatively, a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> class containing such a matrix in its <code class="docutils literal notranslate"><span class="pre">crispr_assay</span></code>.</p>
<p>Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if no CRISPR counts are available.</p>
</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Factor specifying the block of origin (e.g., batch, sample) for each cell in the <code class="docutils literal notranslate"><span class="pre">*_x</span></code> matrices.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>rna_subsets</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>, <a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Gene subsets for quality control, typically used for mitochondrial genes.
Check out the <code class="docutils literal notranslate"><span class="pre">subsets</span></code> arguments in <a class="reference internal" href="#scranpy.rna_quality_control.compute_rna_qc_metrics" title="scranpy.rna_quality_control.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a> for details.</p></li>
<li><p><strong>adt_subsets</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>, <a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – ADT subsets for quality control, typically used for IgG controls.
Check out the <code class="docutils literal notranslate"><span class="pre">subsets</span></code> arguments in <a class="reference internal" href="#scranpy.adt_quality_control.compute_adt_qc_metrics" title="scranpy.adt_quality_control.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a> for details.</p></li>
<li><p><strong>suggest_rna_qc_thresholds_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>suggest_adt_qc_thresholds_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds" title="scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>suggest_crispr_qc_thresholds_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds" title="scranpy.crispr_quality_control.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>filter_cells</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to filter the count matrices to only retain high-quality cells in all modalities.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, QC metrics and thresholds are still computed but are not used to filter the count matrices.</p></li>
<li><p><strong>center_size_factors_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.center_size_factors.center_size_factors" title="scranpy.center_size_factors.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.</p></li>
<li><p><strong>compute_clrm1_factors_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.compute_clrm1_factors.compute_clrm1_factors" title="scranpy.compute_clrm1_factors.compute_clrm1_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_clrm1_factors()</span></code></a>.</p></li>
<li><p><strong>normalized_counts_options</strong> – Arguments to pass to <a class="reference internal" href="#scranpy.normalize_counts.normalize_counts" title="scranpy.normalize_counts.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>model_gene_variances_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.model_gene_variances.model_gene_variances" title="scranpy.model_gene_variances.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>.</p></li>
<li><p><strong>choose_highly_variable_genes_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.choose_highly_variable_genes.choose_highly_variable_genes" title="scranpy.choose_highly_variable_genes.choose_highly_variable_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">choose_highly_variable_genes()</span></code></a>.</p></li>
<li><p><strong>run_pca_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p></li>
<li><p><strong>use_rna_pcs</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to use the RNA-derived PCs for downstream steps (i.e., clustering, visualization).</p></li>
<li><p><strong>use_adt_pcs</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to use the ADT-derived PCs for downstream steps (i.e., clustering, visualization).</p></li>
<li><p><strong>use_crispr_pcs</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to use the CRISPR-derived PCs for downstream steps (i.e., clustering, visualization).</p></li>
<li><p><strong>scale_by_neighbors_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.scale_by_neighbors.scale_by_neighbors" title="scranpy.scale_by_neighbors.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a>.
Only used if multiple modalities are available and their corresponding <code class="docutils literal notranslate"><span class="pre">use_*</span></code> arguments are <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
<li><p><strong>correct_mnn_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.correct_mnn.correct_mnn" title="scranpy.correct_mnn.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is supplied.</p></li>
<li><p><strong>run_umap_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Arguments to pass to <a class="reference internal" href="#scranpy.run_umap.run_umap" title="scranpy.run_umap.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, UMAP is not performed.</p></li>
<li><p><strong>run_tsne_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Arguments to pass to <a class="reference internal" href="#scranpy.run_tsne.run_tsne" title="scranpy.run_tsne.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, t-SNE is not performed.</p></li>
<li><p><strong>build_snn_graph_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Arguments to pass to <a class="reference internal" href="#scranpy.build_snn_graph.build_snn_graph" title="scranpy.build_snn_graph.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.
Ignored if <code class="docutils literal notranslate"><span class="pre">cluster_graph_options</span> <span class="pre">=</span> <span class="pre">None</span></code>.</p></li>
<li><p><strong>cluster_graph_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, graph-based clustering is not performed.</p></li>
<li><p><strong>run_all_neighbor_steps_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.run_all_neighbor_steps.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a>.</p></li>
<li><p><strong>kmeans_clusters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of clusters to use in k-means clustering.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, k-means clustering is not performed.</p></li>
<li><p><strong>cluster_kmeans_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.cluster_kmeans.cluster_kmeans" title="scranpy.cluster_kmeans.cluster_kmeans"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a>.
Ignored if <code class="docutils literal notranslate"><span class="pre">kmeans_clusters</span> <span class="pre">=</span> <span class="pre">None</span></code>.</p></li>
<li><p><strong>clusters_for_markers</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a></span>) – List of clustering algorithms (either <code class="docutils literal notranslate"><span class="pre">graph</span></code> or <code class="docutils literal notranslate"><span class="pre">kmeans</span></code>), specifying the clustering to be used for marker detection.
The first available clustering will be chosen.</p></li>
<li><p><strong>score_markers_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Arguments to pass to <a class="reference internal" href="#scranpy.score_markers.score_markers" title="scranpy.score_markers.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.
Ignored if no suitable clusterings are available.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Algorithm to use for nearest-neighbor searches in the various steps.</p></li>
<li><p><strong>rna_assay</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – </p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use in each step.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.analyze.AnalyzeResults" title="scranpy.analyze.AnalyzeResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The results of the entire analysis, including the results from each step.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>C++ libraries in the <a class="reference external" href="https://github.com/libscran">libscran</a> GitHub organization, which implement all of these steps.</p>
</dd></dl>

</section>
<section id="module-scranpy.build_snn_graph">
<span id="scranpy-build-snn-graph-module"></span><h2>scranpy.build_snn_graph module<a class="headerlink" href="#module-scranpy.build_snn_graph" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.build_snn_graph.</span></span><span class="sig-name descname"><span class="pre">GraphComponents</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vertices</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/build_snn_graph.html#GraphComponents"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Components of a (possibly weighted) graph.
Typically, nodes are cells and edges are formed between cells with similar expression profiles.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'edges':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'vertices':</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">'weights':</span> <span class="pre">typing.Optional[numpy.ndarray]}</span></em><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'edges':</span> <span class="pre">Field(name='edges',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'vertices':</span> <span class="pre">Field(name='vertices',type=&lt;class</span> <span class="pre">'int'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'weights':</span> <span class="pre">Field(name='weights',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('vertices',</span> <span class="pre">'edges',</span> <span class="pre">'weights')</span></em><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.as_igraph">
<span class="sig-name descname"><span class="pre">as_igraph</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/build_snn_graph.html#GraphComponents.as_igraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.as_igraph" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert to a <code class="docutils literal notranslate"><span class="pre">Graph</span></code> from the <strong>igraph</strong> package.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="docutils literal notranslate"><span class="pre">Graph</span></code> for use with methods in the <strong>igraph</strong> package.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.edges">
<span class="sig-name descname"><span class="pre">edges</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of indices for graph edges.
Pairs of values represent the endpoints of an (undirected) edge, i.e., <code class="docutils literal notranslate"><span class="pre">edges[0:2]</span></code> form the first edge, <code class="docutils literal notranslate"><span class="pre">edges[2:4]</span></code> form the second edge and so on.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.vertices">
<span class="sig-name descname"><span class="pre">vertices</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></em><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.vertices" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of vertices in the graph (i.e., cells).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.GraphComponents.weights">
<span class="sig-name descname"><span class="pre">weights</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.build_snn_graph.GraphComponents.weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of weights for each edge. This has length equal to half the length of <code class="docutils literal notranslate"><span class="pre">edges</span></code>; the first weight corresponds to the first edge, and so on.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the graph is assumed to be unweighted.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.build_snn_graph.build_snn_graph">
<span class="sig-prename descclassname"><span class="pre">scranpy.build_snn_graph.</span></span><span class="sig-name descname"><span class="pre">build_snn_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_scheme='ranked'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/build_snn_graph.html#build_snn_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.build_snn_graph.build_snn_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Build a shared nearest neighbor (SNN) graph where each node is a cell.
Edges are formed between cells that share one or more nearest neighbors, weighted by the number or importance of those shared neighbors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>]</span>) – <p>Numeric matrix where rows are dimensions and columns are cells, typically containing a low-dimensional representation from, e.g., <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object containing existing neighbor search results.
The number of neighbors should be the same as <code class="docutils literal notranslate"><span class="pre">num_neighbors</span></code>, otherwise a warning is raised.</p>
<p>Alternatively, a <code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code> object.</p>
</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors in the nearest-neighbor graph.
Larger values generally result in broader clusters during community detection.</p></li>
<li><p><strong>weight_scheme</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'ranked'</span></code>, <code class="docutils literal notranslate"><span class="pre">'number'</span></code>, <code class="docutils literal notranslate"><span class="pre">'jaccard'</span></code>]</span>) – Weighting scheme to use for the edges of the SNN graph, based on the number or ranking of the shared nearest neighbors.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – The algorithm to use for the nearest-neighbor search.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> is not a pre-built nearest-neighbor search index or a list of existing nearest-neighbor search results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents" title="scranpy.build_snn_graph.GraphComponents"><code class="xref py py-class docutils literal notranslate"><span class="pre">GraphComponents</span></code></a></span></p>
</dd>
</dl>
<dl class="simple">
<dt>Results:</dt><dd><p>The components of the SNN graph, to be used in community detection.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">build_snn_graph</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_graph_cluster">scran_graph_cluster</a> C++ library, which provides some more details on the weighting.</p>
</dd></dl>

</section>
<section id="module-scranpy.center_size_factors">
<span id="scranpy-center-size-factors-module"></span><h2>scranpy.center_size_factors module<a class="headerlink" href="#module-scranpy.center_size_factors" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.center_size_factors.center_size_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.center_size_factors.</span></span><span class="sig-name descname"><span class="pre">center_size_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lowest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/center_size_factors.html#center_size_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.center_size_factors.center_size_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Center size factors before computing normalized values from the count matrix.
This ensures that the normalized values are on the same scale as the original counts for easier interpretation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Floating-point array containing size factors for all cells.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell.
If provided, this should have length equal to the number of cells, where cells have the same value if and only if they are in the same block.
Defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>, where all cells are treated as being part of the same block.</p></li>
<li><p><strong>mode</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'lowest'</span></code>, <code class="docutils literal notranslate"><span class="pre">'per-block'</span></code>]</span>) – How to scale size factors across blocks.
<code class="docutils literal notranslate"><span class="pre">lowest</span></code> will scale all size factors by the lowest per-block average.
<code class="docutils literal notranslate"><span class="pre">per-block</span></code> will center the size factors in each block separately.
This argument is only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> in place.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, a new array is returned.
This argument only used if <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> is double-precision, otherwise a new array is always returned.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array containing centered size factors.
If <code class="docutils literal notranslate"><span class="pre">in_place</span> <span class="pre">=</span> <span class="pre">True</span></code>, this is a reference to <code class="docutils literal notranslate"><span class="pre">size_factors</span></code>.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">center_size_factors</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_norm">scran_norm</a> C++ library, which describes the rationale behind centering.</p>
</dd></dl>

</section>
<section id="module-scranpy.choose_highly_variable_genes">
<span id="scranpy-choose-highly-variable-genes-module"></span><h2>scranpy.choose_highly_variable_genes module<a class="headerlink" href="#module-scranpy.choose_highly_variable_genes" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.choose_highly_variable_genes.choose_highly_variable_genes">
<span class="sig-prename descclassname"><span class="pre">scranpy.choose_highly_variable_genes.</span></span><span class="sig-name descname"><span class="pre">choose_highly_variable_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stats</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">larger</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_ties</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/choose_highly_variable_genes.html#choose_highly_variable_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.choose_highly_variable_genes.choose_highly_variable_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Choose highly variable genes (HVGs), typically based on a variance-related statistic.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stats</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Array of variances (or a related statistic) across all genes.
Typically the residuals from <a class="reference internal" href="#scranpy.model_gene_variances.model_gene_variances" title="scranpy.model_gene_variances.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a> used here.</p></li>
<li><p><strong>top</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of top genes to retain.
Note that the actual number of retained genes may not be equal to <code class="docutils literal notranslate"><span class="pre">top</span></code>, depending on the other options.</p></li>
<li><p><strong>larger</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether larger values of <code class="docutils literal notranslate"><span class="pre">stats</span></code> represent more variable genes.
If true, HVGs are defined from the largest values of <code class="docutils literal notranslate"><span class="pre">stats</span></code>.</p></li>
<li><p><strong>keep_ties</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to keep ties at the <code class="docutils literal notranslate"><span class="pre">top</span></code>-th most variable gene.
This avoids arbitrary breaking of tied values.</p></li>
<li><p><strong>bound</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]</span>) – The lower bound (if <code class="docutils literal notranslate"><span class="pre">larger</span> <span class="pre">=</span> <span class="pre">True</span></code>) or upper bound (otherwise) to be applied to <code class="docutils literal notranslate"><span class="pre">stats</span></code>.
Genes are not considered to be HVGs if they do not pass this bound, even if they are within the <code class="docutils literal notranslate"><span class="pre">top</span></code> genes.
Ignored if <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array containing the indices of genes in <code class="docutils literal notranslate"><span class="pre">stats</span></code> that are considered to be highly variable.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">choose_highly_variable_genes</span></code> function from the <a class="reference external" href="https://github.com/libscran/scran_variances">scran_variances</a> library,
which provides the underlying implementation.</p>
</dd></dl>

</section>
<section id="module-scranpy.choose_pseudo_count">
<span id="scranpy-choose-pseudo-count-module"></span><h2>scranpy.choose_pseudo_count module<a class="headerlink" href="#module-scranpy.choose_pseudo_count" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.choose_pseudo_count.choose_pseudo_count">
<span class="sig-prename descclassname"><span class="pre">scranpy.choose_pseudo_count.</span></span><span class="sig-name descname"><span class="pre">choose_pseudo_count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_bias</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/choose_pseudo_count.html#choose_pseudo_count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.choose_pseudo_count.choose_pseudo_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Choose a suitable pseudo-count to control the bias introduced by log-transformation of normalized counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Floating-point array of size factors for all cells.</p></li>
<li><p><strong>quantile</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Quantile to use for defining extreme size factors.</p></li>
<li><p><strong>max_bias</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Maximum allowed bias in the log-fold changes between cells.</p></li>
<li><p><strong>min_value</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum value for the pseudo-count.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Choice of pseudo-count, for use in <a class="reference internal" href="#scranpy.normalize_counts.normalize_counts" title="scranpy.normalize_counts.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">choose_pseudo_count</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_norm">scran_norm</a> C++ library, which describes the rationale behind the choice of pseudo-count.</p>
</dd></dl>

</section>
<section id="module-scranpy.cluster_graph">
<span id="scranpy-cluster-graph-module"></span><h2>scranpy.cluster_graph module<a class="headerlink" href="#module-scranpy.cluster_graph" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.cluster_graph.</span></span><span class="sig-name descname"><span class="pre">ClusterGraphLeidenResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">status</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">membership</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/cluster_graph.html#ClusterGraphLeidenResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults" title="scranpy.cluster_graph.ClusterGraphResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphResults</span></code></a></p>
<p>Clustering results from <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a> when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;leiden&quot;</span></code>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'quality':</span> <span class="pre">&lt;class</span> <span class="pre">'float'&gt;}</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'membership':</span> <span class="pre">Field(name='membership',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'quality':</span> <span class="pre">Field(name='quality',type=&lt;class</span> <span class="pre">'float'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'status':</span> <span class="pre">Field(name='status',type=&lt;class</span> <span class="pre">'int'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('status',</span> <span class="pre">'membership',</span> <span class="pre">'quality')</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphLeidenResults.quality">
<span class="sig-name descname"><span class="pre">quality</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.quality" title="Permalink to this definition">¶</a></dt>
<dd><p>Quality of the clustering.
This is the same as the modularity if <code class="docutils literal notranslate"><span class="pre">leiden_objective</span> <span class="pre">=</span> <span class="pre">&quot;modularity&quot;</span></code>.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.cluster_graph.</span></span><span class="sig-name descname"><span class="pre">ClusterGraphMultilevelResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">status</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">membership</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">levels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modularity</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/cluster_graph.html#ClusterGraphMultilevelResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults" title="scranpy.cluster_graph.ClusterGraphResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphResults</span></code></a></p>
<p>Clustering results from <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a> when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;multilevel&quot;</span></code>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'levels':</span> <span class="pre">tuple[numpy.ndarray],</span> <span class="pre">'modularity':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'levels':</span> <span class="pre">Field(name='levels',type=tuple[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'membership':</span> <span class="pre">Field(name='membership',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'modularity':</span> <span class="pre">Field(name='modularity',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'status':</span> <span class="pre">Field(name='status',type=&lt;class</span> <span class="pre">'int'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('status',</span> <span class="pre">'membership',</span> <span class="pre">'levels',</span> <span class="pre">'modularity')</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.levels">
<span class="sig-name descname"><span class="pre">levels</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.levels" title="Permalink to this definition">¶</a></dt>
<dd><p>Clustering at each level of the algorithm.
Each array corresponds to one level and contains the cluster assignment for each cell at that level.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphMultilevelResults.modularity">
<span class="sig-name descname"><span class="pre">modularity</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.modularity" title="Permalink to this definition">¶</a></dt>
<dd><p>Modularity at each level.
This has length equal to <a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.levels" title="scranpy.cluster_graph.ClusterGraphMultilevelResults.levels"><code class="xref py py-attr docutils literal notranslate"><span class="pre">levels</span></code></a>, and the largest value corresponds to the assignments reported in <a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.membership" title="scranpy.cluster_graph.ClusterGraphResults.membership"><code class="xref py py-attr docutils literal notranslate"><span class="pre">membership</span></code></a>.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.cluster_graph.</span></span><span class="sig-name descname"><span class="pre">ClusterGraphResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">status</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">membership</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/cluster_graph.html#ClusterGraphResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Clustering results from <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'membership':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'status':</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;}</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'membership':</span> <span class="pre">Field(name='membership',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'status':</span> <span class="pre">Field(name='status',type=&lt;class</span> <span class="pre">'int'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('status',</span> <span class="pre">'membership')</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.membership">
<span class="sig-name descname"><span class="pre">membership</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.membership" title="Permalink to this definition">¶</a></dt>
<dd><p>Array containing the cluster assignment for each vertex, i.e., cell.
All values are in [0, N) where N is the total number of clusters.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphResults.status">
<span class="sig-name descname"><span class="pre">status</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphResults.status" title="Permalink to this definition">¶</a></dt>
<dd><p>Status of the clustering.
A non-zero value indicates that the algorithm did not complete successfully.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.cluster_graph.</span></span><span class="sig-name descname"><span class="pre">ClusterGraphWalktrapResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">status</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">membership</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modularity</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/cluster_graph.html#ClusterGraphWalktrapResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults" title="scranpy.cluster_graph.ClusterGraphResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphResults</span></code></a></p>
<p>Clustering results from <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a> when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;walktrap&quot;</span></code>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'merges':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'modularity':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'membership':</span> <span class="pre">Field(name='membership',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'merges':</span> <span class="pre">Field(name='merges',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'modularity':</span> <span class="pre">Field(name='modularity',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'status':</span> <span class="pre">Field(name='status',type=&lt;class</span> <span class="pre">'int'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('status',</span> <span class="pre">'membership',</span> <span class="pre">'merges',</span> <span class="pre">'modularity')</span></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.merges">
<span class="sig-name descname"><span class="pre">merges</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.merges" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix containing two columns.
Each row specifies the pair of cells or clusters that were merged at each step of the algorithm.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_graph.ClusterGraphWalktrapResults.modularity">
<span class="sig-name descname"><span class="pre">modularity</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.modularity" title="Permalink to this definition">¶</a></dt>
<dd><p>Array of modularity scores at each merge step.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.cluster_graph.cluster_graph">
<span class="sig-prename descclassname"><span class="pre">scranpy.cluster_graph.</span></span><span class="sig-name descname"><span class="pre">cluster_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'multilevel'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multilevel_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_objective</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'modularity'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">walktrap_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/cluster_graph.html#cluster_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_graph.cluster_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Identify clusters of cells using a variety of community detection methods from a graph where similar cells are connected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents" title="scranpy.build_snn_graph.GraphComponents"><code class="xref py py-class docutils literal notranslate"><span class="pre">GraphComponents</span></code></a></span>) – Components of the graph to be clustered, typically produced by <code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code>.</p></li>
<li><p><strong>method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'multilevel'</span></code>, <code class="docutils literal notranslate"><span class="pre">'leiden'</span></code>, <code class="docutils literal notranslate"><span class="pre">'walktrap'</span></code>]</span>) – Community detection algorithm to use.</p></li>
<li><p><strong>multilevel_resolution</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Resolution of the clustering when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;multilevel&quot;</span></code>.
Larger values result in finer clusters.</p></li>
<li><p><strong>leiden_resolution</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Resolution of the clustering when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;leiden&quot;</span></code>.
Larger values result in finer clusters.</p></li>
<li><p><strong>leiden_objective</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'modularity'</span></code>, <code class="docutils literal notranslate"><span class="pre">'cpm'</span></code>]</span>) – Objective function to use when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;leiden&quot;</span></code>.</p></li>
<li><p><strong>walktrap_steps</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of steps to use when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;walktrap&quot;</span></code>.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Random seed to use for <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;multilevel&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;leiden&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults" title="scranpy.cluster_graph.ClusterGraphMultilevelResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults</span></code></a>, if <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;multilevel&quot;</span></code>.</p></li>
<li><p><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults" title="scranpy.cluster_graph.ClusterGraphLeidenResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults</span></code></a>, if <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;leiden&quot;</span></code>.</p></li>
<li><p><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults" title="scranpy.cluster_graph.ClusterGraphWalktrapResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults</span></code></a>, if <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;walktrap&quot;</span></code>.</p></li>
</ul>
<p>All objects contain at least <code class="docutils literal notranslate"><span class="pre">status</span></code>, an indicator of whether the
algorithm successfully completed; and <code class="docutils literal notranslate"><span class="pre">membership</span></code>, an array of
cluster assignments for each node in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Clustering results, as a</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The various <code class="docutils literal notranslate"><span class="pre">cluster_*</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_graph_cluster">scran_graph_cluster</a> C++ library, which provides some more details on each algorithm.</p>
</dd></dl>

</section>
<section id="module-scranpy.cluster_kmeans">
<span id="scranpy-cluster-kmeans-module"></span><h2>scranpy.cluster_kmeans module<a class="headerlink" href="#module-scranpy.cluster_kmeans" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.cluster_kmeans.</span></span><span class="sig-name descname"><span class="pre">ClusterKmeansResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">status</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/cluster_kmeans.html#ClusterKmeansResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.cluster_kmeans.cluster_kmeans" title="scranpy.cluster_kmeans.cluster_kmeans"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'centers':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'clusters':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'iterations':</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">'status':</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;}</span></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'centers':</span> <span class="pre">Field(name='centers',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'clusters':</span> <span class="pre">Field(name='clusters',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'iterations':</span> <span class="pre">Field(name='iterations',type=&lt;class</span> <span class="pre">'int'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'status':</span> <span class="pre">Field(name='status',type=&lt;class</span> <span class="pre">'int'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('clusters',</span> <span class="pre">'centers',</span> <span class="pre">'iterations',</span> <span class="pre">'status')</span></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.centers">
<span class="sig-name descname"><span class="pre">centers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.centers" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix contaiing the coordinates of the cluster centroids.
Dimensions are in the rows while centers are in the columns.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.clusters">
<span class="sig-name descname"><span class="pre">clusters</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Array containing the cluster assignment for each cell.
Values are integers in [0, N) where N is the total number of clusters.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.iterations">
<span class="sig-name descname"><span class="pre">iterations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.iterations" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of refinement iterations that were performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.ClusterKmeansResults.status">
<span class="sig-name descname"><span class="pre">status</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></em><a class="headerlink" href="#scranpy.cluster_kmeans.ClusterKmeansResults.status" title="Permalink to this definition">¶</a></dt>
<dd><p>Convergence status.
Any non-zero value indicates a convergence failure though the exact meaning depends on the choice of <code class="docutils literal notranslate"><span class="pre">refine_method</span></code> in <a class="reference internal" href="#scranpy.cluster_kmeans.cluster_kmeans" title="scranpy.cluster_kmeans.cluster_kmeans"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a>.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans.cluster_kmeans">
<span class="sig-prename descclassname"><span class="pre">scranpy.cluster_kmeans.</span></span><span class="sig-name descname"><span class="pre">cluster_kmeans</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'var-part'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refine_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hartigan-wong'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_part_optimize_partition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_part_size_adjustment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lloyd_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hartigan_wong_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hartigan_wong_quick_transfer_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hartigan_wong_quit_quick_transfer_failure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5489</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/cluster_kmeans.html#cluster_kmeans"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_kmeans.cluster_kmeans" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform k-means clustering with a variety of different initialization and refinement algorithms.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Input data matrix where rows are dimensions and columns are observations (i.e., cells).</p></li>
<li><p><strong>k</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of clusters.</p></li>
<li><p><strong>init_method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'var-part'</span></code>, <code class="docutils literal notranslate"><span class="pre">'kmeans++'</span></code>, <code class="docutils literal notranslate"><span class="pre">'random'</span></code>]</span>) – Initialization method for defining the initial centroid coordinates.
Choices are variance partitioning (<code class="docutils literal notranslate"><span class="pre">var-part</span></code>), kmeans++ (<code class="docutils literal notranslate"><span class="pre">kmeans++</span></code>) or random initialization (<code class="docutils literal notranslate"><span class="pre">random</span></code>).</p></li>
<li><p><strong>refine_method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'hartigan-wong'</span></code>, <code class="docutils literal notranslate"><span class="pre">'lloyd'</span></code>]</span>) – Method to use to refine the cluster assignments and centroid coordinates.
Choices are Lloyd’s algorithm (<code class="docutils literal notranslate"><span class="pre">lloyd</span></code>) or the Hartigan-Wong algorithm (<code class="docutils literal notranslate"><span class="pre">hartigan-wong</span></code>).</p></li>
<li><p><strong>var_part_optimize_partition</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether each partition boundary should be optimized to reduce the sum of squares in the child partitions.
Only used if <code class="docutils literal notranslate"><span class="pre">init_method</span> <span class="pre">=</span> <span class="pre">&quot;var-part&quot;</span></code>.</p></li>
<li><p><strong>var_part_size_adjustment</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Floating-point value between 0 and 1, specifying the adjustment to the cluster size when prioritizing the next cluster to partition.
Setting this to 0 will ignore the cluster size while setting this to 1 will generally favor larger clusters.
Only used if <code class="docutils literal notranslate"><span class="pre">init_method</span> <span class="pre">=</span> <span class="pre">&quot;var-part&quot;</span></code>.</p></li>
<li><p><strong>lloyd_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximmum number of iterations for the Lloyd algorithm.</p></li>
<li><p><strong>hartigan_wong_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximmum number of iterations for the Hartigan-Wong algorithm.</p></li>
<li><p><strong>hartigan_wong_quick_transfer_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximmum number of quick transfer iterations for the Hartigan-Wong algorithm.</p></li>
<li><p><strong>hartigan_wong_quit_quick_transfer_failure</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to quit the Hartigan-Wong algorithm upon convergence failure during quick transfer iterations.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Seed to use for random or kmeans++ initialization.</p></li>
<li><p><strong>num.threads</strong> – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults" title="scranpy.cluster_kmeans.ClusterKmeansResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterKmeansResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Results of k-means clustering on the observations.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/LTLA/CppKmeans">https://github.com/LTLA/CppKmeans</a>, which describes the various initialization and refinement algorithms in more detail.</p>
</dd></dl>

</section>
<section id="module-scranpy.combine_factors">
<span id="scranpy-combine-factors-module"></span><h2>scranpy.combine_factors module<a class="headerlink" href="#module-scranpy.combine_factors" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.combine_factors.combine_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.combine_factors.</span></span><span class="sig-name descname"><span class="pre">combine_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_unused</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/combine_factors.html#combine_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.combine_factors.combine_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Combine multiple categorical factors based on the unique combinations of levels from each factor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence containing factors of interest.
Each entry corresponds to a factor and should be a sequence of the same length.
Corresponding elements across all factors represent the combination of levels for a single observation.</p></li>
<li><p><strong>keep_unused</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to report unused combinations of levels.
If any entry of <code class="docutils literal notranslate"><span class="pre">factors</span></code> is a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Factor.Factor" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Factor</span></code></a> object, any unused levels will also be preserved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>Sorted and unique combinations of levels as a tuple.
Each entry of the tuple is a list that corresponds to a factor in <code class="docutils literal notranslate"><span class="pre">factors</span></code>.
Corresponding elements of each list define a single combination, i.e., the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th combination is defined by taking the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th element of each sequence in the tuple.</p></li>
<li><p>Integer array of length equal to each sequence of <code class="docutils literal notranslate"><span class="pre">factors</span></code>, specifying the combination for each observation.
Each entry is an index <code class="docutils literal notranslate"><span class="pre">i</span></code> into the sequences in the previous tuple.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple containing</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">combine_factors</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_aggregate">scran_aggregate</a> library, which provides the underlying implementation.</p>
</dd></dl>

</section>
<section id="module-scranpy.compute_clrm1_factors">
<span id="scranpy-compute-clrm1-factors-module"></span><h2>scranpy.compute_clrm1_factors module<a class="headerlink" href="#module-scranpy.compute_clrm1_factors" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.compute_clrm1_factors.compute_clrm1_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.compute_clrm1_factors.</span></span><span class="sig-name descname"><span class="pre">compute_clrm1_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/compute_clrm1_factors.html#compute_clrm1_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.compute_clrm1_factors.compute_clrm1_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute size factors from an ADT count matrix using the CLRm1 method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object containing ADT count data.
Rows correspond to tags and columns correspond to cells.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array containing the CLRm1 size factor for each cell.
Note that these size factors are not centered and should be passed through, e.g., <a class="reference internal" href="#scranpy.center_size_factors.center_size_factors" title="scranpy.center_size_factors.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a> before normalization.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/clrm1">https://github.com/libscran/clrm1</a>, for a description of the CLRm1 method.</p>
</dd></dl>

</section>
<section id="module-scranpy.correct_mnn">
<span id="scranpy-correct-mnn-module"></span><h2>scranpy.correct_mnn module<a class="headerlink" href="#module-scranpy.correct_mnn" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.correct_mnn.</span></span><span class="sig-name descname"><span class="pre">CorrectMnnResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">corrected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merge_order</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_pairs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/correct_mnn.html#CorrectMnnResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.correct_mnn.correct_mnn" title="scranpy.correct_mnn.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'corrected':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'merge_order':</span> <span class="pre">list[str],</span> <span class="pre">'num_pairs':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'corrected':</span> <span class="pre">Field(name='corrected',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'merge_order':</span> <span class="pre">Field(name='merge_order',type=list[str],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'num_pairs':</span> <span class="pre">Field(name='num_pairs',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('corrected',</span> <span class="pre">'merge_order',</span> <span class="pre">'num_pairs')</span></em><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.corrected">
<span class="sig-name descname"><span class="pre">corrected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.corrected" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point matrix of the same dimensions as the <code class="docutils literal notranslate"><span class="pre">x</span></code> used in <a class="reference internal" href="#scranpy.correct_mnn.correct_mnn" title="scranpy.correct_mnn.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a>, containing the corrected values.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.merge_order">
<span class="sig-name descname"><span class="pre">merge_order</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.merge_order" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge order for the levels of the blocking factor.
The first level in this vector is used as the reference batch; all other batches are iteratively merged and added to the reference.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.correct_mnn.CorrectMnnResults.num_pairs">
<span class="sig-name descname"><span class="pre">num_pairs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.correct_mnn.CorrectMnnResults.num_pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Integer vector of length equal to the number of batches minus 1.
This contains the number of MNN pairs at each merge step.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.correct_mnn.correct_mnn">
<span class="sig-prename descclassname"><span class="pre">scranpy.correct_mnn.</span></span><span class="sig-name descname"><span class="pre">correct_mnn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mads=3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">robust_iterations=2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">robust_trim=0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mass_cap=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_policy='max-rss'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/correct_mnn.html#correct_mnn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.correct_mnn.correct_mnn" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply mutual nearest neighbor (MNN) correction to remove batch effects from a low-dimensional matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Matrix of coordinates where rows are dimensions and columns are cells, typically generated by <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Factor specifying the block of origin (e.g., batch, sample) for each cell. Length should equal the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to use when identifying MNN pairs.</p></li>
<li><p><strong>num_mads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of median absolute deviations to use for removing outliers in the center-of-mass calculations.</p></li>
<li><p><strong>robust_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of iterations for robust calculation of the center of mass.</p></li>
<li><p><strong>robust_trim</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Trimming proportion for robust calculation of the center of mass.
This should be a value in [0, 1).</p></li>
<li><p><strong>mass_cap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Cap on the number of observations to use for center-of-mass calculations on the reference dataset.
A value of 100,000 may be appropriate for speeding up correction of very large datasets.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no cap is used.</p></li>
<li><p><strong>order</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence containing the unique levels of <code class="docutils literal notranslate"><span class="pre">block</span></code> in the desired merge order.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, a suitable merge order is automatically determined.</p></li>
<li><p><strong>reference_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'max-rss'</span></code>, <code class="docutils literal notranslate"><span class="pre">'max-size'</span></code>, <code class="docutils literal notranslate"><span class="pre">'max-variance'</span></code>, <code class="docutils literal notranslate"><span class="pre">'input'</span></code>]</span>) – Policy to use to choose the first reference batch.
This can be based on the largest batch (<code class="docutils literal notranslate"><span class="pre">max-size</span></code>), the most variable batch (<code class="docutils literal notranslate"><span class="pre">max-variance</span></code>), the batch with the largest residual sum of squares (<code class="docutils literal notranslate"><span class="pre">max-rss</span></code>), or the first specified input (<code class="docutils literal notranslate"><span class="pre">input</span></code>).
Only used for automatic merges, i.e., when <code class="docutils literal notranslate"><span class="pre">order</span> <span class="pre">=</span> <span class="pre">None</span></code>.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – The nearest-neighbor algorithm to use.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults" title="scranpy.correct_mnn.CorrectMnnResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">CorrectMnnResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The results of the MNN correction, including a matrix of the corrected coordinates and some additional diagnostics.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/mnncorrect">https://github.com/libscran/mnncorrect</a>, which describes the MNN correction algorithm in more detail.</p>
</dd></dl>

</section>
<section id="module-scranpy.crispr_quality_control">
<span id="scranpy-crispr-quality-control-module"></span><h2>scranpy.crispr_quality_control module<a class="headerlink" href="#module-scranpy.crispr_quality_control" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.crispr_quality_control.</span></span><span class="sig-name descname"><span class="pre">ComputeCrisprQcMetricsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">detected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/crispr_quality_control.html#ComputeCrisprQcMetricsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.crispr_quality_control.compute_crispr_qc_metrics" title="scranpy.crispr_quality_control.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'detected':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'max_index':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'max_value':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'sum':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'detected':</span> <span class="pre">Field(name='detected',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'max_index':</span> <span class="pre">Field(name='max_index',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'max_value':</span> <span class="pre">Field(name='max_value',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'sum':</span> <span class="pre">Field(name='sum',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('sum',</span> <span class="pre">'detected',</span> <span class="pre">'max_value',</span> <span class="pre">'max_index')</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.detected">
<span class="sig-name descname"><span class="pre">detected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.detected" title="Permalink to this definition">¶</a></dt>
<dd><p>Integer array of length equal to the number of cells, containing the number of detected guides in each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.max_index">
<span class="sig-name descname"><span class="pre">max_index</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.max_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Integer array of length equal to the number of cells, containing the row index of the guide with the maximum count in each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.max_value">
<span class="sig-name descname"><span class="pre">max_value</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.max_value" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of cells, containing the maximum count for each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.sum">
<span class="sig-name descname"><span class="pre">sum</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of cells, containing the sum of counts across all guides for each cell.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.to_biocframe">
<span class="sig-name descname"><span class="pre">to_biocframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/crispr_quality_control.html#ComputeCrisprQcMetricsResults.to_biocframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.to_biocframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the results into a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> where each row corresponds to a cell and each column is one of the metrics.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.crispr_quality_control.</span></span><span class="sig-name descname"><span class="pre">SuggestCrisprQcThresholdsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/crispr_quality_control.html#SuggestCrisprQcThresholdsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds" title="scranpy.crispr_quality_control.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'block':</span> <span class="pre">typing.Optional[list],</span> <span class="pre">'max_value':</span> <span class="pre">typing.Union[biocutils.NamedList.NamedList,</span> <span class="pre">float]}</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'block':</span> <span class="pre">Field(name='block',type=typing.Optional[list],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'max_value':</span> <span class="pre">Field(name='max_value',type=typing.Union[biocutils.NamedList.NamedList,</span> <span class="pre">float],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('max_value',</span> <span class="pre">'block')</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.block">
<span class="sig-name descname"><span class="pre">block</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.block" title="Permalink to this definition">¶</a></dt>
<dd><p>Levels of the blocking factor.
Each entry corresponds to a element of <a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.max_value" title="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.max_value"><code class="xref py py-attr docutils literal notranslate"><span class="pre">max_value</span></code></a> if <code class="docutils literal notranslate"><span class="pre">block</span></code> was provided in <a class="reference internal" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds" title="scranpy.crispr_quality_control.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.
This is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> if no blocking was performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.max_value">
<span class="sig-name descname"><span class="pre">max_value</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">,</span> <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.max_value" title="Permalink to this definition">¶</a></dt>
<dd><p>Threshold on the maximum count in each cell.
Cells with lower maxima are considered to be of low quality.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided in <a class="reference internal" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds" title="scranpy.crispr_quality_control.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>, a list is returned containing a separate threshold for each level of the factor.
Otherwise, a single float is returned containing the threshold for all cells.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.compute_crispr_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.crispr_quality_control.</span></span><span class="sig-name descname"><span class="pre">compute_crispr_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/crispr_quality_control.html#compute_crispr_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.crispr_quality_control.compute_crispr_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute quality control metrics from CRISPR count data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object containing CRISPR counts.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>QC metrics computed from the count matrix for each cell.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind these QC metrics.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.filter_crispr_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.crispr_quality_control.</span></span><span class="sig-name descname"><span class="pre">filter_crispr_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/crispr_quality_control.html#filter_crispr_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.crispr_quality_control.filter_crispr_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter for high-quality cells based on CRISPR-derived QC metrics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thresholds</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults" title="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults</span></code></a></span>) – Filter thresholds on the QC metrics, typically computed with <a class="reference internal" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds" title="scranpy.crispr_quality_control.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults" title="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults</span></code></a></span>) – CRISPR-derived QC metrics, typically computed with <a class="reference internal" href="#scranpy.crispr_quality_control.compute_crispr_qc_metrics" title="scranpy.crispr_quality_control.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
The levels should be a subset of those used in <a class="reference internal" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds" title="scranpy.crispr_quality_control.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A NumPy vector of length equal to the number of cells in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>, containing truthy values for putative high-quality cells.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.crispr_quality_control.suggest_crispr_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.crispr_quality_control.</span></span><span class="sig-name descname"><span class="pre">suggest_crispr_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/crispr_quality_control.html#suggest_crispr_qc_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Suggest filter thresholds for the CRISPR-derived QC metrics, typically generated from <a class="reference internal" href="#scranpy.crispr_quality_control.compute_crispr_qc_metrics" title="scranpy.crispr_quality_control.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults" title="scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults</span></code></a></span>) – CRISPR-derived QC metrics from <a class="reference internal" href="#scranpy.crispr_quality_control.compute_crispr_qc_metrics" title="scranpy.crispr_quality_control.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>num_mads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number of MADs from the median to define the threshold for outliers in each QC metric.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults" title="scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Suggested filters on the relevant QC metrics.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_filters</span></code> and <code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_filters_blocked</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind the suggested filters.</p>
</dd></dl>

</section>
<section id="module-scranpy.fit_variance_trend">
<span id="scranpy-fit-variance-trend-module"></span><h2>scranpy.fit_variance_trend module<a class="headerlink" href="#module-scranpy.fit_variance_trend" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.fit_variance_trend.fit_variance_trend">
<span class="sig-prename descclassname"><span class="pre">scranpy.fit_variance_trend.</span></span><span class="sig-name descname"><span class="pre">fit_variance_trend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mean</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_min_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_window_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/fit_variance_trend.html#fit_variance_trend"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.fit_variance_trend.fit_variance_trend" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit a trend to the per-cell variances with respect to the mean.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Array containing the mean (log-)expression for each gene.</p></li>
<li><p><strong>variance</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Array containing the variance in the (log-)expression for each gene.
This should have length equal to <code class="docutils literal notranslate"><span class="pre">mean</span></code>.</p></li>
<li><p><strong>mean_filter</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to filter on the means before trend fitting.</p></li>
<li><p><strong>min_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – The minimum mean of genes to use in trend fitting.
Only used if <code class="docutils literal notranslate"><span class="pre">mean_filter</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>transform</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether a quarter-root transformation should be applied before trend fitting.</p></li>
<li><p><strong>span</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Span of the LOWESS smoother.
Ignored if <code class="docutils literal notranslate"><span class="pre">use_min_width</span> <span class="pre">=</span> <span class="pre">TRUE</span></code>.</p></li>
<li><p><strong>use_min_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether a minimum width constraint should be applied to the LOWESS smoother.
This is useful to avoid overfitting in high-density intervals.</p></li>
<li><p><strong>min_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum width of the window to use when <code class="docutils literal notranslate"><span class="pre">use_min_width</span> <span class="pre">=</span> <span class="pre">TRUE</span></code>.</p></li>
<li><p><strong>min_window_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum number of observations in each window.
Only used if <code class="docutils literal notranslate"><span class="pre">use_min_width=TRUE</span></code>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Tuple" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A tuple of two arrays.
The first array contains the fitted value of the trend for each gene while the second array contains the residual.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/LTLA/CppWeightedLowess">https://github.com/LTLA/CppWeightedLowess</a>, which provides the underlying implementation of the LOWESS smoother.</p>
</dd></dl>

</section>
<section id="module-scranpy.lib_scranpy">
<span id="scranpy-lib-scranpy-module"></span><h2>scranpy.lib_scranpy module<a class="headerlink" href="#module-scranpy.lib_scranpy" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.aggregate_across_cells">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.aggregate_across_cells" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.aggregate_across_genes">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><span class="pre">list</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><span class="pre">list</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.aggregate_across_genes" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.build_snn_graph">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">build_snn_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.build_snn_graph" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.center_size_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">center_size_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.center_size_factors" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.choose_highly_variable_genes">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">choose_highly_variable_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.choose_highly_variable_genes" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.choose_pseudo_count">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">choose_pseudo_count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.choose_pseudo_count" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.cluster_kmeans">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">cluster_kmeans</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg8</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg9</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg10</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg11</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.cluster_kmeans" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.cluster_leiden">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">cluster_leiden</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.cluster_leiden" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.cluster_multilevel">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">cluster_multilevel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.cluster_multilevel" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.cluster_walktrap">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">cluster_walktrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.cluster_walktrap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.combine_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">combine_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.combine_factors" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.compute_adt_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_adt_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><span class="pre">list</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.compute_adt_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.compute_clrm1_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_clrm1_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.compute_clrm1_factors" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.compute_crispr_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_crispr_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.compute_crispr_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.compute_rna_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_rna_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><span class="pre">list</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.compute_rna_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.correct_mnn">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">correct_mnn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg8</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg9</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg10</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.correct_mnn" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.filter_adt_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">filter_adt_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.filter_adt_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.filter_crispr_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">filter_crispr_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.filter_crispr_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.filter_rna_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">filter_rna_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.filter_rna_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.fit_variance_trend">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">fit_variance_trend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg8</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg9</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.fit_variance_trend" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.model_gene_variances">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">model_gene_variances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg8</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg9</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg10</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg11</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg12</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.model_gene_variances" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.normalize_counts">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">normalize_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.normalize_counts" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.perplexity_to_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">perplexity_to_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.perplexity_to_neighbors" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.run_pca">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">run_pca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg8</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg9</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg10</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg11</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.run_pca" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.run_tsne">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">run_tsne</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.run_tsne" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.run_umap">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">run_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.run_umap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.sanitize_size_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">sanitize_size_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.sanitize_size_factors" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.scale_by_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">scale_by_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><span class="pre">list</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.scale_by_neighbors" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.score_gene_set">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">score_gene_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg8</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg9</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg10</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.score_gene_set" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.score_markers_pairwise">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">score_markers_pairwise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg8</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg9</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg10</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg11</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.score_markers_pairwise" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.score_markers_summary">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">score_markers_summary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#str" title="(in Python v3.11)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg6</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg7</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg8</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg9</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg10</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg11</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.score_markers_summary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.subsample_by_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">subsample_by_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.subsample_by_neighbors" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.suggest_adt_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">suggest_adt_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.suggest_adt_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.suggest_crispr_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">suggest_crispr_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.suggest_crispr_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.suggest_rna_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">suggest_rna_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><span class="pre">None</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><span class="pre">float</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.suggest_rna_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.summarize_effects">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">summarize_effects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#tuple" title="(in Python v3.11)"><span class="pre">tuple</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.summarize_effects" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.lib_scranpy.test_enrichment">
<span class="sig-prename descclassname"><span class="pre">scranpy.lib_scranpy.</span></span><span class="sig-name descname"><span class="pre">test_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg0</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg3</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg4</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><span class="pre">bool</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">arg5</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><span class="pre">numpy.ndarray</span></a></span></span><a class="headerlink" href="#scranpy.lib_scranpy.test_enrichment" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</section>
<section id="module-scranpy.model_gene_variances">
<span id="scranpy-model-gene-variances-module"></span><h2>scranpy.model_gene_variances module<a class="headerlink" href="#module-scranpy.model_gene_variances" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.model_gene_variances.</span></span><span class="sig-name descname"><span class="pre">ModelGeneVariancesResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mean</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitted</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">per_block</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/model_gene_variances.html#ModelGeneVariancesResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.model_gene_variances.model_gene_variances" title="scranpy.model_gene_variances.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'fitted':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'mean':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'per_block':</span> <span class="pre">typing.Optional[biocutils.NamedList.NamedList],</span> <span class="pre">'residual':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'variance':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'fitted':</span> <span class="pre">Field(name='fitted',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'mean':</span> <span class="pre">Field(name='mean',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'per_block':</span> <span class="pre">Field(name='per_block',type=typing.Optional[biocutils.NamedList.NamedList],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'residual':</span> <span class="pre">Field(name='residual',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'variance':</span> <span class="pre">Field(name='variance',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('mean',</span> <span class="pre">'variance',</span> <span class="pre">'fitted',</span> <span class="pre">'residual',</span> <span class="pre">'per_block')</span></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.fitted">
<span class="sig-name descname"><span class="pre">fitted</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.fitted" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes, containing the fitted value of the mean-variance trend for each gene.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.mean">
<span class="sig-name descname"><span class="pre">mean</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes, containing the mean (log-)expression for each gene.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.per_block">
<span class="sig-name descname"><span class="pre">per_block</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.per_block" title="Permalink to this definition">¶</a></dt>
<dd><p>List of per-block results, obtained from modelling the variances separately for each block of cells.
Each entry is another <a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults" title="scranpy.model_gene_variances.ModelGeneVariancesResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ModelGeneVariancesResults</span></code></a> object, containing the statistics for the corresponding block.
This is only filled if <code class="docutils literal notranslate"><span class="pre">block</span></code> was used in <a class="reference internal" href="#scranpy.model_gene_variances.model_gene_variances" title="scranpy.model_gene_variances.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>, otherwise it is set to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.residual">
<span class="sig-name descname"><span class="pre">residual</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.residual" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes, containing the residual from the mean-variance trend for each gene.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.to_biocframe">
<span class="sig-name descname"><span class="pre">to_biocframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_per_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/model_gene_variances.html#ModelGeneVariancesResults.to_biocframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.to_biocframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the results into a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>include_per_block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to include the per-block results as a nested dataframe.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> where each row corresponds to a gene and each column corresponds to a statistic.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.ModelGeneVariancesResults.variance">
<span class="sig-name descname"><span class="pre">variance</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.variance" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes, containing the variance in (log-)expression for each gene.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.model_gene_variances.model_gene_variances">
<span class="sig-prename descclassname"><span class="pre">scranpy.model_gene_variances.</span></span><span class="sig-name descname"><span class="pre">model_gene_variances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_min_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_window_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/model_gene_variances.html#model_gene_variances"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.model_gene_variances.model_gene_variances" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the variance in (log-)expression values for each gene, and model the trend in the variances with respect to the mean.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
It is typically expected to contain log-expression values, e.g., from <a class="reference internal" href="#scranpy.normalize_counts.normalize_counts" title="scranpy.normalize_counts.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the block of origin (e.g., batch, sample) for each cell.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_weight_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'variable'</span></code>, <code class="docutils literal notranslate"><span class="pre">'equal'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – Policy to use for weighting different blocks when computing the average for each statistic.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>variable_block_weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Tuple" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span>) – Parameters for variable block weighting.
This should be a tuple of length 2 where the first and second values are used as the lower and upper bounds, respectively, for the variable weight calculation.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_weight_policy</span> <span class="pre">=</span> <span class="pre">&quot;variable&quot;</span></code>.</p></li>
<li><p><strong>mean_filter</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to filter on the means before trend fitting.</p></li>
<li><p><strong>min_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – The minimum mean of genes to use in trend fitting.
Only used if <code class="docutils literal notranslate"><span class="pre">mean_filter</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>transform</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether a quarter-root transformation should be applied before trend fitting.</p></li>
<li><p><strong>span</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Span of the LOWESS smoother for trend fitting, see <a class="reference internal" href="#scranpy.fit_variance_trend.fit_variance_trend" title="scranpy.fit_variance_trend.fit_variance_trend"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a>.</p></li>
<li><p><strong>use_min_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether a minimum width constraint should be applied during trend fitting, see <a class="reference internal" href="#scranpy.fit_variance_trend.fit_variance_trend" title="scranpy.fit_variance_trend.fit_variance_trend"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a>.</p></li>
<li><p><strong>min_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum width of the smoothing window for trend fitting, see <a class="reference internal" href="#scranpy.fit_variance_trend.fit_variance_trend" title="scranpy.fit_variance_trend.fit_variance_trend"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a>.</p></li>
<li><p><strong>min_window_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum number of observations in each smoothing window for trend fitting, see <a class="reference internal" href="#scranpy.fit_variance_trend.fit_variance_trend" title="scranpy.fit_variance_trend.fit_variance_trend"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults" title="scranpy.model_gene_variances.ModelGeneVariancesResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ModelGeneVariancesResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The results of the variance modelling for each gene.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/LTLA/CppWeightedLowess">https://github.com/LTLA/CppWeightedLowess</a>, which provides the underlying implementation of the LOWESS smoother.</p>
</dd></dl>

</section>
<section id="module-scranpy.normalize_counts">
<span id="scranpy-normalize-counts-module"></span><h2>scranpy.normalize_counts module<a class="headerlink" href="#module-scranpy.normalize_counts" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.normalize_counts.normalize_counts">
<span class="sig-prename descclassname"><span class="pre">scranpy.normalize_counts.</span></span><span class="sig-name descname"><span class="pre">normalize_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudo_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_sparsity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/normalize_counts.html#normalize_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.normalize_counts.normalize_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a matrix of (log-transformed) normalized expression values.
The normalization removes uninteresting per-cell differences due to sequencing efficiency and library size.
The log-transformation ensures that any differences represent log-fold changes in downstream analysis steps; such relative changes in expression are more relevant than absolute changes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – <p>Matrix-like object containing cells in columns and features in rows, typically with count data.</p>
<p>Alternatively, a <a class="reference external" href="https://tatami-inc.github.io/mattress/api/mattress.html#mattress.InitializedMatrix.InitializedMatrix" title="(in mattress vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitializedMatrix</span></code></a> representing a count matrix, typically created by <code class="xref py py-class docutils literal notranslate"><span class="pre">initialize</span></code>.</p>
</p></li>
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Size factor for each cell. This should have length equal to the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>log</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether log-transformation should be performed.</p></li>
<li><p><strong>pseudo_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Positive pseudo-count to add before log-transformation.
Ignored if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>log_base</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Base of the log-transformation, ignored if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>preserve_sparsity</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to preserve sparsity when <code class="docutils literal notranslate"><span class="pre">pseudo_count</span> <span class="pre">!=</span> <span class="pre">1</span></code>.
If <code class="docutils literal notranslate"><span class="pre">True</span></code>, users should manually add <code class="docutils literal notranslate"><span class="pre">log(pseudo_count,</span> <span class="pre">log_base)</span></code> to the returned matrix to obtain the desired log-transformed expression values.
Ignored if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">False</span></code> or <code class="docutils literal notranslate"><span class="pre">pseudo_count</span> <span class="pre">=</span> <span class="pre">1</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a>, <a class="reference external" href="https://tatami-inc.github.io/mattress/api/mattress.html#mattress.InitializedMatrix.InitializedMatrix" title="(in mattress vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitializedMatrix</span></code></a>]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>If <code class="docutils literal notranslate"><span class="pre">x</span></code> is a matrix-like object, a <a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a> is returned containing the (log-transformed) normalized expression matrix.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">x</span></code> is an <code class="docutils literal notranslate"><span class="pre">InitializedMatrix</span></code>, a new <code class="docutils literal notranslate"><span class="pre">InitializedMatrix</span></code> is returned containing the normalized expression matrix.</p>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">normalize_counts</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_norm">scran_norm</a> C++ library, which provides the reference implementation.</p>
</dd></dl>

</section>
<section id="module-scranpy.rna_quality_control">
<span id="scranpy-rna-quality-control-module"></span><h2>scranpy.rna_quality_control module<a class="headerlink" href="#module-scranpy.rna_quality_control" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.rna_quality_control.</span></span><span class="sig-name descname"><span class="pre">ComputeRnaQcMetricsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">detected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_proportion</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/rna_quality_control.html#ComputeRnaQcMetricsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.rna_quality_control.compute_rna_qc_metrics" title="scranpy.rna_quality_control.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'detected':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'subset_proportion':</span> <span class="pre">&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,</span> <span class="pre">'sum':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'detected':</span> <span class="pre">Field(name='detected',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'subset_proportion':</span> <span class="pre">Field(name='subset_proportion',type=&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'sum':</span> <span class="pre">Field(name='sum',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('sum',</span> <span class="pre">'detected',</span> <span class="pre">'subset_proportion')</span></em><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.detected">
<span class="sig-name descname"><span class="pre">detected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.detected" title="Permalink to this definition">¶</a></dt>
<dd><p>Integer array of length equal to the number of cells, containing the number of detected genes in each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.subset_proportion">
<span class="sig-name descname"><span class="pre">subset_proportion</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></em><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.subset_proportion" title="Permalink to this definition">¶</a></dt>
<dd><p>Proportion of counts in each gene subset in each cell.
Each list element corresponds to a gene subset and is a NumPy array of length equal to the number of cells.
Each entry of the array contains the proportion of counts in that subset in each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.sum">
<span class="sig-name descname"><span class="pre">sum</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of cells, containing the sum of counts across all genes for each cell.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.to_biocframe">
<span class="sig-name descname"><span class="pre">to_biocframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flatten</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/rna_quality_control.html#ComputeRnaQcMetricsResults.to_biocframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.to_biocframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the results into a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>flatten</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to flatten the subset proportions into separate columns.
If <code class="docutils literal notranslate"><span class="pre">True</span></code>, each entry of <a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.subset_proportion" title="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.subset_proportion"><code class="xref py py-attr docutils literal notranslate"><span class="pre">subset_proportion</span></code></a> is represented by a <code class="docutils literal notranslate"><span class="pre">subset_proportion_&lt;NAME&gt;</span></code> column,
where <code class="docutils literal notranslate"><span class="pre">&lt;NAME&gt;</span></code> is the the name of each entry (if available) or its index (otherwise).
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, <a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.subset_proportion" title="scranpy.rna_quality_control.ComputeRnaQcMetricsResults.subset_proportion"><code class="xref py py-attr docutils literal notranslate"><span class="pre">subset_proportion</span></code></a> is represented by a nested <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> where each row corresponds to a cell and each column is one of the metrics.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.rna_quality_control.</span></span><span class="sig-name descname"><span class="pre">SuggestRnaQcThresholdsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">detected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_proportion</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/rna_quality_control.html#SuggestRnaQcThresholdsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'block':</span> <span class="pre">typing.Optional[list],</span> <span class="pre">'detected':</span> <span class="pre">typing.Union[biocutils.NamedList.NamedList,</span> <span class="pre">float],</span> <span class="pre">'subset_proportion':</span> <span class="pre">&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,</span> <span class="pre">'sum':</span> <span class="pre">typing.Union[biocutils.NamedList.NamedList,</span> <span class="pre">float]}</span></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'block':</span> <span class="pre">Field(name='block',type=typing.Optional[list],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'detected':</span> <span class="pre">Field(name='detected',type=typing.Union[biocutils.NamedList.NamedList,</span> <span class="pre">float],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'subset_proportion':</span> <span class="pre">Field(name='subset_proportion',type=&lt;class</span> <span class="pre">'biocutils.NamedList.NamedList'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'sum':</span> <span class="pre">Field(name='sum',type=typing.Union[biocutils.NamedList.NamedList,</span> <span class="pre">float],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('sum',</span> <span class="pre">'detected',</span> <span class="pre">'subset_proportion',</span> <span class="pre">'block')</span></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.block">
<span class="sig-name descname"><span class="pre">block</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.block" title="Permalink to this definition">¶</a></dt>
<dd><p>Levels of the blocking factor.
Each entry corresponds to a element of <a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.sum" title="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.sum"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sum</span></code></a>, <a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.detected" title="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.detected"><code class="xref py py-attr docutils literal notranslate"><span class="pre">detected</span></code></a>, etc., if <code class="docutils literal notranslate"><span class="pre">block</span></code> was provided in <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.
This is set to <code class="docutils literal notranslate"><span class="pre">None</span></code> if no blocking was performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.detected">
<span class="sig-name descname"><span class="pre">detected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">,</span> <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.detected" title="Permalink to this definition">¶</a></dt>
<dd><p>Threshold on the number of detected genes.
Cells with lower numbers of detected genes are considered to be of low quality.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided in <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>, a list is returned containing a separate threshold for each level of the factor.
Otherwise, a single float is returned containing the threshold for all cells.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.subset_proportion">
<span class="sig-name descname"><span class="pre">subset_proportion</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.subset_proportion" title="Permalink to this definition">¶</a></dt>
<dd><p>Thresholds on the sum of counts in each gene subset.
Each element of the list corresponds to a gene subset.
Cells with higher sums than the threshold for any subset are considered to be of low quality.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided in <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>, each entry of the returned list is another <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>  containing a separate threshold for each level.
Otherwise, each entry of the list is a single float containing the threshold for all cells.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.sum">
<span class="sig-name descname"><span class="pre">sum</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">,</span> <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Threshold on the sum of counts in each cell.
Cells with lower totals are considered to be of low quality.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided in <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>, a list is returned containing a separate threshold for each level of the factor.
Otherwise, a single float is returned containing the threshold for all cells.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.compute_rna_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.rna_quality_control.</span></span><span class="sig-name descname"><span class="pre">compute_rna_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/rna_quality_control.html#compute_rna_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.rna_quality_control.compute_rna_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute quality control metrics from RNA count data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object containing RNA counts.</p></li>
<li><p><strong>subsets</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>, <a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – <p>Subsets of genes corresponding to “control” features like mitochondrial genes.
This may be either:</p>
<ul>
<li><p>A list of arrays.
Each array corresponds to an gene subset and can either contain boolean or integer values.
For booleans, the array should be of length equal to the number of rows, and values should be truthy for rows that belong in the subset.
For integers, each element of the array is treated the row index of an gene in the subset.</p></li>
<li><p>A dictionary where keys are the names of each gene subset and the values are arrays as described above.</p></li>
<li><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> where each element is an array as described above, possibly with names.</p></li>
</ul>
</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>QC metrics computed from the count matrix for each cell.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_rna_qc_metrics</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind these QC metrics.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.filter_rna_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.rna_quality_control.</span></span><span class="sig-name descname"><span class="pre">filter_rna_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/rna_quality_control.html#filter_rna_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.rna_quality_control.filter_rna_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter for high-quality cells based on RNA-derived QC metrics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thresholds</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults" title="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults</span></code></a></span>) – Filter thresholds on the QC metrics, typically computed with <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults" title="scranpy.rna_quality_control.ComputeRnaQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults</span></code></a></span>) – RNA-derived QC metrics, typically computed with <a class="reference internal" href="#scranpy.rna_quality_control.compute_rna_qc_metrics" title="scranpy.rna_quality_control.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
The levels should be a subset of those used in <a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A NumPy vector of length equal to the number of cells in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>, containing truthy values for putative high-quality cells.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.rna_quality_control.suggest_rna_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.rna_quality_control.</span></span><span class="sig-name descname"><span class="pre">suggest_rna_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/rna_quality_control.html#suggest_rna_qc_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Suggest filter thresholds for the RNA-derived QC metrics, typically generated from <a class="reference internal" href="#scranpy.rna_quality_control.compute_rna_qc_metrics" title="scranpy.rna_quality_control.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults" title="scranpy.rna_quality_control.ComputeRnaQcMetricsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults</span></code></a></span>) – RNA-derived QC metrics from <a class="reference internal" href="#scranpy.rna_quality_control.compute_rna_qc_metrics" title="scranpy.rna_quality_control.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>num_mads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number of MADs from the median to define the threshold for outliers in each QC metric.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults" title="scranpy.rna_quality_control.SuggestRnaQcThresholdsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Suggested filters on the relevant QC metrics.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_rna_qc_filters</span></code> and <code class="docutils literal notranslate"><span class="pre">compute_rna_qc_filters_blocked</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind the suggested filters.</p>
</dd></dl>

</section>
<section id="module-scranpy.run_all_neighbor_steps">
<span id="scranpy-run-all-neighbor-steps-module"></span><h2>scranpy.run_all_neighbor_steps module<a class="headerlink" href="#module-scranpy.run_all_neighbor_steps" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.run_all_neighbor_steps.</span></span><span class="sig-name descname"><span class="pre">RunAllNeighborStepsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">run_tsne</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_umap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_snn_graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_graph</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/run_all_neighbor_steps.html#RunAllNeighborStepsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.run_all_neighbor_steps.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'build_snn_graph':</span> <span class="pre">typing.Optional[scranpy.build_snn_graph.GraphComponents],</span> <span class="pre">'cluster_graph':</span> <span class="pre">typing.Optional[scranpy.cluster_graph.ClusterGraphResults],</span> <span class="pre">'run_tsne':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'run_umap':</span> <span class="pre">typing.Optional[numpy.ndarray]}</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'build_snn_graph':</span> <span class="pre">Field(name='build_snn_graph',type=typing.Optional[scranpy.build_snn_graph.GraphComponents],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'cluster_graph':</span> <span class="pre">Field(name='cluster_graph',type=typing.Optional[scranpy.cluster_graph.ClusterGraphResults],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'run_tsne':</span> <span class="pre">Field(name='run_tsne',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'run_umap':</span> <span class="pre">Field(name='run_umap',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('run_tsne',</span> <span class="pre">'run_umap',</span> <span class="pre">'build_snn_graph',</span> <span class="pre">'cluster_graph')</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.build_snn_graph">
<span class="sig-name descname"><span class="pre">build_snn_graph</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents" title="scranpy.build_snn_graph.GraphComponents"><code class="xref py py-class docutils literal notranslate"><span class="pre">GraphComponents</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.build_snn_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.build_snn_graph.build_snn_graph" title="scranpy.build_snn_graph.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if clustering was not performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.cluster_graph">
<span class="sig-name descname"><span class="pre">cluster_graph</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults" title="scranpy.cluster_graph.ClusterGraphResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterGraphResults</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.cluster_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if clustering was not performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.run_tsne">
<span class="sig-name descname"><span class="pre">run_tsne</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.run_tsne" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.run_tsne.run_tsne" title="scranpy.run_tsne.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if t-SNE was not performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.run_umap">
<span class="sig-name descname"><span class="pre">run_umap</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.run_umap" title="Permalink to this definition">¶</a></dt>
<dd><p>Results of <a class="reference internal" href="#scranpy.run_umap.run_umap" title="scranpy.run_umap.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if UMAP was not performed.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps.run_all_neighbor_steps">
<span class="sig-prename descclassname"><span class="pre">scranpy.run_all_neighbor_steps.</span></span><span class="sig-name descname"><span class="pre">run_all_neighbor_steps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_umap_options={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_tsne_options={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_snn_graph_options={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_graph_options={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_search=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=3</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/run_all_neighbor_steps.html#run_all_neighbor_steps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_all_neighbor_steps.run_all_neighbor_steps" title="Permalink to this definition">¶</a></dt>
<dd><p>Run all steps that depend on the nearest neighbor search -
namely, <a class="reference internal" href="#scranpy.run_tsne.run_tsne" title="scranpy.run_tsne.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>, <a class="reference internal" href="#scranpy.run_umap.run_umap" title="scranpy.run_umap.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>, <a class="reference internal" href="#scranpy.build_snn_graph.build_snn_graph" title="scranpy.build_snn_graph.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>, and <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
This builds the index once and re-uses it for the neighbor search in each step; the various steps are also run in parallel to save more time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>]</span>) – <p>Matrix of principal components where rows are cells and columns are PCs, typically produced by <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code> instance containing a prebuilt search index for the cells.</p>
</p></li>
<li><p><strong>run_umap_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Optional arguments for <a class="reference internal" href="#scranpy.run_umap.run_umap" title="scranpy.run_umap.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, UMAP is not performed.</p></li>
<li><p><strong>run_tsne_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Optional arguments for <a class="reference internal" href="#scranpy.run_tsne.run_tsne" title="scranpy.run_tsne.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, t-SNE is not performed.</p></li>
<li><p><strong>build_snn_graph_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Optional arguments for <a class="reference internal" href="#scranpy.build_snn_graph.build_snn_graph" title="scranpy.build_snn_graph.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.
Ignored if <code class="docutils literal notranslate"><span class="pre">cluster_graph_options</span> <span class="pre">=</span> <span class="pre">None</span></code>.</p></li>
<li><p><strong>cluster_graph_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Optional arguments for <a class="reference internal" href="#scranpy.cluster_graph.cluster_graph" title="scranpy.cluster_graph.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, graph-based clustering is not performed.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Parameters for the nearest-neighbor search.</p></li>
<li><p><strong>collapse_search</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to collapse the nearest-neighbor search for each step into a single search.
Steps that need fewer neighbors will use a subset of the neighbors from the collapsed search.
This is faster but may not give the same results as separate searches for some approximate search algorithms.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for the parallel execution of UMAP, t-SNE and SNN graph construction.
This overrides the specified number of threads in the various <code class="docutils literal notranslate"><span class="pre">*_options</span></code> arguments.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults" title="scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The results of each step.
These should be equivalent to the result of running each step in serial.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scranpy.run_pca">
<span id="scranpy-run-pca-module"></span><h2>scranpy.run_pca module<a class="headerlink" href="#module-scranpy.run_pca" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.run_pca.</span></span><span class="sig-name descname"><span class="pre">RunPcaResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">components</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variance_explained</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_variance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/run_pca.html#RunPcaResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_pca.RunPcaResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'block':</span> <span class="pre">typing.Optional[list],</span> <span class="pre">'center':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'components':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'rotation':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'scale':</span> <span class="pre">typing.Optional[numpy.ndarray],</span> <span class="pre">'total_variance':</span> <span class="pre">&lt;class</span> <span class="pre">'float'&gt;,</span> <span class="pre">'variance_explained':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'block':</span> <span class="pre">Field(name='block',type=typing.Optional[list],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'center':</span> <span class="pre">Field(name='center',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'components':</span> <span class="pre">Field(name='components',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'rotation':</span> <span class="pre">Field(name='rotation',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'scale':</span> <span class="pre">Field(name='scale',type=typing.Optional[numpy.ndarray],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'total_variance':</span> <span class="pre">Field(name='total_variance',type=&lt;class</span> <span class="pre">'float'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'variance_explained':</span> <span class="pre">Field(name='variance_explained',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('components',</span> <span class="pre">'rotation',</span> <span class="pre">'variance_explained',</span> <span class="pre">'total_variance',</span> <span class="pre">'center',</span> <span class="pre">'scale',</span> <span class="pre">'block')</span></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.block">
<span class="sig-name descname"><span class="pre">block</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.block" title="Permalink to this definition">¶</a></dt>
<dd><p>Levels of the blocking factor, corresponding to each row of <code class="docutils literal notranslate"><span class="pre">center</span></code>.
This is <code class="docutils literal notranslate"><span class="pre">None</span></code> if no blocking was performed.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.center">
<span class="sig-name descname"><span class="pre">center</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.center" title="Permalink to this definition">¶</a></dt>
<dd><p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> was used in <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>, this is a floating-point matrix containing the mean for each gene (column) in each block of cells (row).
Otherwise, this is a floating-point array of length equal to the number of genes, containing the mean for each gene across all cells.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.components">
<span class="sig-name descname"><span class="pre">components</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.components" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point matrix of principal component (PC) scores.
Rows are dimensions (i.e., PCs) and columns are cells.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.rotation">
<span class="sig-name descname"><span class="pre">rotation</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.rotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point rotation matrix.
Rows are genes and columns are dimensions (i.e., PCs).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.scale">
<span class="sig-name descname"><span class="pre">scale</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.scale" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array containing the scaling factor applied to each gene.
Only reported if <code class="docutils literal notranslate"><span class="pre">scale</span> <span class="pre">=</span> <span class="pre">True</span></code>, otherwise it is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.total_variance">
<span class="sig-name descname"><span class="pre">total_variance</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.total_variance" title="Permalink to this definition">¶</a></dt>
<dd><p>Total variance in the dataset.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.run_pca.RunPcaResults.variance_explained">
<span class="sig-name descname"><span class="pre">variance_explained</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.run_pca.RunPcaResults.variance_explained" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of PCs, containing the variance explained by each successive PC.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_pca.run_pca">
<span class="sig-prename descclassname"><span class="pre">scranpy.run_pca.</span></span><span class="sig-name descname"><span class="pre">run_pca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">components_from_residuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_work</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5489</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">realized</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/run_pca.html#run_pca"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_pca.run_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Run a PCA on the gene-by-cell log-expression matrix to obtain a low-dimensional representation for downstream analyses.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
Typically, the matrix is expected to contain log-expression values, and the rows should be filtered to relevant (e.g., highly variable) genes.</p></li>
<li><p><strong>number</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of PCs to retain.</p></li>
<li><p><strong>scale</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to scale all genes to have the same variance.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the block of origin (e.g., batch, sample) for each cell.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_weight_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'variable'</span></code>, <code class="docutils literal notranslate"><span class="pre">'equal'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – Policy to use for weighting different blocks when computing the average for each statistic.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>variable_block_weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Tuple" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span>) – Parameters for variable block weighting.
This should be a tuple of length 2 where the first and second values are used as the lower and upper bounds, respectively, for the variable weight calculation.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_weight_policy</span> <span class="pre">=</span> <span class="pre">&quot;variable&quot;</span></code>.</p></li>
<li><p><strong>components_from_residuals</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the PC scores from the residuals in the presence of a blocking factor.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, the residuals are only used to compute the rotation matrix, and the original expression values of the cells are projected onto this new space.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>extra_work</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of extra dimensions for the IRLBA workspace.</p></li>
<li><p><strong>iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum number of restart iterations for IRLBA.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Seed for the initial random vector in IRLBA.</p></li>
<li><p><strong>realized</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to realize <code class="docutils literal notranslate"><span class="pre">x</span></code> into an optimal memory layout for IRLBA.
This speeds up computation at the cost of increased memory usage.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.run_pca.RunPcaResults" title="scranpy.run_pca.RunPcaResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">RunPcaResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The results of the PCA.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/scran_pca">https://github.com/libscran/scran_pca</a>, which describes the approach in more detail.
In particular, see the documentation for the <code class="docutils literal notranslate"><span class="pre">blocked_pca</span></code> function for an explanation of the blocking strategy.</p>
</dd></dl>

</section>
<section id="module-scranpy.run_tsne">
<span id="scranpy-run-tsne-module"></span><h2>scranpy.run_tsne module<a class="headerlink" href="#module-scranpy.run_tsne" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_tsne.run_tsne">
<span class="sig-prename descclassname"><span class="pre">scranpy.run_tsne.</span></span><span class="sig-name descname"><span class="pre">run_tsne</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity=30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_depth=20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_approximation=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iterations=500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed=42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/run_tsne.html#run_tsne"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_tsne.run_tsne" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute t-SNE coordinates to visualize similarities between cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>]</span>) – <p>Numeric matrix where rows are dimensions and columns are cells, typically containing a low-dimensional representation from, e.g., <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object containing existing neighbor search results.
The number of neighbors should be the same as <code class="docutils literal notranslate"><span class="pre">num_neighbors</span></code>, otherwise a warning is raised.</p>
<p>Alternatively, a <code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code> object.</p>
</p></li>
<li><p><strong>perplexity</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Perplexity to use in the t-SNE algorithm.
Larger values cause the embedding to focus on global structure.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of neighbors in the nearest-neighbor graph.
Typically derived from <code class="docutils literal notranslate"><span class="pre">perplexity</span></code> using <a class="reference internal" href="#scranpy.run_tsne.tsne_perplexity_to_neighbors" title="scranpy.run_tsne.tsne_perplexity_to_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsne_perplexity_to_neighbors()</span></code></a>.</p></li>
<li><p><strong>max_depth</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum depth of the Barnes-Hut quadtree.
Smaller values (7-10) improve speed at the cost of accuracy.</p></li>
<li><p><strong>leaf_approximation</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to use the “leaf approximation” approach, which sacrifices some accuracy for greater speed.
Only effective when <code class="docutils literal notranslate"><span class="pre">max_depth</span></code> is small enough for multiple cells to be assigned to the same leaf node of the quadtree.</p></li>
<li><p><strong>max_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum number of iterations to perform.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Random seed to use for generating the initial coordinates.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – The algorithm to use for the nearest-neighbor search.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> is not a pre-built nearest-neighbor search index or a list of existing nearest-neighbor search results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array containing the coordinates of each cell in a 2-dimensional embedding.
Each row corresponds to a dimension and each column represents a cell.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/qdtsne">https://github.com/libscran/qdtsne</a>, for some more details on the approximations.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_tsne.tsne_perplexity_to_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.run_tsne.</span></span><span class="sig-name descname"><span class="pre">tsne_perplexity_to_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">perplexity</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/run_tsne.html#tsne_perplexity_to_neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_tsne.tsne_perplexity_to_neighbors" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine the number of nearest neighbors required to support a given
perplexity in the t-SNE algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>perplexity</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Perplexity to use in <a class="reference internal" href="#scranpy.run_tsne.run_tsne" title="scranpy.run_tsne.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The corresponding number of nearest neighbors.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scranpy.run_umap">
<span id="scranpy-run-umap-module"></span><h2>scranpy.run_umap module<a class="headerlink" href="#module-scranpy.run_umap" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_umap.run_umap">
<span class="sig-prename descclassname"><span class="pre">scranpy.run_umap.</span></span><span class="sig-name descname"><span class="pre">run_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_dim=2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_epochs=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist=0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed=1234567890</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel_optimization=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/run_umap.html#run_umap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_umap.run_umap" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute UMAP coordinates to visualize similarities between cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>]</span>) – <p>Numeric matrix where rows are dimensions and columns are cells, typically containing a low-dimensional representation from, e.g., <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object containing existing neighbor search results.
The number of neighbors should be the same as <code class="docutils literal notranslate"><span class="pre">num_neighbors</span></code>, otherwise a warning is raised.</p>
<p>Alternatively, a <code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code> object.</p>
</p></li>
<li><p><strong>num_dim</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of dimensions in the UMAP embedding.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to use in the UMAP algorithm.
Larger values cause the embedding to focus on global structure.</p></li>
<li><p><strong>num_epochs</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of epochs to perform.
If set to None, an appropriate number of epochs is chosen based on the number of points in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>min_dist</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum distance between points in the embedding.
Larger values result in more visual clusters that are more dispersed.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Integer scalar specifying the seed to use.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
<li><p><strong>parallel_optimization</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to parallelize the optimization step.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – The algorithm to use for the nearest-neighbor search.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> is not a pre-built nearest-neighbor search index or a list of existing nearest-neighbor search results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array containing the coordinates of each cell in a 2-dimensional embedding.
Each row corresponds to a dimension and each column represents a cell.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/umappp">https://github.com/libscran/umappp</a>, for the underlying implementation.</p>
</dd></dl>

</section>
<section id="module-scranpy.sanitize_size_factors">
<span id="scranpy-sanitize-size-factors-module"></span><h2>scranpy.sanitize_size_factors module<a class="headerlink" href="#module-scranpy.sanitize_size_factors" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.sanitize_size_factors.sanitize_size_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.sanitize_size_factors.</span></span><span class="sig-name descname"><span class="pre">sanitize_size_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_zero</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_negative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_infinite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/sanitize_size_factors.html#sanitize_size_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.sanitize_size_factors.sanitize_size_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace invalid size factors, i.e., zero, negative, infinite or NaNs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Floating-point array containing size factors for all cells.</p></li>
<li><p><strong>replace_zero</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to replace size factors of zero with the lowest positive factor. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, zeros are retained.</p></li>
<li><p><strong>replace_negative</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to replace negative size factors with the lowest positive factor. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, negative values are retained.</p></li>
<li><p><strong>replace_infinite</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to replace infinite size factors with the largest positive factor. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, infinite values are retained.</p></li>
<li><p><strong>replace_nan</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to replace NaN size factors with unity. If False, NaN values are retained.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> in place. If False, a new array
is returned. This argument only used if <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> is
double-precision, otherwise a new array is always returned.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array containing sanitized size factors. If <code class="docutils literal notranslate"><span class="pre">in_place</span> <span class="pre">=</span> <span class="pre">True</span></code>, this
is a reference to <code class="docutils literal notranslate"><span class="pre">size_factors</span></code>.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sanitize_size_factors</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_norm">scran_norm</a> C++ library, which provides the underlying implementation.</p>
</dd></dl>

</section>
<section id="module-scranpy.scale_by_neighbors">
<span id="scranpy-scale-by-neighbors-module"></span><h2>scranpy.scale_by_neighbors module<a class="headerlink" href="#module-scranpy.scale_by_neighbors" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.scale_by_neighbors.</span></span><span class="sig-name descname"><span class="pre">ScaleByNeighborsResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scaling</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combined</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/scale_by_neighbors.html#ScaleByNeighborsResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.scale_by_neighbors.scale_by_neighbors" title="scranpy.scale_by_neighbors.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'combined':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'scaling':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'combined':</span> <span class="pre">Field(name='combined',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'scaling':</span> <span class="pre">Field(name='scaling',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('scaling',</span> <span class="pre">'combined')</span></em><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.combined">
<span class="sig-name descname"><span class="pre">combined</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.combined" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point matrix of scaled embeddings.
Each row corresponds to a dimension and each column corresponds to a cell.
Formed by scaling each embedding in the <code class="docutils literal notranslate"><span class="pre">x</span></code> supplied to <a class="reference internal" href="#scranpy.scale_by_neighbors.scale_by_neighbors" title="scranpy.scale_by_neighbors.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a> by its corresponding entry of <code class="docutils literal notranslate"><span class="pre">scaling</span></code>, and then concatenating them together by row.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.ScaleByNeighborsResults.scaling">
<span class="sig-name descname"><span class="pre">scaling</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.scaling" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array containing the scaling factor to be applied to each embedding.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors.scale_by_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.scale_by_neighbors.</span></span><span class="sig-name descname"><span class="pre">scale_by_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/scale_by_neighbors.html#scale_by_neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.scale_by_neighbors.scale_by_neighbors" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale multiple embeddings (usually derived from different modalities across the same set of cells) so that their within-population variances are comparable.
Then, combine them into a single embedding matrix for combined downstream analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of of numeric matrices of principal components or other embeddings, one for each modality.
For each entry, rows are dimensions and columns are cells.
All entries should have the same number of columns but may have different numbers of rows.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to use to define the scaling factor.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Algorithm for the nearest-neighbor search.</p></li>
<li><p><strong>weights</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to <code class="docutils literal notranslate"><span class="pre">x</span></code>, specifying the weights to apply to each modality.
Each value represents a multiplier of the within-population variance of its modality, i.e., larger values increase the contribution of that modality in the combined output matrix.
The default of <code class="docutils literal notranslate"><span class="pre">None</span></code> is equivalent to an all-1 vector, i.e., all modalities are scaled to have the same within-population variance.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults" title="scranpy.scale_by_neighbors.ScaleByNeighborsResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ScaleByNeighborsResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Scaling factors and the combined matrix from all modalities.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/mumosa">https://github.com/libscran/mumosa</a>, for the basis and caveats of this approach.</p>
</dd></dl>

</section>
<section id="module-scranpy.score_gene_set">
<span id="scranpy-score-gene-set-module"></span><h2>scranpy.score_gene_set module<a class="headerlink" href="#module-scranpy.score_gene_set" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.score_gene_set.</span></span><span class="sig-name descname"><span class="pre">ScoreGeneSetResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/score_gene_set.html#ScoreGeneSetResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.score_gene_set.score_gene_set" title="scranpy.score_gene_set.score_gene_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'scores':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'weights':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'scores':</span> <span class="pre">Field(name='scores',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'weights':</span> <span class="pre">Field(name='weights',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('scores',</span> <span class="pre">'weights')</span></em><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of cells, containing the gene set score for each cell.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_gene_set.ScoreGeneSetResults.weights">
<span class="sig-name descname"><span class="pre">weights</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.score_gene_set.ScoreGeneSetResults.weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes in the set, containing the weight of each gene in terms of its contribution to the score.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.score_gene_set.score_gene_set">
<span class="sig-prename descclassname"><span class="pre">scranpy.score_gene_set.</span></span><span class="sig-name descname"><span class="pre">score_gene_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_work</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5489</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">realized</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/score_gene_set.html#score_gene_set"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.score_gene_set.score_gene_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute per-cell scores for a gene set, defined as the column sums of a rank-1 approximation to the submatrix for the feature set.
This uses the same approach implemented in the GSDecon package by Jason Hackney.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
The matrix is expected to contain log-expression values.</p></li>
<li><p><strong>set</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Array of integer indices specifying the rows of <code class="docutils literal notranslate"><span class="pre">x</span></code> belonging to the gene set.
Alternatively, a sequence of boolean values of length equal to the number of rows, where truthy elements indicate that the corresponding row belongs to the gene set.</p></li>
<li><p><strong>rank</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Rank of the approximation.</p></li>
<li><p><strong>scale</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to scale all genes to have the same variance.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the block of origin (e.g., batch, sample) for each cell.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_weight_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'variable'</span></code>, <code class="docutils literal notranslate"><span class="pre">'equal'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – Policy to use for weighting different blocks when computing the average for each statistic.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>variable_block_weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Tuple" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span>) – Parameters for variable block weighting.
This should be a tuple of length 2 where the first and second values are used as the lower and upper bounds, respectively, for the variable weight calculation.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_weight_policy</span> <span class="pre">=</span> <span class="pre">&quot;variable&quot;</span></code>.</p></li>
<li><p><strong>extra_work</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of extra dimensions for the IRLBA workspace.</p></li>
<li><p><strong>iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum number of restart iterations for IRLBA.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Seed for the initial random vector in IRLBA.</p></li>
<li><p><strong>realized</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to realize <code class="docutils literal notranslate"><span class="pre">x</span></code> into an optimal memory layout for IRLBA.
This speeds up computation at the cost of increased memory usage.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults" title="scranpy.score_gene_set.ScoreGeneSetResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ScoreGeneSetResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array of per-cell scores and per-gene weights.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/gsdecon">https://github.com/libscran/gsdecon</a>, which describes the approach in more detail.
In particular, see the documentation for the <code class="docutils literal notranslate"><span class="pre">compute_blocked</span></code> function for an explanation of the blocking strategy.</p>
</dd></dl>

</section>
<section id="module-scranpy.score_markers">
<span id="scranpy-score-markers-module"></span><h2>scranpy.score_markers module<a class="headerlink" href="#module-scranpy.score_markers" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.score_markers.</span></span><span class="sig-name descname"><span class="pre">ScoreMarkersResults</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">detected</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cohens_d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta_mean</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta_detected</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/score_markers.html#ScoreMarkersResults"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Results of <a class="reference internal" href="#scranpy.score_markers.score_markers" title="scranpy.score_markers.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'auc':</span> <span class="pre">typing.Union[numpy.ndarray,</span> <span class="pre">biocutils.NamedList.NamedList,</span> <span class="pre">NoneType],</span> <span class="pre">'cohens_d':</span> <span class="pre">typing.Union[numpy.ndarray,</span> <span class="pre">biocutils.NamedList.NamedList,</span> <span class="pre">NoneType],</span> <span class="pre">'delta_detected':</span> <span class="pre">typing.Union[numpy.ndarray,</span> <span class="pre">biocutils.NamedList.NamedList,</span> <span class="pre">NoneType],</span> <span class="pre">'delta_mean':</span> <span class="pre">typing.Union[numpy.ndarray,</span> <span class="pre">biocutils.NamedList.NamedList,</span> <span class="pre">NoneType],</span> <span class="pre">'detected':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'groups':</span> <span class="pre">&lt;class</span> <span class="pre">'list'&gt;,</span> <span class="pre">'mean':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'auc':</span> <span class="pre">Field(name='auc',type=typing.Union[numpy.ndarray,</span> <span class="pre">biocutils.NamedList.NamedList,</span> <span class="pre">NoneType],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'cohens_d':</span> <span class="pre">Field(name='cohens_d',type=typing.Union[numpy.ndarray,</span> <span class="pre">biocutils.NamedList.NamedList,</span> <span class="pre">NoneType],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'delta_detected':</span> <span class="pre">Field(name='delta_detected',type=typing.Union[numpy.ndarray,</span> <span class="pre">biocutils.NamedList.NamedList,</span> <span class="pre">NoneType],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'delta_mean':</span> <span class="pre">Field(name='delta_mean',type=typing.Union[numpy.ndarray,</span> <span class="pre">biocutils.NamedList.NamedList,</span> <span class="pre">NoneType],default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'detected':</span> <span class="pre">Field(name='detected',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'groups':</span> <span class="pre">Field(name='groups',type=&lt;class</span> <span class="pre">'list'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'mean':</span> <span class="pre">Field(name='mean',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('groups',</span> <span class="pre">'mean',</span> <span class="pre">'detected',</span> <span class="pre">'cohens_d',</span> <span class="pre">'auc',</span> <span class="pre">'delta_mean',</span> <span class="pre">'delta_detected')</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.auc">
<span class="sig-name descname"><span class="pre">auc</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">,</span> <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">,</span> <a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.auc" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as <a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.cohens_d" title="scranpy.score_markers.ScoreMarkersResults.cohens_d"><code class="xref py py-attr docutils literal notranslate"><span class="pre">cohens_d</span></code></a> but for the AUCs.
If <code class="docutils literal notranslate"><span class="pre">compute_auc</span> <span class="pre">=</span> <span class="pre">False</span></code>, this is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.cohens_d">
<span class="sig-name descname"><span class="pre">cohens_d</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">,</span> <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">,</span> <a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.cohens_d" title="Permalink to this definition">¶</a></dt>
<dd><p>If <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code>, this is a named list of <a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects" title="scranpy.summarize_effects.GroupwiseSummarizedEffects"><code class="xref py py-class docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects</span></code></a> objects.
Each object corresponds to a group in the same order as <a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.groups" title="scranpy.score_markers.ScoreMarkersResults.groups"><code class="xref py py-attr docutils literal notranslate"><span class="pre">groups</span></code></a>, and contains a summary of Cohen’s d from pairwise comparisons to all other groups.
This includes the min, mean, median, max and min-rank.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">True</span></code>, this is a 3-dimensional numeric array containing the Cohen’s d from each pairwise comparison between groups.
The extents of the first two dimensions are equal to the number of groups, while the extent of the final dimension is equal to the number of genes.
The entry <code class="docutils literal notranslate"><span class="pre">[i,</span> <span class="pre">j,</span> <span class="pre">k]</span></code> represents Cohen’s d from the comparison of group <code class="docutils literal notranslate"><span class="pre">j</span></code> over group <code class="docutils literal notranslate"><span class="pre">i</span></code> for gene <code class="docutils literal notranslate"><span class="pre">k</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">compute_cohens_d</span> <span class="pre">=</span> <span class="pre">False</span></code>, this is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.delta_detected">
<span class="sig-name descname"><span class="pre">delta_detected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">,</span> <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">,</span> <a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.delta_detected" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as <a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.cohens_d" title="scranpy.score_markers.ScoreMarkersResults.cohens_d"><code class="xref py py-attr docutils literal notranslate"><span class="pre">cohens_d</span></code></a> but for the delta-detected.
If <code class="docutils literal notranslate"><span class="pre">compute_delta_detected</span> <span class="pre">=</span> <span class="pre">False</span></code>, this is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.delta_mean">
<span class="sig-name descname"><span class="pre">delta_mean</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a><span class="pre">[</span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a><span class="pre">,</span> <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a><span class="pre">,</span> <a class="reference external" href="https://docs.python.org/3.11/library/constants.html#None" title="(in Python v3.11)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.delta_mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as <a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.cohens_d" title="scranpy.score_markers.ScoreMarkersResults.cohens_d"><code class="xref py py-attr docutils literal notranslate"><span class="pre">cohens_d</span></code></a> but for the delta-mean.
If <code class="docutils literal notranslate"><span class="pre">compute_delta_mean</span> <span class="pre">=</span> <span class="pre">False</span></code>, this is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.detected">
<span class="sig-name descname"><span class="pre">detected</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.detected" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point matrix containing the proportion of cells with detected expression for each gene in each group.
Each row is a gene and each column is a group, ordered as in <a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.groups" title="scranpy.score_markers.ScoreMarkersResults.groups"><code class="xref py py-attr docutils literal notranslate"><span class="pre">groups</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.groups">
<span class="sig-name descname"><span class="pre">groups</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.groups" title="Permalink to this definition">¶</a></dt>
<dd><p>Identities of the groups.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.mean">
<span class="sig-name descname"><span class="pre">mean</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point matrix containing the mean expression for each gene in each group.
Each row is a gene and each column is a group, ordered as in <a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.groups" title="scranpy.score_markers.ScoreMarkersResults.groups"><code class="xref py py-attr docutils literal notranslate"><span class="pre">groups</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.score_markers.ScoreMarkersResults.to_biocframes">
<span class="sig-name descname"><span class="pre">to_biocframes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">effect_sizes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summaries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_detected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/score_markers.html#ScoreMarkersResults.to_biocframes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.score_markers.ScoreMarkersResults.to_biocframes" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the effect size summaries into a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> for each group.
This should only be used if <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code> in <a class="reference internal" href="#scranpy.score_markers.score_markers" title="scranpy.score_markers.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>effect_sizes</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>]</span>) – List of effect sizes to include in each <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.
This can contain any of <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code>, <code class="docutils literal notranslate"><span class="pre">auc</span></code>, <code class="docutils literal notranslate"><span class="pre">delta_mean</span></code>, and <code class="docutils literal notranslate"><span class="pre">delta_detected</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all non-<code class="docutils literal notranslate"><span class="pre">None</span></code> effect sizes are reported.</p></li>
<li><p><strong>summaries</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>]</span>) – List of summary statistics to include in each <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.
This can contain any of <code class="docutils literal notranslate"><span class="pre">min</span></code>, <code class="docutils literal notranslate"><span class="pre">mean</span></code>, <code class="docutils literal notranslate"><span class="pre">median</span></code>, <code class="docutils literal notranslate"><span class="pre">max</span></code>, and <code class="docutils literal notranslate"><span class="pre">min_rank</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all summary statistics are reported.</p></li>
<li><p><strong>include_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to include the mean for each group.</p></li>
<li><p><strong>include_detected</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to include the detected proportion for each group.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A list of length equal to <a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.groups" title="scranpy.score_markers.ScoreMarkersResults.groups"><code class="xref py py-attr docutils literal notranslate"><span class="pre">groups</span></code></a>, containing a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with the effect size summaries for each group.
Each row of the <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> corresponds toa  gene.
Each effect size summary is represented by a column named <code class="docutils literal notranslate"><span class="pre">&lt;EFFECT&gt;_&lt;SUMMARY&gt;</span></code>.
If <code class="docutils literal notranslate"><span class="pre">include_mean</span> <span class="pre">=</span> <span class="pre">True</span></code> or <code class="docutils literal notranslate"><span class="pre">include_detected</span> <span class="pre">=</span> <span class="pre">True</span></code>, additional columns will be present with the mean and detected proportion, respectively.</p>
<p>The list itself is named according to <a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.groups" title="scranpy.score_markers.ScoreMarkersResults.groups"><code class="xref py py-attr docutils literal notranslate"><span class="pre">groups</span></code></a> if the elements can be converted to strings, otherwise it is unnamed.</p>
</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.score_markers.score_markers">
<span class="sig-prename descclassname"><span class="pre">scranpy.score_markers.</span></span><span class="sig-name descname"><span class="pre">score_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_delta_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_delta_detected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_cohens_d</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_auc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_pairwise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/score_markers.html#score_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.score_markers.score_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Score marker genes for each group using a variety of effect sizes from pairwise comparisons between groups.
This includes Cohen’s d, the area under the curve (AUC), the difference in the means (delta-mean) and the difference in the proportion of detected cells (delta-detected).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Any" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
It is typically expected to contain log-expression values, e.g., from <a class="reference internal" href="#scranpy.normalize_counts.normalize_counts" title="scranpy.normalize_counts.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>groups</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Group assignment for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
This should have length equal to the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Optional" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the block of origin (e.g., batch, sample) for each cell.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_weight_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Literal" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'variable'</span></code>, <code class="docutils literal notranslate"><span class="pre">'equal'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – Policy to use for weighting different blocks when computing the average for each statistic.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>variable_block_weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Tuple" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span>) – Parameters for variable block weighting.
This should be a tuple of length 2 where the first and second values are used as the lower and upper bounds, respectively, for the variable weight calculation.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_weight_policy</span> <span class="pre">=</span> <span class="pre">&quot;variable&quot;</span></code>.</p></li>
<li><p><strong>compute_delta_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the delta-means, i.e., the log-fold change when <code class="docutils literal notranslate"><span class="pre">x</span></code> contains log-expression values.</p></li>
<li><p><strong>compute_delta_detected</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the delta-detected, i.e., differences in the proportion of cells with detected expression.</p></li>
<li><p><strong>cohens_d</strong> – Whether to compute Cohen’s d.</p></li>
<li><p><strong>compute_auc</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the AUC.
Setting this to <code class="docutils literal notranslate"><span class="pre">False</span></code> can improve speed and memory efficiency.</p></li>
<li><p><strong>threshold</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Non-negative value specifying the minimum threshold on the differences in means (i.e., the log-fold change, if <code class="docutils literal notranslate"><span class="pre">x</span></code> contains log-expression values).
This is incorporated into the calculation for Cohen’s d and the AUC.</p></li>
<li><p><strong>all_pairwise</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to report the full effects for every pairwise comparison between groups.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, only summaries are reported.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults" title="scranpy.score_markers.ScoreMarkersResults"><code class="xref py py-class docutils literal notranslate"><span class="pre">ScoreMarkersResults</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Scores for ranking marker genes in each group, based on the effect sizes for pairwise comparisons between groups.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">score_markers_summary</span></code> and <code class="docutils literal notranslate"><span class="pre">score_markers_pairwise</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_markers">scran_markers</a> C++ library,
which describes the rationale behind the choice of effect sizes and summary statistics.</p>
</dd></dl>

</section>
<section id="module-scranpy.subsample_by_neighbors">
<span id="scranpy-subsample-by-neighbors-module"></span><h2>scranpy.subsample_by_neighbors module<a class="headerlink" href="#module-scranpy.subsample_by_neighbors" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.subsample_by_neighbors.subsample_by_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.subsample_by_neighbors.</span></span><span class="sig-name descname"><span class="pre">subsample_by_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_remaining=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/subsample_by_neighbors.html#subsample_by_neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.subsample_by_neighbors.subsample_by_neighbors" title="Permalink to this definition">¶</a></dt>
<dd><p>Subsample a dataset by selecting cells to represent all of their nearest neighbors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>]</span>) – <p>Numeric matrix where rows are dimensions and columns are cells, typically containing a low-dimensional representation from, e.g., <a class="reference internal" href="#scranpy.run_pca.run_pca" title="scranpy.run_pca.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code> object containing a pre-built search index for a dataset.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object containing pre-computed search results for a dataset.
The number of neighbors should be equal to <code class="docutils literal notranslate"><span class="pre">num_neighbors</span></code>, otherwise a warning is raised.</p>
</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to use.
Larger values result in greater downsampling.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> does not contain existing neighbor search results.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle vunknown)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Neighbor search algorithm to use.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> does not contain existing neighbor search results.</p></li>
<li><p><strong>min_remaining</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum number of remaining (i.e., unselected) neighbors that a cell must have in order to be considered for selection.
This should be less than or equal to <code class="docutils literal notranslate"><span class="pre">num_neighbors</span></code>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for the nearest-neighbor search.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> does not contain existing neighbor search results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Integer array with indices of the cells selected to be in the subsample.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/nenesub">https://github.com/libscran/nenesub</a>, for the rationale behind this approach.</p>
</dd></dl>

</section>
<section id="module-scranpy.summarize_effects">
<span id="scranpy-summarize-effects-module"></span><h2>scranpy.summarize_effects module<a class="headerlink" href="#module-scranpy.summarize_effects" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">scranpy.summarize_effects.</span></span><span class="sig-name descname"><span class="pre">GroupwiseSummarizedEffects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">median</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_rank</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/summarize_effects.html#GroupwiseSummarizedEffects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.11/library/functions.html#object" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Summarized effect sizes for a single group, typically created by <a class="reference internal" href="#scranpy.summarize_effects.summarize_effects" title="scranpy.summarize_effects.summarize_effects"><code class="xref py py-func docutils literal notranslate"><span class="pre">summarize_effects()</span></code></a> or <a class="reference internal" href="#scranpy.score_markers.score_markers" title="scranpy.score_markers.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'max':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'mean':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'median':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'min':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,</span> <span class="pre">'min_rank':</span> <span class="pre">&lt;class</span> <span class="pre">'numpy.ndarray'&gt;}</span></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__annotations__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.__dataclass_fields__">
<span class="sig-name descname"><span class="pre">__dataclass_fields__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'max':</span> <span class="pre">Field(name='max',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'mean':</span> <span class="pre">Field(name='mean',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'median':</span> <span class="pre">Field(name='median',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'min':</span> <span class="pre">Field(name='min',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD),</span> <span class="pre">'min_rank':</span> <span class="pre">Field(name='min_rank',type=&lt;class</span> <span class="pre">'numpy.ndarray'&gt;,default=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,default_factory=&lt;dataclasses._MISSING_TYPE</span> <span class="pre">object&gt;,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}</span></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__dataclass_fields__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.__dataclass_params__">
<span class="sig-name descname"><span class="pre">__dataclass_params__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_DataclassParams(init=True,repr=True,eq=True,order=False,unsafe_hash=False,frozen=False)</span></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__dataclass_params__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return self==value.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.__match_args__">
<span class="sig-name descname"><span class="pre">__match_args__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('min',</span> <span class="pre">'mean',</span> <span class="pre">'median',</span> <span class="pre">'max',</span> <span class="pre">'min_rank')</span></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__match_args__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return repr(self).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.max">
<span class="sig-name descname"><span class="pre">max</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.max" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes.
Each entry is the maximum effect size for that gene from all pairwise comparisons to other groups.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.mean">
<span class="sig-name descname"><span class="pre">mean</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes.
Each entry is the mean effect size for that gene from all pairwise comparisons to other groups.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.median">
<span class="sig-name descname"><span class="pre">median</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.median" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes.
Each entry is the median effect size for that gene from all pairwise comparisons to other groups.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.min">
<span class="sig-name descname"><span class="pre">min</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.min" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes.
Each entry is the minimum effect size for that gene from all pairwise comparisons to other groups.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.min_rank">
<span class="sig-name descname"><span class="pre">min_rank</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></em><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.min_rank" title="Permalink to this definition">¶</a></dt>
<dd><p>Floating-point array of length equal to the number of genes.
Each entry is the minimum rank of the gene from all pairwise comparisons to other groups.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scranpy.summarize_effects.GroupwiseSummarizedEffects.to_biocframe">
<span class="sig-name descname"><span class="pre">to_biocframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/summarize_effects.html#GroupwiseSummarizedEffects.to_biocframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.to_biocframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the results to a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> where each row is a gene and each column is a summary statistic.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.summarize_effects.summarize_effects">
<span class="sig-prename descclassname"><span class="pre">scranpy.summarize_effects.</span></span><span class="sig-name descname"><span class="pre">summarize_effects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">effects</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/summarize_effects.html#summarize_effects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.summarize_effects.summarize_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>For each group, summarize the effect sizes for all pairwise comparisons
to other groups. This yields a set of summary statistics that can be used
to rank marker genes for each group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>effects</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – A 3-dimensional numeric containing the effect sizes from each pairwise comparison between groups.
The extents of the first two dimensions should be equal to the number of groups, while the extent of the final dimension is equal to the number of genes.
The entry <code class="docutils literal notranslate"><span class="pre">[i,</span> <span class="pre">j,</span> <span class="pre">k]</span></code> should represent the effect size from the comparison of group <code class="docutils literal notranslate"><span class="pre">j</span></code> against group <code class="docutils literal notranslate"><span class="pre">i</span></code> for gene <code class="docutils literal notranslate"><span class="pre">k</span></code>.
See also the output of <a class="reference internal" href="#scranpy.score_markers.score_markers" title="scranpy.score_markers.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a> with <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/stdtypes.html#list" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects" title="scranpy.summarize_effects.GroupwiseSummarizedEffects"><code class="xref py py-class docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects</span></code></a>]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of length equal to the number of groups (i.e., the extents of the first two dimensions of <code class="docutils literal notranslate"><span class="pre">effects</span></code>).
Each entry contains the summary statistics of the effect sizes of the comparisons involving the corresponding group.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">summarize_effects</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_markers">scran_markers</a> C++ library, which provides the underlying implementation.</p>
</dd></dl>

</section>
<section id="module-scranpy.test_enrichment">
<span id="scranpy-test-enrichment-module"></span><h2>scranpy.test_enrichment module<a class="headerlink" href="#module-scranpy.test_enrichment" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.test_enrichment.test_enrichment">
<span class="sig-prename descclassname"><span class="pre">scranpy.test_enrichment.</span></span><span class="sig-name descname"><span class="pre">test_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/test_enrichment.html#test_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.test_enrichment.test_enrichment" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a hypergeometric test for enrichment of interesting genes (e.g., markers) in one or more pre-defined gene sets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of identifiers for the interesting genes.</p></li>
<li><p><strong>sets</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of gene sets, where each entry corresponds to a gene set and contains a sequence of identifiers for genes in that set.</p></li>
<li><p><strong>universe</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Union" title="(in Python v3.11)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.11/library/typing.html#typing.Sequence" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of identifiers for the universe of genes in the dataset.
It is expected that <code class="docutils literal notranslate"><span class="pre">x</span></code> is a subset of <code class="docutils literal notranslate"><span class="pre">universe</span></code>.
Alternatively, an integer specifying the number of genes in the universe.</p></li>
<li><p><strong>log</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#bool" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to report the log-transformed p-values.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.11/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array of (log-transformed) p-values to test for significant enrichment of <code class="docutils literal notranslate"><span class="pre">x</span></code> in each entry of <code class="docutils literal notranslate"><span class="pre">sets</span></code>.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/phyper">https://github.com/libscran/phyper</a>, for the underlying implementation.</p>
</dd></dl>

</section>
<section id="module-scranpy">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-scranpy" title="Permalink to this heading">¶</a></h2>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="modules.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">scranpy</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Jayaram Kancherla
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">scranpy package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-scranpy.adt_quality_control">scranpy.adt_quality_control module</a><ul>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.detected"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.detected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.subset_sum"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.subset_sum</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.sum"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.sum</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.ComputeAdtQcMetricsResults.to_biocframe"><code class="docutils literal notranslate"><span class="pre">ComputeAdtQcMetricsResults.to_biocframe()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.block"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.block</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.detected"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.detected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.SuggestAdtQcThresholdsResults.subset_sum"><code class="docutils literal notranslate"><span class="pre">SuggestAdtQcThresholdsResults.subset_sum</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.compute_adt_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.filter_adt_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_adt_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.adt_quality_control.suggest_adt_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.aggregate_across_cells">scranpy.aggregate_across_cells module</a><ul>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.combinations"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.combinations</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.counts"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.counts</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.detected"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.detected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.index"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.index</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.sum"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.sum</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.AggregateAcrossCellsResults.to_summarizedexperiment"><code class="docutils literal notranslate"><span class="pre">AggregateAcrossCellsResults.to_summarizedexperiment()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells.aggregate_across_cells"><code class="docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.aggregate_across_genes">scranpy.aggregate_across_genes module</a><ul>
<li><a class="reference internal" href="#scranpy.aggregate_across_genes.aggregate_across_genes"><code class="docutils literal notranslate"><span class="pre">aggregate_across_genes()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.analyze">scranpy.analyze module</a><ul>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.__eq__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.__hash__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.__repr__"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_filtered"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_filtered</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_markers"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_markers</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_normalized"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_normalized</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_pca"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_pca</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_qc_filter"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_qc_filter</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_qc_metrics"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_qc_metrics</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_qc_thresholds</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_row_names"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_row_names</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.adt_size_factors"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.adt_size_factors</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.block"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.block</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.clusters"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.clusters</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.column_names"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.column_names</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.combined_pca"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.combined_pca</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.combined_qc_filter"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.combined_qc_filter</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_filtered"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_filtered</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_markers"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_markers</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_normalized"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_normalized</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_pca"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_pca</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_qc_filter"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_qc_filter</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_qc_metrics"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_qc_metrics</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_qc_thresholds</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_row_names"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_row_names</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.crispr_size_factors"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.crispr_size_factors</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.graph_clusters"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.graph_clusters</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.kmeans_clusters"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.kmeans_clusters</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.mnn_corrected"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.mnn_corrected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_filtered"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_filtered</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_gene_variances"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_gene_variances</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_highly_variable_genes"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_highly_variable_genes</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_markers"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_markers</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_normalized"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_normalized</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_pca"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_pca</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_qc_filter"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_qc_filter</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_qc_metrics"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_qc_metrics</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_qc_thresholds</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_row_names"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_row_names</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.rna_size_factors"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.rna_size_factors</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.snn_graph"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.snn_graph</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.to_singlecellexperiment"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.to_singlecellexperiment()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.tsne"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.tsne</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze.AnalyzeResults.umap"><code class="docutils literal notranslate"><span class="pre">AnalyzeResults.umap</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.analyze.analyze"><code class="docutils literal notranslate"><span class="pre">analyze()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.build_snn_graph">scranpy.build_snn_graph module</a><ul>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents"><code class="docutils literal notranslate"><span class="pre">GraphComponents</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.__annotations__"><code class="docutils literal notranslate"><span class="pre">GraphComponents.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">GraphComponents.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">GraphComponents.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.__eq__"><code class="docutils literal notranslate"><span class="pre">GraphComponents.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.__hash__"><code class="docutils literal notranslate"><span class="pre">GraphComponents.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.__match_args__"><code class="docutils literal notranslate"><span class="pre">GraphComponents.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.__repr__"><code class="docutils literal notranslate"><span class="pre">GraphComponents.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.as_igraph"><code class="docutils literal notranslate"><span class="pre">GraphComponents.as_igraph()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.edges"><code class="docutils literal notranslate"><span class="pre">GraphComponents.edges</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.vertices"><code class="docutils literal notranslate"><span class="pre">GraphComponents.vertices</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.GraphComponents.weights"><code class="docutils literal notranslate"><span class="pre">GraphComponents.weights</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.build_snn_graph.build_snn_graph"><code class="docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.center_size_factors">scranpy.center_size_factors module</a><ul>
<li><a class="reference internal" href="#scranpy.center_size_factors.center_size_factors"><code class="docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.choose_highly_variable_genes">scranpy.choose_highly_variable_genes module</a><ul>
<li><a class="reference internal" href="#scranpy.choose_highly_variable_genes.choose_highly_variable_genes"><code class="docutils literal notranslate"><span class="pre">choose_highly_variable_genes()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.choose_pseudo_count">scranpy.choose_pseudo_count module</a><ul>
<li><a class="reference internal" href="#scranpy.choose_pseudo_count.choose_pseudo_count"><code class="docutils literal notranslate"><span class="pre">choose_pseudo_count()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.cluster_graph">scranpy.cluster_graph module</a><ul>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphLeidenResults.quality"><code class="docutils literal notranslate"><span class="pre">ClusterGraphLeidenResults.quality</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.levels"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.levels</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphMultilevelResults.modularity"><code class="docutils literal notranslate"><span class="pre">ClusterGraphMultilevelResults.modularity</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.membership"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.membership</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphResults.status"><code class="docutils literal notranslate"><span class="pre">ClusterGraphResults.status</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.merges"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.merges</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph.ClusterGraphWalktrapResults.modularity"><code class="docutils literal notranslate"><span class="pre">ClusterGraphWalktrapResults.modularity</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.cluster_graph.cluster_graph"><code class="docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.cluster_kmeans">scranpy.cluster_kmeans module</a><ul>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.centers"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.centers</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.clusters"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.clusters</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.iterations"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.iterations</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.ClusterKmeansResults.status"><code class="docutils literal notranslate"><span class="pre">ClusterKmeansResults.status</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans.cluster_kmeans"><code class="docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.combine_factors">scranpy.combine_factors module</a><ul>
<li><a class="reference internal" href="#scranpy.combine_factors.combine_factors"><code class="docutils literal notranslate"><span class="pre">combine_factors()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.compute_clrm1_factors">scranpy.compute_clrm1_factors module</a><ul>
<li><a class="reference internal" href="#scranpy.compute_clrm1_factors.compute_clrm1_factors"><code class="docutils literal notranslate"><span class="pre">compute_clrm1_factors()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.correct_mnn">scranpy.correct_mnn module</a><ul>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.__eq__"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.__hash__"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.__repr__"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.corrected"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.corrected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.merge_order"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.merge_order</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn.CorrectMnnResults.num_pairs"><code class="docutils literal notranslate"><span class="pre">CorrectMnnResults.num_pairs</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.correct_mnn.correct_mnn"><code class="docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.crispr_quality_control">scranpy.crispr_quality_control module</a><ul>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.detected"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.detected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.max_index"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.max_index</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.max_value"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.max_value</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.sum"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.sum</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.ComputeCrisprQcMetricsResults.to_biocframe"><code class="docutils literal notranslate"><span class="pre">ComputeCrisprQcMetricsResults.to_biocframe()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.block"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.block</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.SuggestCrisprQcThresholdsResults.max_value"><code class="docutils literal notranslate"><span class="pre">SuggestCrisprQcThresholdsResults.max_value</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.compute_crispr_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.filter_crispr_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_crispr_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.crispr_quality_control.suggest_crispr_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.fit_variance_trend">scranpy.fit_variance_trend module</a><ul>
<li><a class="reference internal" href="#scranpy.fit_variance_trend.fit_variance_trend"><code class="docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.lib_scranpy">scranpy.lib_scranpy module</a><ul>
<li><a class="reference internal" href="#scranpy.lib_scranpy.aggregate_across_cells"><code class="docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.aggregate_across_genes"><code class="docutils literal notranslate"><span class="pre">aggregate_across_genes()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.build_snn_graph"><code class="docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.center_size_factors"><code class="docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.choose_highly_variable_genes"><code class="docutils literal notranslate"><span class="pre">choose_highly_variable_genes()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.choose_pseudo_count"><code class="docutils literal notranslate"><span class="pre">choose_pseudo_count()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.cluster_kmeans"><code class="docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.cluster_leiden"><code class="docutils literal notranslate"><span class="pre">cluster_leiden()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.cluster_multilevel"><code class="docutils literal notranslate"><span class="pre">cluster_multilevel()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.cluster_walktrap"><code class="docutils literal notranslate"><span class="pre">cluster_walktrap()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.combine_factors"><code class="docutils literal notranslate"><span class="pre">combine_factors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.compute_adt_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.compute_clrm1_factors"><code class="docutils literal notranslate"><span class="pre">compute_clrm1_factors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.compute_crispr_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.compute_rna_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.correct_mnn"><code class="docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.filter_adt_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_adt_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.filter_crispr_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_crispr_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.filter_rna_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_rna_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.fit_variance_trend"><code class="docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.model_gene_variances"><code class="docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.normalize_counts"><code class="docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.perplexity_to_neighbors"><code class="docutils literal notranslate"><span class="pre">perplexity_to_neighbors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.run_pca"><code class="docutils literal notranslate"><span class="pre">run_pca()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.run_tsne"><code class="docutils literal notranslate"><span class="pre">run_tsne()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.run_umap"><code class="docutils literal notranslate"><span class="pre">run_umap()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.sanitize_size_factors"><code class="docutils literal notranslate"><span class="pre">sanitize_size_factors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.scale_by_neighbors"><code class="docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.score_gene_set"><code class="docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.score_markers_pairwise"><code class="docutils literal notranslate"><span class="pre">score_markers_pairwise()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.score_markers_summary"><code class="docutils literal notranslate"><span class="pre">score_markers_summary()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.subsample_by_neighbors"><code class="docutils literal notranslate"><span class="pre">subsample_by_neighbors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.suggest_adt_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.suggest_crispr_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.suggest_rna_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.summarize_effects"><code class="docutils literal notranslate"><span class="pre">summarize_effects()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.lib_scranpy.test_enrichment"><code class="docutils literal notranslate"><span class="pre">test_enrichment()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.model_gene_variances">scranpy.model_gene_variances module</a><ul>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.fitted"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.fitted</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.mean"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.mean</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.per_block"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.per_block</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.residual"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.residual</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.to_biocframe"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.to_biocframe()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.ModelGeneVariancesResults.variance"><code class="docutils literal notranslate"><span class="pre">ModelGeneVariancesResults.variance</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.model_gene_variances.model_gene_variances"><code class="docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.normalize_counts">scranpy.normalize_counts module</a><ul>
<li><a class="reference internal" href="#scranpy.normalize_counts.normalize_counts"><code class="docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.rna_quality_control">scranpy.rna_quality_control module</a><ul>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.detected"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.detected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.subset_proportion"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.subset_proportion</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.sum"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.sum</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.ComputeRnaQcMetricsResults.to_biocframe"><code class="docutils literal notranslate"><span class="pre">ComputeRnaQcMetricsResults.to_biocframe()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.block"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.block</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.detected"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.detected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.subset_proportion"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.subset_proportion</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.SuggestRnaQcThresholdsResults.sum"><code class="docutils literal notranslate"><span class="pre">SuggestRnaQcThresholdsResults.sum</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.compute_rna_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.filter_rna_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_rna_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.rna_quality_control.suggest_rna_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.run_all_neighbor_steps">scranpy.run_all_neighbor_steps module</a><ul>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.build_snn_graph"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.build_snn_graph</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.cluster_graph"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.cluster_graph</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.run_tsne"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.run_tsne</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.RunAllNeighborStepsResults.run_umap"><code class="docutils literal notranslate"><span class="pre">RunAllNeighborStepsResults.run_umap</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps.run_all_neighbor_steps"><code class="docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.run_pca">scranpy.run_pca module</a><ul>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults"><code class="docutils literal notranslate"><span class="pre">RunPcaResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.__eq__"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.__hash__"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.__repr__"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.block"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.block</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.center"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.center</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.components"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.components</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.rotation"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.rotation</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.scale"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.scale</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.total_variance"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.total_variance</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca.RunPcaResults.variance_explained"><code class="docutils literal notranslate"><span class="pre">RunPcaResults.variance_explained</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.run_pca.run_pca"><code class="docutils literal notranslate"><span class="pre">run_pca()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.run_tsne">scranpy.run_tsne module</a><ul>
<li><a class="reference internal" href="#scranpy.run_tsne.run_tsne"><code class="docutils literal notranslate"><span class="pre">run_tsne()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_tsne.tsne_perplexity_to_neighbors"><code class="docutils literal notranslate"><span class="pre">tsne_perplexity_to_neighbors()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.run_umap">scranpy.run_umap module</a><ul>
<li><a class="reference internal" href="#scranpy.run_umap.run_umap"><code class="docutils literal notranslate"><span class="pre">run_umap()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.sanitize_size_factors">scranpy.sanitize_size_factors module</a><ul>
<li><a class="reference internal" href="#scranpy.sanitize_size_factors.sanitize_size_factors"><code class="docutils literal notranslate"><span class="pre">sanitize_size_factors()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.scale_by_neighbors">scranpy.scale_by_neighbors module</a><ul>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.combined"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.combined</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.ScaleByNeighborsResults.scaling"><code class="docutils literal notranslate"><span class="pre">ScaleByNeighborsResults.scaling</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors.scale_by_neighbors"><code class="docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.score_gene_set">scranpy.score_gene_set module</a><ul>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.scores"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.scores</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set.ScoreGeneSetResults.weights"><code class="docutils literal notranslate"><span class="pre">ScoreGeneSetResults.weights</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.score_gene_set.score_gene_set"><code class="docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.score_markers">scranpy.score_markers module</a><ul>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.__annotations__"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.__eq__"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.__hash__"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.__match_args__"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.__repr__"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.auc"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.auc</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.cohens_d"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.cohens_d</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.delta_detected"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.delta_detected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.delta_mean"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.delta_mean</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.detected"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.detected</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.groups"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.groups</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.mean"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.mean</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers.ScoreMarkersResults.to_biocframes"><code class="docutils literal notranslate"><span class="pre">ScoreMarkersResults.to_biocframes()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.score_markers.score_markers"><code class="docutils literal notranslate"><span class="pre">score_markers()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.subsample_by_neighbors">scranpy.subsample_by_neighbors module</a><ul>
<li><a class="reference internal" href="#scranpy.subsample_by_neighbors.subsample_by_neighbors"><code class="docutils literal notranslate"><span class="pre">subsample_by_neighbors()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.summarize_effects">scranpy.summarize_effects module</a><ul>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects</span></code></a><ul>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__annotations__"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__dataclass_fields__"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.__dataclass_fields__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__dataclass_params__"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.__dataclass_params__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__eq__"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.__eq__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__hash__"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.__hash__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__match_args__"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.__match_args__</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.__repr__"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.max"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.max</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.mean"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.mean</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.median"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.median</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.min"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.min</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.min_rank"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.min_rank</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects.GroupwiseSummarizedEffects.to_biocframe"><code class="docutils literal notranslate"><span class="pre">GroupwiseSummarizedEffects.to_biocframe()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#scranpy.summarize_effects.summarize_effects"><code class="docutils literal notranslate"><span class="pre">summarize_effects()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy.test_enrichment">scranpy.test_enrichment module</a><ul>
<li><a class="reference internal" href="#scranpy.test_enrichment.test_enrichment"><code class="docutils literal notranslate"><span class="pre">test_enrichment()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-scranpy">Module contents</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=b1f64a84"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>